--- sys/netinet/ip_input.c.old	2011-08-19 20:43:50.000000000 +0000
+++ sys/netinet/ip_input.c	2011-08-19 20:44:07.000000000 +0000
@@ -554,6 +554,7 @@
 		m->m_flags &= ~M_FASTFWD_OURS;
 		goto ours;
 	}
+#if 0
 	if (m_tag_find(m, PACKET_TAG_IPFORWARD, NULL) != NULL) {
 		/*
 		 * Directly ship the packet on.  This allows forwarding
@@ -563,6 +564,7 @@
 		ip_forward(m, 1);
 		return;
 	}
+#endif
 #endif /* IPFIREWALL_FORWARD */
 
 passin:
--- sys/netinet/ip_output.c.old	2011-08-19 20:44:18.000000000 +0000
+++ sys/netinet/ip_output.c	2011-08-19 20:44:57.000000000 +0000
@@ -566,6 +566,7 @@
 		error = netisr_queue(NETISR_IP, m);
 		goto done;
 	}
+#if 0
 	/* Or forward to some other address? */
 	fwd_tag = m_tag_find(m, PACKET_TAG_IPFORWARD, NULL);
 	if (fwd_tag) {
@@ -575,6 +576,7 @@
 		m_tag_delete(m, fwd_tag);
 		goto again;
 	}
+#endif
 #endif /* IPFIREWALL_FORWARD */
 
 passout:
