--- pf.c.orig	2012-01-27 20:48:41.000000000 +0000
+++ pf.c	2012-01-27 20:59:44.000000000 +0000
@@ -3599,6 +3599,8 @@
 		if (nr != NULL)
 			s->log |= nr->log & PF_LOG_ALL;
 		s->proto = IPPROTO_TCP;
+		s->qid = act.qid;
+		s->pqid = act.pqid;
 		s->direction = direction;
 		s->af = af;
 		if (direction == PF_OUT) {
@@ -4015,6 +4017,8 @@
 		if (nr != NULL)
 			s->log |= nr->log & PF_LOG_ALL;
 		s->proto = IPPROTO_UDP;
+		s->qid = act.qid;
+		s->pqid = act.pqid;
 		s->direction = direction;
 		s->af = af;
 		if (direction == PF_OUT) {
@@ -7278,33 +7282,32 @@
         u_int8_t		 divflags = 0;
 	int			 loopedfrom = 0;
 
-	PF_LOCK();
-#endif
-	if (!pf_status.running)
-#ifdef __FreeBSD__
-	{
-		PF_UNLOCK();
-#endif
+	if (m->m_flags & M_SKIP_FIREWALL)
 		return (PF_PASS);
-#ifdef __FreeBSD__
-	}
-#endif
 
+#endif
 	memset(&pd, 0, sizeof(pd));
 	if ((pd.pf_mtag = pf_get_mtag(m)) == NULL) {
-#ifdef __FreeBSD__
-		PF_UNLOCK();
-#endif
 		DPFPRINTF(PF_DEBUG_URGENT,
 		    ("pf_test: pf_get_mtag returned NULL\n"));
 		return (PF_DROP);
 	}
+
+#ifdef __FreeBSD__
+	PF_LOCK();
+#endif
+
+	if (!pf_status.running)
 #ifdef __FreeBSD__
-	if (m->m_flags & M_SKIP_FIREWALL) {
+	{
 		PF_UNLOCK();
+#endif
 		return (PF_PASS);
+#ifdef __FreeBSD__
 	}
-#else
+#endif
+
+#if 0
 	if (pd.pf_mtag->flags & PF_TAG_GENERATED)
 		return (PF_PASS);
 #endif
