Index: dhclient.c
===================================================================
RCS file: /home/ncvs/src/sbin/dhclient/dhclient.c,v
retrieving revision 1.21.2.3.4.1
diff -u -r1.21.2.3.4.1 dhclient.c
--- dhclient.c	15 Apr 2009 03:14:26 -0000	1.21.2.3.4.1
+++ dhclient.c	9 Jun 2009 22:40:17 -0000
@@ -1977,7 +1977,7 @@
 					len = ip->client->
 					    config->defaults[i].len +
 					    lease->options[i].len;
-					if (len > sizeof(dbuf)) {
+					if (len >= sizeof(dbuf)) {
 						warning("no space to %s %s",
 						    "prepend option",
 						    dhcp_options[i].name);
@@ -1996,24 +1996,35 @@
 					dp[len] = '\0';
 					break;
 				case ACTION_APPEND:
+                                       /*
+                                        * When we append, we assume that we're
+                                        * appending to text.  Some MS servers
+                                        * include a NUL byte at the end of
+                                        * the search string provided.
+                                        */
 					len = ip->client->
 					    config->defaults[i].len +
 					    lease->options[i].len;
-					if (len > sizeof(dbuf)) {
+					if (len >= sizeof(dbuf)) {
 						warning("no space to %s %s",
 						    "append option",
 						    dhcp_options[i].name);
 						goto supersede;
 					}
 					dp = dbuf;
-					memcpy(dp,
+					memcpy(dbuf,
 						lease->options[i].data,
 						lease->options[i].len);
-					memcpy(dp + lease->options[i].len,
+                                       for (dp = dbuf + lease->options[i].len;
+                                           dp > dbuf; dp--, len--)
+                                               if (dp[-1] != '\0')
+                                                       break;
+					memcpy(dp,
 						ip->client->
 						config->defaults[i].data,
 						ip->client->
 						config->defaults[i].len);
+					dp = dbuf;
 					dp[len] = '\0';
 				}
 			} else {
