Index: if.c
===================================================================
RCS file: /home/ncvs/src/sys/net/if.c,v
retrieving revision 1.273.2.11.2.1
diff -u -r1.273.2.11.2.1 if.c
--- if.c	15 Apr 2009 03:14:26 -0000	1.273.2.11.2.1
+++ if.c	5 Sep 2009 20:41:13 -0000
@@ -1123,6 +1123,7 @@
 static int
 if_rtdel(struct radix_node *rn, void *arg)
 {
+	return (0);
 	struct rtentry	*rt = (struct rtentry *)rn;
 	struct ifnet	*ifp = arg;
 	int		err;
