Index: if_ether.c
===================================================================
RCS file: /home/ncvs/src/sys/netinet/if_ether.c,v
retrieving revision 1.162.2.6.2.2
diff -u -r1.162.2.6.2.2 if_ether.c
--- if_ether.c	20 Dec 2008 19:45:22 -0000	1.162.2.6.2.2
+++ if_ether.c	27 Mar 2009 21:41:31 -0000
@@ -195,10 +195,10 @@
 	case RTM_RESOLVE:
 		if (gate->sa_family != AF_LINK ||
 		    gate->sa_len < sizeof(null_sdl)) {
-			log(LOG_DEBUG, "%s: bad gateway %s%s\n", __func__,
-			    inet_ntoa(SIN(rt_key(rt))->sin_addr),
-			    (gate->sa_family != AF_LINK) ?
-			    " (!AF_LINK)": "");
+			//log(LOG_DEBUG, "%s: bad gateway %s%s\n", __func__,
+			//    inet_ntoa(SIN(rt_key(rt))->sin_addr),
+			//    (gate->sa_family != AF_LINK) ?
+			//    " (!AF_LINK)": "");
 			break;
 		}
 		SDL(gate)->sdl_type = rt->rt_ifp->if_type;
