Index: kern_conf.c
===================================================================
--- kern_conf.c	(revision 192098)
+++ kern_conf.c	(working copy)
@@ -529,7 +529,9 @@
 	if (cold)
 		return;
 	namelen = strlen(dev->si_name);
-	data = malloc(namelen + sizeof(prefix), M_TEMP, M_WAITOK);
+	data = malloc(namelen + sizeof(prefix), M_TEMP, M_NOWAIT);
+	if (data == NULL)
+		return;
 	memcpy(data, prefix, sizeof(prefix) - 1);
 	memcpy(data + sizeof(prefix) - 1, dev->si_name, namelen + 1);
 	devctl_notify("DEVFS", "CDEV", ev, data);
