Index: if.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/if.c,v
retrieving revision 1.69.2.1.4.1
diff -u -r1.69.2.1.4.1 if.c
--- if.c	15 Apr 2009 03:14:26 -0000	1.69.2.1.4.1
+++ if.c	10 Jan 2010 20:07:54 -0000
@@ -683,6 +683,8 @@
 	sigsetmask(oldmask);
 	line++;
 	first = 0;
+	if (qflag && (qval-- == 0))
+		exit(0);
 	if (line == 21)
 		goto banner;
 	else
Index: main.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/main.c,v
retrieving revision 1.87.2.4.2.1
diff -u -r1.87.2.4.2.1 main.c
--- main.c	15 Apr 2009 03:14:26 -0000	1.87.2.4.2.1
+++ main.c	10 Jan 2010 20:07:54 -0000
@@ -329,6 +329,8 @@
 int	numeric_addr;	/* show addresses numerically */
 int	numeric_port;	/* show ports numerically */
 static int pflag;	/* show given protocol */
+int	qflag;		/* whether to quit after specified outputs */
+int	qval;		/* how much outputs before we exit */
 int	rflag;		/* show routing tables (or routing stats) */
 int	sflag;		/* show protocol statistics */
 int	tflag;		/* show i/f watchdog timers */
@@ -351,7 +353,7 @@
 
 	af = AF_UNSPEC;
 
-	while ((ch = getopt(argc, argv, "AaBbdf:ghI:iLlM:mN:np:rSstuWw:z")) != -1)
+	while ((ch = getopt(argc, argv, "AaBbdf:ghI:iLlM:mN:np:q:rSstuWw:z")) != -1)
 		switch(ch) {
 		case 'A':
 			Aflag = 1;
@@ -437,6 +439,10 @@
 			}
 			pflag = 1;
 			break;
+		case 'q':
+			qval = atoi(optarg);
+			qflag = 1;
+			break;
 		case 'r':
 			rflag = 1;
 			break;
Index: netstat.1
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/netstat.1,v
retrieving revision 1.58.8.1
diff -u -r1.58.8.1 netstat.1
--- netstat.1	15 Apr 2009 03:14:26 -0000	1.58.8.1
+++ netstat.1	10 Jan 2010 20:07:54 -0000
@@ -130,6 +130,7 @@
 .Nm
 .Fl w Ar wait
 .Op Fl I Ar interface
+.Op Fl q Ar howmany
 .Op Fl d
 .Op Fl M Ar core
 .Op Fl N Ar system
@@ -143,6 +144,11 @@
 or a single
 .Ar interface .
 If
+.Fl q
+is also present, exit after
+.Ar howmany
+outputs.
+If
 .Fl d
 is also present, show the number of dropped packets.
 .It Xo
Index: netstat.h
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/netstat.h,v
retrieving revision 1.51.2.3.2.1
diff -u -r1.51.2.3.2.1 netstat.h
--- netstat.h	15 Apr 2009 03:14:26 -0000	1.51.2.3.2.1
+++ netstat.h	10 Jan 2010 20:07:54 -0000
@@ -47,6 +47,8 @@
 extern int	mflag;	/* show memory stats */
 extern int	numeric_addr;	/* show addresses numerically */
 extern int	numeric_port;	/* show ports numerically */
+extern int	qflag;	/* whether to quit after specified outputs */
+extern int	qval;	/* how much outputs before we exit */
 extern int	rflag;	/* show routing tables (or routing stats) */
 extern int	sflag;	/* show protocol statistics */
 extern int	tflag;	/* show i/f watchdog timers */
