Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /root/pfsensesrc/7/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.2
diff -u -r1.2 pf.c
--- sys/contrib/pf/net/pf.c	14 Oct 2009 15:05:29 -0000	1.2
+++ sys/contrib/pf/net/pf.c	14 Oct 2009 15:50:29 -0000
@@ -6237,6 +6237,9 @@
 
 	ip = mtod(m0, struct ip *);
 
+	if (IN_MULTICAST(ntohl(ip->ip_dst.s_addr)))
+                return;
+
 	ro = &iproute;
 	bzero((caddr_t)ro, sizeof(*ro));
 	dst = satosin(&ro->ro_dst);
@@ -6278,6 +6281,21 @@
 	}
 	if (ifp == NULL)
 		goto bad;
+	else if (r->rt == PF_REPLYTO) {
+		register u_long i = ntohl(ip->ip_dst);
+		struct in_ifaddr *ia;
+		/* XXX: copied from in_localaddr. */
+		TAILQ_FOREACH(ia, &in_ifaddrhead, ia_link) {
+			if ((i & ia->ia_subnetmask) == ia->ia_subnet) {
+				if (ia->ia_ifp == ifp)
+					goto donelocal;
+				break;
+			}
+		}
+	} else if (r->rt == PF_ROUTETO && r->direction == dir && in_localip(ip->ip_dst))
+		goto donelocal;
+	else if (in_broadcast(ip->ip_dst, oifp)) /* XXX: LOCKING of address list?! */
+		goto donelocal;
 
 	if (oifp != ifp) {
 #ifdef __FreeBSD__
@@ -6459,6 +6477,7 @@
 done:
 	if (r->rt != PF_DUPTO)
 		*m = NULL;
+donelocal:
 	if (ro == &iproute && ro->ro_rt)
 		RTFREE(ro->ro_rt);
 	return;
