--- vfs_mount.c.orig	2008-11-24 21:59:29.000000000 -0500
+++ vfs_mount.c	2009-02-19 18:28:46.000000000 -0500
@@ -58,6 +58,7 @@
 #include <sys/sysent.h>
 #include <sys/systm.h>
 #include <sys/vnode.h>
+#include <sys/cons.h>
 #include <vm/uma.h>
 
 #include <geom/geom.h>
@@ -1611,7 +1612,11 @@
 vfs_mountroot(void)
 {
 	char *cp;
-	int error, i, asked = 0;
+	const char *rootdevname_orig;
+	int error;
+	unsigned int i;
+	unsigned char asked = 0; /* set if asked for mount point */
+	unsigned char timeout = 16; /* seconds */
 
 	root_mount_prepare();
 
@@ -1629,6 +1634,10 @@
 		asked = 1;
 	}
 
+	/* store a copy of the initial root device name */
+	rootdevname_orig = ctrootdevname;
+retry:
+
 	/*
 	 * The root filesystem information is compiled in, and we are
 	 * booted with instructions to use it.
@@ -1678,6 +1687,20 @@
 	if (ctrootdevname != NULL)
 		if (!vfs_mountroot_try(ctrootdevname))
 			goto mounted;
+
+	/*
+	* Check if we should try more times.
+	*/
+       if (timeout != 0) {
+               timeout--;
+               pause("WROOT", hz);
+               if (cncheckc() == -1) {
+                       /* no key press - try again */
+                       ctrootdevname = rootdevname_orig;
+                       goto retry;
+               }
+       }
+
 	/*
 	 * Everything so far has failed, prompt on the console if we haven't
 	 * already tried that.
