--- MakefileORIG	Thu Sep  1 19:20:21 2005
+++ Makefile	Thu Sep  1 19:25:47 2005
@@ -23,7 +23,6 @@
 CFLAGS=	-Os \
 	-fno-guess-branch-probability \
 	-fomit-frame-pointer \
-	-fno-unit-at-a-time \
 	-mno-align-long-strings \
 	-mrtd \
 	-mno-mmx -mno-3dnow -mno-sse -mno-sse2 \
--- boot2.c.orig	Mon Aug 15 06:14:07 2005
+++ boot2.c	Thu Sep  1 19:25:47 2005
@@ -139,8 +139,8 @@
 static int xgetc(int);
 static int getc(int);
 
-static void memcpy(void *, const void *, int);
-static void
+void memcpy(void *, const void *, int);
+void
 memcpy(void *dst, const void *src, int len)
 {
     const char *s = src;
