Index: fdisk.c
===================================================================
RCS file: /home/freebsd-cvs/src/sbin/fdisk/fdisk.c,v
retrieving revision 1.81
diff -u -r1.81 fdisk.c
--- fdisk.c	17 Sep 2005 14:56:10 -0000	1.81
+++ fdisk.c	2 Oct 2005 06:58:47 -0000
@@ -1043,9 +1043,8 @@
 	}
 	    if (dos_cyls > 1024) {
 	    warnx(
-	"WARNING line %d: number of cylinders (%d) may be out-of-range\n\
-    (must be within 1-1024 for normal BIOS operation, unless the entire disk\n\
-    is dedicated to FreeBSD)",
+	"WARNING line %d: number of cylinders (%d) may be out-of-range\n"
+	"    (must be within 1-1024 for normal BIOS operation)",
 		    current_line_number, dos_cyls);
 	}
 
@@ -1054,9 +1053,10 @@
 		    current_line_number);
 	    status = 0;
 	    } else if (dos_heads > 256) {
-	    warnx("ERROR line %d: number of heads must be within (1-256)",
-		    current_line_number);
-	    status = 0;
+	    warnx(
+		"WARNING line %d: number of heads (%d) may be out-of-range\n"
+		"    (must be within 1-256 for normal BIOS operation)",
+		    current_line_number, dos_heads);
 	}
 
 	    if (dos_sectors == 0) {
@@ -1064,9 +1064,10 @@
 		    current_line_number);
 	    status = 0;
 	    } else if (dos_sectors > 63) {
-	    warnx("ERROR line %d: number of sectors must be within (1-63)",
-		    current_line_number);
-	    status = 0;
+	    warnx(
+		"WARNING line %d: number of sectors (%d) may be out-of-range\n"
+		"    (must be within 1-63 for normal BIOS operation)",
+		    current_line_number, dos_sectors);
 	}
 
 	break;
