--- if_em.c.orig	Thu Apr  6 00:42:14 2006
+++ if_em.c	Thu Apr  6 00:42:33 2006
@@ -1038,7 +1038,8 @@
 	INIT_DEBUGOUT("em_media_status: begin");
 
 	em_check_for_link(&adapter->hw);
-	if (E1000_READ_REG(&adapter->hw, STATUS) & E1000_STATUS_LU) {
+       if (adapter->hw.media_type != em_media_type_fiber &&
+           (E1000_READ_REG(&adapter->hw, STATUS) & E1000_STATUS_LU))
 		if (adapter->link_active == 0) {
 			em_get_speed_and_duplex(&adapter->hw, 
 						&adapter->link_speed, 
