Index: mibII_interfaces.c
===================================================================
RCS file: /home/ncvs/src/contrib/bsnmp/snmp_mibII/mibII_interfaces.c,v
retrieving revision 1.1.1.6
diff -u -r1.1.1.6 mibII_interfaces.c
--- mibII_interfaces.c	23 May 2005 11:19:09 -0000	1.1.1.6
+++ mibII_interfaces.c	23 Oct 2005 21:00:34 -0000
@@ -280,8 +280,11 @@
 		break;
 
 	  case LEAF_ifOperStatus:
-		value->v.integer =
-		    (ifp->mib.ifmd_flags & IFF_RUNNING) ? 1 : 2;
+		if (ifp->mib.ifmd_data.ifi_link_state == LINK_STATE_UP &&
+		    (ifp->mib.ifmd_flags & IFF_RUNNING) != 0)
+			value->v.integer = 1;
+		else
+			value->v.integer = 2;
 		break;
 
 	  case LEAF_ifLastChange:
