--- pf_ioctl.c.orig	Fri May 12 15:47:31 2006
+++ pf_ioctl.c	Fri May 12 15:47:49 2006
@@ -1463,14 +1463,10 @@
 			break;
 		}
 		if (pr->ticket != ruleset->rules[rs_num].inactive.ticket) {
-			printf("ticket: %d != [%d]%d\n", pr->ticket,
-			    rs_num, ruleset->rules[rs_num].inactive.ticket);
 			error = EBUSY;
 			break;
 		}
 		if (pr->pool_ticket != ticket_pabuf) {
-			printf("pool_ticket: %d != %d\n", pr->pool_ticket,
-			    ticket_pabuf);
 			error = EBUSY;
 			break;
 		}
