--- usr.sbin/bsnmpd/modules/snmp_pf/pf_tree.def.orig	Sun Dec 18 14:37:25 2005
+++ usr.sbin/bsnmpd/modules/snmp_pf/pf_tree.def	Sun Dec 18 14:37:44 2005
@@ -155,22 +155,22 @@
 		    (20 pfTablesTblPktsOutXPass COUNTER64 GET)
 		  )
 		)
-		(3 pfTablesAddrTable
-		  (1 pfTablesAddrEntry : INTEGER32 pf_tbladdr
-		    (1 pfTablesAddrIndex INTEGER32)
-		    (2 pfTablesAddrNet IPADDRESS GET)
-		    (3 pfTablesAddrMask INTEGER32 GET)
-		    (4 pfTablesAddrTZero TIMETICKS GET)
-		    (5 pfTablesAddrBytesInPass COUNTER64 GET)
-		    (6 pfTablesAddrBytesInBlock COUNTER64 GET)
-		    (7 pfTablesAddrBytesOutPass COUNTER64 GET)
-		    (8 pfTablesAddrBytesOutBlock COUNTER64 GET)
-		    (9 pfTablesAddrPktsInPass COUNTER64 GET)
-		    (10 pfTablesAddrPktsInBlock COUNTER64 GET)
-		    (11 pfTablesAddrPktsOutPass COUNTER64 GET)
-		    (12 pfTablesAddrPktsOutBlock COUNTER64 GET)
-		  )
-		)
+#		(3 pfTablesAddrTable
+#		  (1 pfTablesAddrEntry : INTEGER32 pf_tbladdr
+#		    (1 pfTablesAddrIndex INTEGER32)
+#		    (2 pfTablesAddrNet IPADDRESS GET)
+#		    (3 pfTablesAddrMask INTEGER32 GET)
+#		    (4 pfTablesAddrTZero TIMETICKS GET)
+#		    (5 pfTablesAddrBytesInPass COUNTER64 GET)
+#		    (6 pfTablesAddrBytesInBlock COUNTER64 GET)
+#		    (7 pfTablesAddrBytesOutPass COUNTER64 GET)
+#		    (8 pfTablesAddrBytesOutBlock COUNTER64 GET)
+#		    (9 pfTablesAddrPktsInPass COUNTER64 GET)
+#		    (10 pfTablesAddrPktsInBlock COUNTER64 GET)
+#		    (11 pfTablesAddrPktsOutPass COUNTER64 GET)
+#		    (12 pfTablesAddrPktsOutBlock COUNTER64 GET)
+#		  )
+#		)
 	      )
 	      (10 pfAltq
 	        (1 pfAltqQueueNumber INTEGER32 pf_altq GET)
--- usr.sbin/bsnmpd/modules/snmp_pf/pf_snmp.c.orig	Sun Dec 18 14:36:49 2005
+++ usr.sbin/bsnmpd/modules/snmp_pf/pf_snmp.c	Sun Dec 18 14:37:07 2005
@@ -772,12 +772,14 @@
 	return (SNMP_ERR_NOERROR);
 }
 
+/*
 int
 pf_tbladdr(struct snmp_context __unused *ctx, struct snmp_value __unused *val,
 	u_int __unused sub, u_int __unused vindex, enum snmp_op __unused op)
 {
 	return (SNMP_ERR_GENERR);
 }
+*/
 
 int
 pf_altq(struct snmp_context __unused *ctx, struct snmp_value *val,
