--- ipsec_output.c.afterenc	Sat May 20 19:55:17 2006
+++ ipsec_output.c	Sat May 20 19:55:55 2006
@@ -357,6 +357,7 @@
 	IPSECREQUEST_LOCK(isr);		/* insure SA contents don't change */
 
 #ifdef DEV_ENC
+	sav = isr->sav;
 	if (encif->if_bpf) {
 		/*
 		 * We need to prepend the address family as
@@ -562,6 +563,7 @@
 	int flags,
 	int *tun)
 {
+	struct secasvar *sav;
 	struct ipsecrequest *isr;
 	struct secasindex saidx;
 	int error = 0;
@@ -589,6 +591,7 @@
 	m = state->m;
 
 #ifdef DEV_ENC
+	sav = isr->sav;
 	if (encif->if_bpf) {
 		/*
 		 * We need to prepend the address family as
