Index: if_ethersubr.c
===================================================================
RCS file: /usr/store/mlaier/fcvs/src/sys/net/if_ethersubr.c,v
retrieving revision 1.196
diff -u -r1.196 if_ethersubr.c
--- if_ethersubr.c	9 Aug 2005 10:19:58 -0000	1.196
+++ if_ethersubr.c	15 Aug 2005 15:14:34 -0000
@@ -310,7 +310,8 @@
 	 * on the wire). However, we don't do that here for security
 	 * reasons and compatibility with the original behavior.
 	 */
-	if ((ifp->if_flags & IFF_SIMPLEX) && (loop_copy != -1)) {
+	if ((ifp->if_flags & IFF_SIMPLEX) && (loop_copy != -1) &&
+	    m_tag_find(m, PACKET_TAG_PF_ROUTED, NULL) == NULL) {
 		int csum_flags = 0;
 
 		if (m->m_pkthdr.csum_flags & CSUM_IP)


