Index: tcp_input.c
===================================================================
RCS file: /home/ncvs/src/sys/netinet/tcp_input.c,v
retrieving revision 1.281.2.5
diff -u -r1.281.2.5 tcp_input.c
--- tcp_input.c	1 Mar 2006 21:13:29 -0000	1.281.2.5
+++ tcp_input.c	19 Jun 2006 13:43:46 -0000
@@ -1771,6 +1771,8 @@
 		if (tp->t_state == TCPS_SYN_RECEIVED ||
 		    (tp->t_flags & TF_NEEDSYN))
 			goto step6;
+		else if (tp->t_flags & TF_ACKNOW)
+			goto dropafterack;
 		else
 			goto drop;
 	}
