Index: opencrypto/xform.c
===================================================================
RCS file: /usr/repo/src/sys/opencrypto/xform.c,v
retrieving revision 1.5
diff -u -p -r1.5 xform.c
--- opencrypto/xform.c	11 Mar 2005 17:24:46 -0000	1.5
+++ opencrypto/xform.c	16 Aug 2005 18:39:35 -0000
@@ -512,6 +512,8 @@ rijndael128_setkey(u_int8_t **sched, u_i
 {
 	int err;
 
+	if (len != 16 && len != 24 && len != 32)
+		return (EINVAL);
 	MALLOC(*sched, u_int8_t *, sizeof(rijndael_ctx), M_CRYPTO_DATA,
 	    M_NOWAIT|M_ZERO);
 	if (*sched != NULL) {
