--- work/ipsec-tools-0.8.0/src/racoon/isakmp_cfg.c	2010-09-21 15:14:17.000000000 +0200
+++ src/racoon/isakmp_cfg.c	2012-01-18 18:39:06.000000000 +0100
@@ -1,4 +1,4 @@
-/*	$NetBSD: isakmp_cfg.c,v 1.24 2010/09/21 13:14:17 vanhu Exp $	*/
+/*     $NetBSD: isakmp_cfg.c,v 1.23 2010/08/04 09:16:58 vanhu Exp $    */
 
 /* Id: isakmp_cfg.c,v 1.55 2006/08/22 18:17:17 manubsd Exp */
 
@@ -38,7 +38,12 @@
 #include <sys/socket.h>
 #include <sys/queue.h>
 
+#if __FreeBSD_version >= 900007
 #include <utmpx.h>
+#else
+#include <utmp.h>
+#include <libutil.h>
+#endif
 #if defined(__APPLE__) && defined(__MACH__)
 #include <util.h>
 #endif
@@ -1661,7 +1666,8 @@
 	int inout;
 {
 	int error = 0;
-	struct utmpx ut;
+       struct utmpx ut;
+       char term[UT_LINESIZE];
 	char addr[NI_MAXHOST];
 	
 	if (usr == NULL || usr[0]=='\0') {
@@ -1670,31 +1676,35 @@
 		return -1;
 	}
 
-	memset(&ut, 0, sizeof ut);
-	gettimeofday((struct timeval *)&ut.ut_tv, NULL);
-	snprintf(ut.ut_id, sizeof ut.ut_id, TERMSPEC, port);
+       sprintf(term, TERMSPEC, port);
 
 	switch (inout) {
 	case ISAKMP_CFG_LOGIN:
-		ut.ut_type = USER_PROCESS;
-		strncpy(ut.ut_user, usr, sizeof ut.ut_user);
+               strncpy(ut.ut_user, usr, UT_NAMESIZE);
+               ut.ut_user[UT_NAMESIZE - 1] = '\0';
+
+               strncpy(ut.ut_line, term, UT_LINESIZE);
+               ut.ut_line[UT_LINESIZE - 1] = '\0';
 
 		GETNAMEINFO_NULL(raddr, addr);
-		strncpy(ut.ut_host, addr, sizeof ut.ut_host);
+               strncpy(ut.ut_host, addr, UT_HOSTSIZE);
+               ut.ut_host[UT_HOSTSIZE - 1] = '\0';
+
+		gettimeofday(&ut.ut_tv, NULL);
 
 		plog(LLV_INFO, LOCATION, NULL,
 			"Accounting : '%s' logging on '%s' from %s.\n",
-			ut.ut_user, ut.ut_id, addr);
+                       ut.ut_user, ut.ut_line, ut.ut_host);
 
 		pututxline(&ut);
 
 		break;
 	case ISAKMP_CFG_LOGOUT:	
-		ut.ut_type = DEAD_PROCESS;
 
+		ut.ut_type = DEAD_PROCESS;
 		plog(LLV_INFO, LOCATION, NULL,
 			"Accounting : '%s' unlogging from '%s'.\n",
-			usr, ut.ut_id);
+                       usr, term);
 
 		pututxline(&ut);
 
