diff -ur ../ipsec-tools-0.8.0.old/src/racoon/cfparse.y ./src/racoon/cfparse.y
--- ../ipsec-tools-0.8.0.old/src/racoon/cfparse.y	2012-01-27 18:36:22.000000000 +0000
+++ ./src/racoon/cfparse.y	2012-01-27 18:52:39.000000000 +0000
@@ -201,9 +201,11 @@
 %token LDAP_ATTR_USER LDAP_ATTR_ADDR LDAP_ATTR_MASK LDAP_ATTR_GROUP LDAP_ATTR_MEMBER
 	/* radius config */
 %token RADCFG RAD_AUTH RAD_ACCT RAD_TIMEOUT RAD_RETRIES
+	/* external cofnig */
+%token EXTERNALCFG
 	/* modecfg */
 %token MODECFG CFG_NET4 CFG_MASK4 CFG_DNS4 CFG_NBNS4 CFG_DEFAULT_DOMAIN
-%token CFG_AUTH_SOURCE CFG_AUTH_GROUPS CFG_SYSTEM CFG_RADIUS CFG_PAM CFG_LDAP CFG_LOCAL CFG_NONE
+%token CFG_AUTH_SOURCE CFG_AUTH_GROUPS CFG_SYSTEM CFG_EXTERNAL CFG_RADIUS CFG_PAM CFG_LDAP CFG_LOCAL CFG_NONE
 %token CFG_GROUP_SOURCE CFG_ACCOUNTING CFG_CONF_SOURCE CFG_MOTD CFG_POOL_SIZE CFG_AUTH_THROTTLE
 %token CFG_SPLIT_NETWORK CFG_SPLIT_LOCAL CFG_SPLIT_INCLUDE CFG_SPLIT_DNS
 %token CFG_PFS_GROUP CFG_SAVE_PASSWD
@@ -278,6 +280,7 @@
 	|	listen_statement
 	|	ldapcfg_statement
 	|	radcfg_statement
+	|	external_statement
 	|	modecfg_statement
 	|	timer_statement
 	|	sainfo_statement
@@ -485,6 +488,19 @@
 	|	PORT		{ $$ = $1; }
 	;
 
+	/* external script parameters */
+external_statement
+	: EXTERNALCFG BOC SCRIPT QUOTEDSTRING EOC {
+#ifdef ENABLE_HYBRID
+		vfree(xauth_external_config.script);
+		xauth_external_config.script = vdup($4);
+#else
+		yyerror("racoon not configured with --enable-hybrid");
+		return -1;
+#endif
+	}
+	;
+
 	/* radius configuration */
 radcfg_statement
 	:	RADCFG {
@@ -830,6 +846,15 @@
 #endif
 		}
 		EOS
+	|	CFG_AUTH_SOURCE CFG_EXTERNAL
+		{
+#ifdef ENABLE_HYBRID
+			isakmp_cfg_config.authsource = ISAKMP_CFG_AUTH_EXTERNAL;
+#else
+			yyerror("racoon not configured with --enable-hybrid");
+#endif
+		}
+		EOS
 	|	CFG_AUTH_SOURCE CFG_RADIUS
 		{
 #ifdef ENABLE_HYBRID
diff -ur ../ipsec-tools-0.8.0.old/src/racoon/cftoken.l ./src/racoon/cftoken.l
--- ../ipsec-tools-0.8.0.old/src/racoon/cftoken.l	2012-01-27 18:36:23.000000000 +0000
+++ ./src/racoon/cftoken.l	2012-01-27 18:36:34.000000000 +0000
@@ -132,7 +132,7 @@
 decstring	{digit}+
 hexstring	0x{hexdigit}+
 
-%s S_INI S_PRIV S_PTH S_LOG S_PAD S_LST S_RTRY S_CFG S_LDAP S_RAD
+%s S_INI S_PRIV S_PTH S_LOG S_PAD S_LST S_RTRY S_CFG S_LDAP S_RAD S_EXT
 %s S_ALGST S_ALGCL
 %s S_SAINF S_SAINFS
 %s S_RMT S_RMTS S_RMTP
@@ -237,6 +237,12 @@
 <S_LDAP>attr_member	{ YYD; return(LDAP_ATTR_MEMBER); }
 <S_LDAP>{ecl}		{ BEGIN S_INI; return(EOC); }
 
+	/* external script */
+<S_INI>extcfg		{ BEGIN S_EXT; YYDB; return(EXTERNALCFG); }
+<S_EXT>{bcl}		{ return(BOC); }
+<S_EXT>script		{ YYD; return(SCRIPT); }
+<S_EXT>{ecl}		{ BEGIN S_INI; return(EOC); }
+
 	/* mode_cfg */
 <S_INI>mode_cfg		{ BEGIN S_CFG; YYDB; return(MODECFG); }
 <S_CFG>{bcl}		{ return(BOC); }
@@ -252,6 +258,7 @@
 <S_CFG>conf_source	{ YYD; return(CFG_CONF_SOURCE); }
 <S_CFG>accounting	{ YYD; return(CFG_ACCOUNTING); }
 <S_CFG>system		{ YYD; return(CFG_SYSTEM); }
+<S_CFG>external		{ YYD; return(CFG_EXTERNAL); }
 <S_CFG>local		{ YYD; return(CFG_LOCAL); }
 <S_CFG>none		{ YYD; return(CFG_NONE); }
 <S_CFG>radius		{ YYD; return(CFG_RADIUS); }
diff -ur ../ipsec-tools-0.8.0.old/src/racoon/isakmp_cfg.h ./src/racoon/isakmp_cfg.h
--- ../ipsec-tools-0.8.0.old/src/racoon/isakmp_cfg.h	2012-01-27 18:36:23.000000000 +0000
+++ ./src/racoon/isakmp_cfg.h	2012-01-27 18:36:34.000000000 +0000
@@ -111,6 +111,7 @@
 #define ISAKMP_CFG_AUTH_RADIUS	1
 #define ISAKMP_CFG_AUTH_PAM	2
 #define ISAKMP_CFG_AUTH_LDAP	4
+#define ISAKMP_CFG_AUTH_EXTERNAL	5
 
 /* For groupsource */
 #define ISAKMP_CFG_GROUP_SYSTEM	0
diff -ur ../ipsec-tools-0.8.0.old/src/racoon/isakmp_xauth.c ./src/racoon/isakmp_xauth.c
--- ../ipsec-tools-0.8.0.old/src/racoon/isakmp_xauth.c	2012-01-27 18:36:22.000000000 +0000
+++ ./src/racoon/isakmp_xauth.c	2012-01-27 19:00:22.000000000 +0000
@@ -37,6 +37,7 @@
 #include <sys/param.h>
 #include <sys/socket.h>
 #include <sys/queue.h>
+#include <sys/wait.h>
 
 #include <netinet/in.h>
 
@@ -101,6 +102,10 @@
 struct xauth_rad_config xauth_rad_config;
 #endif
 
+#ifdef ENABLE_HYBRID
+struct xauth_ext_config xauth_external_config;
+#endif
+
 #ifdef HAVE_LIBPAM
 #include <security/pam_appl.h>
 
@@ -293,6 +298,11 @@
 			res = xauth_login_ldap(iph1, usr, pwd);
 			break;
 #endif
+#ifdef ENABLE_HYBRID
+		case ISAKMP_CFG_AUTH_EXTERNAL:
+			res = xauth_login_script(iph1, usr, pwd);
+			break;
+#endif
 		default:
 			plog(LLV_ERROR, LOCATION, NULL, 
 			    "Unexpected authentication source\n");
@@ -1256,6 +1266,64 @@
 
 #endif
 
+#ifdef ENABLE_HYBRID
+int
+xauth_login_script(struct ph1handle *iph1, char *user, char *pwd)
+{
+	pid_t pid, rpid;
+	char **envp, **c;
+	char *argv[] = { NULL, NULL, NULL };
+	int envc = 0;
+	int ret = -1;
+
+	argv[0] = xauth_external_config.script->v;
+	argv[1] = xauth_external_config.script->v;
+	argv[2] = NULL;
+
+        (void)isakmp_cfg_setenv(iph1, &envp, &envc);
+
+	if (script_env_append(&envp, &envc, "username", user) != 0) {
+                plog(LLV_ERROR, LOCATION, NULL, "Cannot set username\n");
+                goto outscript;
+        }
+	if (script_env_append(&envp, &envc, "password", pwd) != 0) {
+                plog(LLV_ERROR, LOCATION, NULL, "Cannot set password\n");
+                goto outscript;
+        }
+
+	pid = fork();
+	if (pid == (pid_t)0) {
+		execve(argv[0], argv, envp);
+		plog(LLV_ERROR, LOCATION, NULL,
+                    "execve(\"%s\") failed: %s\n",
+                    argv[0], strerror(errno));
+		exit (127);
+	} else if (pid <= (pid_t)0)
+		ret = -1;
+	else {
+		do {
+			rpid = waitpid (pid, &ret, 0);
+		} while (rpid == -1 && errno == EINTR);
+		if (rpid != pid)
+			ret = -1;
+	}
+
+	if (WIFEXITED(ret)) {
+		if (WEXITSTATUS(ret) > 0)
+			ret = -1;
+		else
+			ret = 0;
+	}
+
+outscript:
+	for (c = envp; *c; c++)
+		racoon_free(*c);
+	racoon_free(envp);
+
+	return (ret);
+}
+#endif
+
 int
 xauth_login_system(usr, pwd)
 	char *usr;
diff -ur ../ipsec-tools-0.8.0.old/src/racoon/isakmp_xauth.h ./src/racoon/isakmp_xauth.h
--- ../ipsec-tools-0.8.0.old/src/racoon/isakmp_xauth.h	2012-01-27 18:36:22.000000000 +0000
+++ ./src/racoon/isakmp_xauth.h	2012-01-27 18:39:50.000000000 +0000
@@ -120,6 +120,14 @@
 int xauth_login_pam(int, struct sockaddr *, char *, char *);
 #endif
 
+#ifdef  ENABLE_HYBRID
+struct xauth_ext_config {
+	vchar_t  *script;
+};
+extern struct xauth_ext_config xauth_external_config;
+int xauth_login_script(struct ph1handle *, char *, char *);
+#endif
+
 #ifdef HAVE_LIBRADIUS
 
 #define RADIUS_MAX_SERVERS 5
