Index: ip_input.c
===================================================================
RCS file: /home/ncvs/src/sys/netinet/ip_input.c,v
retrieving revision 1.301.2.7
diff -u -r1.301.2.7 ip_input.c
--- ip_input.c	4 Mar 2006 09:15:08 -0000	1.301.2.7
+++ ip_input.c	19 Nov 2006 18:34:38 -0000
@@ -1990,8 +1990,6 @@
 #else /* FAST_IPSEC */
 				KEY_FREESP(&sp);
 #endif
-				ipstat.ips_cantfrag++;
-				break;
 			}
 		}
 #endif /*IPSEC || FAST_IPSEC*/
