Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /home/eri/development/divert/tmprepo/RELENG_7/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.6
diff -u -r1.6 pf.c
--- sys/contrib/pf/net/pf.c	30 Nov 2008 15:17:26 -0000	1.6
+++ sys/contrib/pf/net/pf.c	30 Nov 2008 15:18:31 -0000
@@ -3227,6 +3227,7 @@
 pf_set_rt_ifp(struct pf_state *s, struct pf_addr *saddr)
 {
 	struct pf_rule *r = s->rule.ptr;
+	struct pf_src_node *sn = NULL;
 
 	s->rt_kif = NULL;
 	if (!r->rt || r->rt == PF_FASTROUTE)
@@ -3234,15 +3235,13 @@
 	switch (s->af) {
 #ifdef INET
 	case AF_INET:
-		pf_map_addr(AF_INET, r, saddr, &s->rt_addr, NULL,
-		    &s->nat_src_node);
+		pf_map_addr(AF_INET, r, saddr, &s->rt_addr, NULL, &sn);
 		s->rt_kif = r->rpool.cur->kif;
 		break;
 #endif /* INET */
 #ifdef INET6
 	case AF_INET6:
-		pf_map_addr(AF_INET6, r, saddr, &s->rt_addr, NULL,
-		    &s->nat_src_node);
+		pf_map_addr(AF_INET6, r, saddr, &s->rt_addr, NULL, &sn);
 		s->rt_kif = r->rpool.cur->kif;
 		break;
 #endif /* INET6 */
