Index: geom_ctl.c
===================================================================
RCS file: /v/freebsd/cvs/src/sys/geom/geom_ctl.c,v
retrieving revision 1.33
diff -u -d -r1.33 geom_ctl.c
--- geom_ctl.c	23 Oct 2004 20:51:20 -0000	1.33
+++ geom_ctl.c	5 Jan 2005 20:58:38 -0000
@@ -378,7 +378,7 @@
 		if (!strcmp(p, cp->name))
 			return (cp);
 	}
-	gctl_error(req, "Class not found");
+	gctl_error(req, "Class not found: \"%s\"", p);
 	return (NULL);
 }
 
@@ -400,7 +400,7 @@
 			}
 		}
 	}
-	gctl_error(req, "Geom not found");
+	gctl_error(req, "Geom not found: \"%s\"", p);
 	return (NULL);
 }
 
@@ -416,7 +416,7 @@
 	pp = g_provider_by_name(p);
 	if (pp != NULL)
 		return (pp);
-	gctl_error(req, "Provider not found");
+	gctl_error(req, "Provider not found: \"%s\"", p);
 	return (NULL);
 }
 
