Index: if_sis.c
===================================================================
RCS file: /home/ncvs/src/sys/pci/if_sis.c,v
retrieving revision 1.132.2.7
diff -u -r1.132.2.7 if_sis.c
--- if_sis.c	17 Mar 2006 21:30:57 -0000	1.132.2.7
+++ if_sis.c	9 Jun 2007 23:12:14 -0000
@@ -1856,7 +1856,6 @@
 	sis_stop(sc);
 	sc->sis_stopped = 0;
 
-#ifdef notyet
 	if (sc->sis_type == SIS_TYPE_83815 && sc->sis_srr >= NS_SRR_16A) {
 		/*
 		 * Configure 400usec of interrupt holdoff.  This is based
@@ -1864,7 +1863,6 @@
  		 */
 		CSR_WRITE_4(sc, NS_IHR, 0x100 | 4);
 	}
-#endif
 
 	mii = device_get_softc(sc->sis_miibus);
 
