Index: sys/conf/files
===================================================================
RCS file: /home/eri/development/FreeBSD/src/sys/conf/files,v
retrieving revision 1.1243.2.2
diff -u -r1.1243.2.2 files
--- sys/conf/files	6 Dec 2007 10:15:29 -0000	1.1243.2.2
+++ sys/conf/files	28 Apr 2008 23:11:50 -0000
@@ -1608,7 +1608,8 @@
 net/if_disc.c			optional disc
 net/if_edsc.c			optional edsc
 net/if_ef.c			optional ef
-net/if_enc.c			optional enc
+net/if_enc.c			optional enc \
+	compile-with "${NORMAL_C} -I$S/contrib/pf"
 net/if_ethersubr.c		optional ether \
 	compile-with "${NORMAL_C} -I$S/contrib/pf"
 net/if_faith.c			optional faith
@@ -1925,7 +1926,8 @@
 netipsec/ipsec.c		optional ipsec
 netipsec/ipsec_input.c		optional ipsec
 netipsec/ipsec_mbuf.c		optional ipsec
-netipsec/ipsec_output.c		optional ipsec
+netipsec/ipsec_output.c		optional ipsec \
+	compile-with "${NORMAL_C} -I$S/contrib/pf"
 netipsec/key.c			optional ipsec
 netipsec/key_debug.c		optional ipsec
 netipsec/keysock.c		optional ipsec
Index: sys/net/if_enc.c
===================================================================
RCS file: /home/eri/development/FreeBSD/src/sys/net/if_enc.c,v
retrieving revision 1.6.4.1
diff -u -r1.6.4.1 if_enc.c
--- sys/net/if_enc.c	29 Dec 2007 17:28:38 -0000	1.6.4.1
+++ sys/net/if_enc.c	28 Apr 2008 23:11:09 -0000
@@ -48,6 +48,9 @@
 #include <net/bpf.h>
 #include <net/bpfdesc.h>
 
+#include <altq/if_altq.h>
+#include <net/pf_mtag.h>
+
 #include <netinet/in.h>
 #include <netinet/in_systm.h>
 #include <netinet/ip.h>
@@ -194,11 +197,12 @@
 }
 
 int
-ipsec_filter(struct mbuf **mp, int dir)
+ipsec_filter(struct mbuf **mp, struct secasindex *saidx, int dir)
 {
 	int error, i;
 	struct ip *ip;
-
+ 	struct pf_mtag *atag;
+	
 	KASSERT(encif != NULL, ("%s: encif is null", __func__));
 
 	if ((encif->if_drv_flags & IFF_DRV_RUNNING) == 0)
@@ -267,6 +271,9 @@
 	if (error != 0)
 		goto bad;
 
+	if (saidx && (atag = pf_find_mtag(*mp)) != NULL) 
+		saidx->qid = atag->qid; 
+
 	return (error);
 
 bad:
Index: sys/netipsec/ipsec.h
===================================================================
RCS file: /home/eri/development/FreeBSD/src/sys/netipsec/ipsec.h,v
retrieving revision 1.13
diff -u -r1.13 ipsec.h
--- sys/netipsec/ipsec.h	1 Jul 2007 11:38:29 -0000	1.13
+++ sys/netipsec/ipsec.h	28 Apr 2008 23:11:09 -0000
@@ -410,7 +410,7 @@
 extern	struct mbuf *m_makespace(struct mbuf *m0, int skip, int hlen, int *off);
 extern	caddr_t m_pad(struct mbuf *m, int n);
 extern	int m_striphdr(struct mbuf *m, int skip, int hlen);
-extern	int ipsec_filter(struct mbuf **, int);
+extern	int ipsec_filter(struct mbuf **, struct secasindex *, int);
 extern	void ipsec_bpf(struct mbuf *, struct secasvar *, int);
 #endif /* _KERNEL */
 
Index: sys/netipsec/ipsec_input.c
===================================================================
RCS file: /home/eri/development/FreeBSD/src/sys/netipsec/ipsec_input.c,v
retrieving revision 1.19
diff -u -r1.19 ipsec_input.c
--- sys/netipsec/ipsec_input.c	12 Sep 2007 05:54:53 -0000	1.19
+++ sys/netipsec/ipsec_input.c	28 Apr 2008 23:11:09 -0000
@@ -461,7 +461,7 @@
 	ipsec_bpf(m, sav, AF_INET);
 
 	if (prot != IPPROTO_IPIP)
-		if ((error = ipsec_filter(&m, PFIL_IN)) != 0)
+		if ((error = ipsec_filter(&m, &sav->sah->saidx, PFIL_IN)) != 0)
 			return (error);
 #endif
 
Index: sys/netipsec/ipsec_output.c
===================================================================
RCS file: /home/eri/development/FreeBSD/src/sys/netipsec/ipsec_output.c,v
retrieving revision 1.16
diff -u -r1.16 ipsec_output.c
--- sys/netipsec/ipsec_output.c	19 Jul 2007 09:57:54 -0000	1.16
+++ sys/netipsec/ipsec_output.c	28 Apr 2008 23:11:09 -0000
@@ -43,6 +43,11 @@
 #include <sys/errno.h>
 #include <sys/syslog.h>
 
+#ifdef DEV_ENC
+#include <altq/if_altq.h>
+#include <net/pf_mtag.h>
+#endif
+
 #include <net/if.h>
 #include <net/pfil.h>
 #include <net/route.h>
@@ -89,6 +94,7 @@
 	struct m_tag *mtag;
 	struct secasvar *sav;
 	struct secasindex *saidx;
+	struct pf_mtag *atag = NULL;
 	int error;
 
 	IPSEC_SPLASSERT_SOFTNET(__func__);
@@ -161,6 +167,14 @@
 	}
 	key_sa_recordxfer(sav, m);		/* record data transfer */
 
+#ifdef DEV_ENC
+	if (saidx->qid && (atag = pf_get_mtag(m)) != NULL) {
+        	atag->qid = saidx->qid;
+                /* add hints for ecn */
+                atag->af = saidx->dst.sa.sa_family;
+                atag->hdr = NULL; /* This should be safe! */
+	}
+#endif
 	/*
 	 * We're done with IPsec processing, transmit the packet using the
 	 * appropriate network protocol (IP or IPv6). SPD lookup will be
@@ -363,7 +377,7 @@
 
 #ifdef DEV_ENC
 	/* pass the mbuf to enc0 for packet filtering */
-	if ((error = ipsec_filter(&m, PFIL_OUT)) != 0)
+	if ((error = ipsec_filter(&m, &sav->sah->saidx, PFIL_OUT)) != 0)
 		goto bad;
 #endif
 
Index: sys/netipsec/keydb.h
===================================================================
RCS file: /home/eri/development/FreeBSD/src/sys/netipsec/keydb.h,v
retrieving revision 1.6
diff -u -r1.6 keydb.h
--- sys/netipsec/keydb.h	25 Mar 2006 13:38:52 -0000	1.6
+++ sys/netipsec/keydb.h	28 Apr 2008 23:11:09 -0000
@@ -58,6 +58,8 @@
 	u_int8_t mode;			/* mode of protocol, see ipsec.h */
 	u_int32_t reqid;		/* reqid id who owned this SA */
 					/* see IPSEC_MANUAL_REQID_MAX. */
+	u_int32_t qid;			/* used for ALTQ shaping inside */
+					/* tunnel */
 };
 
 /* 
Index: sys/netipsec/xform_ipip.c
===================================================================
RCS file: /home/eri/development/FreeBSD/src/sys/netipsec/xform_ipip.c,v
retrieving revision 1.15
diff -u -r1.15 xform_ipip.c
--- sys/netipsec/xform_ipip.c	19 Jul 2007 09:57:54 -0000	1.15
+++ sys/netipsec/xform_ipip.c	28 Apr 2008 23:11:09 -0000
@@ -349,7 +349,7 @@
 
 #ifdef DEV_ENC
 	/* pass the mbuf to enc0 for packet filtering */
-	if (ipsec_filter(&m, PFIL_IN) != 0)
+	if (ipsec_filter(&m, NULL, PFIL_IN) != 0)
 		return;
 #endif
 
