--- cam_xpt.c	2010-11-20 14:54:47.000000000 -0500
+++ cam_xpt.c	2010-11-20 14:55:07.000000000 -0500
@@ -152,7 +152,7 @@
 
 TUNABLE_INT("kern.cam.boot_delay", &xsoftc.boot_delay);
 SYSCTL_INT(_kern_cam, OID_AUTO, boot_delay, CTLFLAG_RDTUN,
-           &xsoftc.boot_delay, 0, "Bus registration wait time");
+           &xsoftc.boot_delay, 10000, "Bus registration wait time");
 
 /* Queues for our software interrupt handler */
 typedef TAILQ_HEAD(cam_isrq, ccb_hdr) cam_isrq_t;
