Index: if.c
===================================================================
RCS file: /root/freebsd/test/repo/src/sys/net/if.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 if.c
--- if.c	16 Mar 2010 23:54:56 -0000	1.1.1.1
+++ if.c	17 Mar 2010 00:13:10 -0000
@@ -1343,6 +1343,7 @@
 static int
 if_rtdel(struct radix_node *rn, void *arg)
 {
+	return (0);
 	struct rtentry	*rt = (struct rtentry *)rn;
 	struct ifnet	*ifp = arg;
 	int		err;
