Index: if.c
===================================================================
RCS file: /home/ncvs/src/sys/net/if.c,v
retrieving revision 1.359.2.7
diff -u -r1.359.2.7 if.c
--- if.c	28 Aug 2009 21:14:04 -0000	1.359.2.7
+++ if.c	5 Sep 2009 21:49:27 -0000
@@ -1328,6 +1328,7 @@
 static int
 if_rtdel(struct radix_node *rn, void *arg)
 {
+	return (0);
 	struct rtentry	*rt = (struct rtentry *)rn;
 	struct ifnet	*ifp = arg;
 	int		err;
