Index: if.h
===================================================================
RCS file: /root/freebsd/repo/src/sys/net/if.h,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 if.h
--- if.h	14 Jun 2011 09:06:53 -0000	1.1.1.1
+++ if.h	15 Jun 2011 18:40:35 -0000
@@ -226,7 +226,7 @@
 #define	IFCAP_WOL	(IFCAP_WOL_UCAST | IFCAP_WOL_MCAST | IFCAP_WOL_MAGIC)
 #define	IFCAP_TOE	(IFCAP_TOE4 | IFCAP_TOE6)
 
-#define	IFQ_MAXLEN	50
+#define	IFQ_MAXLEN	128
 #define	IFNET_SLOWHZ	1		/* granularity is 1 second */
 
 /*
