Index: sys/netinet/ip_carp.c
===================================================================
RCS file: /root/freebsd/freebsd/repo/src/sys/netinet/ip_carp.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 ip_carp.c
--- sys/netinet/ip_carp.c	22 Jan 2010 11:00:06 -0000	1.1.1.1
+++ sys/netinet/ip_carp.c	22 Jan 2010 11:21:32 -0000
@@ -501,7 +501,7 @@
 			ifpromisc(sc->sc_carpdev, 0);
 			sc->sc_carpdev->if_carp = NULL;
 			CARP_LOCK_DESTROY(cif);
-			free(cif, M_CARP);
+			//free(cif, M_CARP);
 		} else if (unlock)
 			CARP_UNLOCK(cif);
 		sc->sc_carpdev = NULL;
