Index: pf.c
===================================================================
RCS file: /root/freebsd/test/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.7
diff -u -r1.7 pf.c
--- pf.c	1 Apr 2010 23:24:32 -0000	1.7
+++ pf.c	1 Apr 2010 23:24:55 -0000
@@ -7577,7 +7577,9 @@
 		m_freem(*m0);
 		*m0 = NULL;
 		action = PF_PASS;
-	} else if (r->rt)
+	} else if (r->rt &&
+		/* Check to avoid breaking ipfw forward rules needed on CP and proxies listening on loclhost*/
+		!((m->m_flags & M_FASTFWD_OURS) || (pd.pf_mtag->flags & PF_TAG_TRANSLATE_LOCALHOST)))
 		/* pf_route can free the mbuf causing *m0 to become NULL */
 		pf_route(m0, r, dir, ifp, s, &pd);
 
