Index: Makefile.inc1
===================================================================
RCS file: /home/ncvs/src/Makefile.inc1,v
retrieving revision 1.628.2.9.2.2
diff -u -r1.628.2.9.2.2 Makefile.inc1
--- Makefile.inc1	14 Jun 2010 02:26:33 -0000	1.628.2.9.2.2
+++ Makefile.inc1	30 May 2012 17:07:43 -0000
@@ -604,7 +604,7 @@
 ITOOLS=	[ awk cap_mkdb cat chflags chmod chown \
 	date echo egrep find grep ${_install-info} \
 	ln lockf make mkdir mtree mv pwd_mkdb rm sed sh sysctl \
-	test true uname wc zic
+	test true uname wc zic uuencode uudecode ex
 
 #
 # distributeworld
