Index: if.c
===================================================================
RCS file: /root/freebsd/freebsd/repo/src/usr.bin/netstat/if.c,v
retrieving revision 1.2
diff -u -r1.2 if.c
--- if.c	22 Jan 2010 11:13:41 -0000	1.2
+++ if.c	22 Jan 2010 11:25:00 -0000
@@ -692,6 +692,8 @@
 	sigsetmask(oldmask);
 	line++;
 	first = 0;
+	if (qflag && (qval-- == 0))
+		exit(0);
 	if (line == 21)
 		goto banner;
 	else
Index: main.c
===================================================================
RCS file: /root/freebsd/freebsd/repo/src/usr.bin/netstat/main.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 main.c
--- main.c	22 Jan 2010 11:00:09 -0000	1.1.1.1
+++ main.c	22 Jan 2010 11:25:00 -0000
@@ -331,6 +331,8 @@
 int	numeric_addr;	/* show addresses numerically */
 int	numeric_port;	/* show ports numerically */
 static int pflag;	/* show given protocol */
+int	qflag;		/* whether to quit after specified outputs */
+int	qval;		/* how much outputs before we exit */
 int	rflag;		/* show routing tables (or routing stats) */
 int	sflag;		/* show protocol statistics */
 int	tflag;		/* show i/f watchdog timers */
@@ -354,7 +356,7 @@
 
 	af = AF_UNSPEC;
 
-	while ((ch = getopt(argc, argv, "AaBbdf:ghI:iLlM:mN:np:rSstuWw:xz")) != -1)
+	while ((ch = getopt(argc, argv, "AaBbdf:ghI:iLlM:mN:np:q:rSstuWw:xz")) != -1)
 		switch(ch) {
 		case 'A':
 			Aflag = 1;
@@ -440,6 +442,10 @@
 			}
 			pflag = 1;
 			break;
+		case 'q':
+			qval = atoi(optarg);
+			qflag = 1;
+			break;
 		case 'r':
 			rflag = 1;
 			break;
Index: netstat.1
===================================================================
RCS file: /root/freebsd/freebsd/repo/src/usr.bin/netstat/netstat.1,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 netstat.1
--- netstat.1	22 Jan 2010 11:00:09 -0000	1.1.1.1
+++ netstat.1	22 Jan 2010 11:25:00 -0000
@@ -133,6 +133,7 @@
 .Nm
 .Fl w Ar wait
 .Op Fl I Ar interface
+.Op Fl q Ar howmany
 .Op Fl d
 .Op Fl M Ar core
 .Op Fl N Ar system
@@ -146,6 +147,11 @@
 or a single
 .Ar interface .
 If
+.Fl q
+is also present, exit after
+.Ar howmany
+outputs.
+If
 .Fl d
 is also present, show the number of dropped packets.
 .It Xo
Index: netstat.h
===================================================================
RCS file: /root/freebsd/freebsd/repo/src/usr.bin/netstat/netstat.h,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 netstat.h
--- netstat.h	22 Jan 2010 11:00:09 -0000	1.1.1.1
+++ netstat.h	22 Jan 2010 11:25:00 -0000
@@ -47,6 +47,8 @@
 extern int	mflag;	/* show memory stats */
 extern int	numeric_addr;	/* show addresses numerically */
 extern int	numeric_port;	/* show ports numerically */
+extern int	qflag;	/* whether to quit after specified outputs */
+extern int	qval;	/* how much outputs before we exit */
 extern int	rflag;	/* show routing tables (or routing stats) */
 extern int	sflag;	/* show protocol statistics */
 extern int	tflag;	/* show i/f watchdog timers */
Index: route.c
===================================================================
RCS file: /root/freebsd/freebsd/repo/src/usr.bin/netstat/route.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 route.c
--- route.c	22 Jan 2010 11:00:09 -0000	1.1.1.1
+++ route.c	22 Jan 2010 11:25:00 -0000
@@ -1008,11 +1008,11 @@
 #define	p(f, m) if (rtstat.f || sflag <= 1) \
 	printf(m, rtstat.f, plural(rtstat.f))
 
-	p(rts_badredirect, "\t%hu bad routing redirect%s\n");
-	p(rts_dynamic, "\t%hu dynamically created route%s\n");
-	p(rts_newgateway, "\t%hu new gateway%s due to redirects\n");
-	p(rts_unreach, "\t%hu destination%s found unreachable\n");
-	p(rts_wildcard, "\t%hu use%s of a wildcard route\n");
+	p(rts_badredirect, "\t%u bad routing redirect%s\n");
+	p(rts_dynamic, "\t%u dynamically created route%s\n");
+	p(rts_newgateway, "\t%u new gateway%s due to redirects\n");
+	p(rts_unreach, "\t%u destination%s found unreachable\n");
+	p(rts_wildcard, "\t%u use%s of a wildcard route\n");
 #undef p
 
 	if (rttrash || sflag <= 1)
