Index: usr.bin/netstat/if.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/if.c,v
retrieving revision 1.70.2.1
diff -u -r1.70.2.1 if.c
--- usr.bin/netstat/if.c	3 Aug 2009 08:13:06 -0000	1.70.2.1
+++ usr.bin/netstat/if.c	23 Sep 2009 17:16:33 -0000
@@ -222,9 +222,9 @@
 
 	if (!pfunc) {
 		if (Wflag)
-			printf("%-7.7s", "Name");
+			printf("%-16.16s", "Name");
 		else
-			printf("%-5.5s", "Name");
+			printf("%-16.16s", "Name");
 		printf(" %5.5s %-13.13s %-17.17s %8.8s %5.5s",
 		    "Mtu", "Network", "Address", "Ipkts", "Ierrs");
 		if (bflag)
@@ -291,9 +291,9 @@
 
 		if (ifaddraddr == 0) {
 			if (Wflag)
-				printf("%-7.7s", name);
+				printf("%-16.16s", name);
 			else
-				printf("%-5.5s", name);
+				printf("%-16.16s", name);
 			printf(" %5lu ", ifnet.if_mtu);
 			printf("%-13.13s ", "none");
 			printf("%-17.17s ", "none");
@@ -313,7 +313,7 @@
 				continue;
 			}
 			if (Wflag)
-				printf("%-7.7s", name);
+				printf("%-16.16s", name);
 			else
 				printf("%-5.5s", name);
 			printf(" %5lu ", ifnet.if_mtu);
