Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.10
diff -u -r1.10 pf.c
--- sys/contrib/pf/net/pf.c	8 Feb 2011 12:36:14 -0000	1.10
+++ sys/contrib/pf/net/pf.c	8 Feb 2011 12:38:34 -0000
@@ -6652,6 +6652,13 @@
 	if (oifp != ifp) {
 		if (in_broadcast(ip->ip_dst, oifp)) /* XXX: LOCKING of address list?! */
 			goto donelocal;
+
+		if (s && r->rt == PF_ROUTETO && pd->nat_rule != NULL &&
+		    r->direction == PF_OUT && r->direction == dir && pd->pf_mtag->routed < 2) {
+			int off = ip->ip_hl << 2;
+			PACKET_UNDO_NAT(m0, (*pd), s, dir);
+			pf_unlink_state(s);
+		}
 #ifdef __FreeBSD__
 		PF_UNLOCK();
 		if (pf_test(PF_OUT, ifp, &m0, NULL, NULL) != PF_PASS) {
