Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /root/freebsd/test/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.10
diff -u -r1.10 pf.c
--- sys/contrib/pf/net/pf.c	22 Apr 2010 18:13:35 -0000	1.10
+++ sys/contrib/pf/net/pf.c	22 Apr 2010 18:14:02 -0000
@@ -6653,7 +6653,7 @@
 			m0->m_flags &= ~(M_PROTOFLAGS);
 			PF_UNLOCK();
 			error = (*ifp->if_output)(ifp, m0, sintosa(dst),
-			    NULL);
+			    ro);
 			PF_LOCK();
 		} else
 #else
