--- pf_ioctl.c.old	2010-12-08 17:28:53.000000000 +0000
+++ pf_ioctl.c	2010-12-08 17:29:41.000000000 +0000
@@ -3752,11 +3752,6 @@
 	struct ip *h = NULL;
 	int chk;
 
-	/* We need a proper CSUM befor we start (s. OpenBSD ip_output) */
-	if ((*m)->m_pkthdr.csum_flags & CSUM_DELAY_DATA) {
-		in_delayed_cksum(*m);
-		(*m)->m_pkthdr.csum_flags &= ~CSUM_DELAY_DATA;
-	}
 	if ((*m)->m_pkthdr.len >= (int)sizeof(*h)) {
 		/* if m_pkthdr.len is less than ip header, pf will handle. */
 		h = mtod(*m, struct ip *);
