Index: sys/contrib/pf/net/pf_mtag.h
===================================================================
RCS file: /root/freebsd/8/src/sys/contrib/pf/net/pf_mtag.h,v
retrieving revision 1.4
diff -u -r1.4 pf_mtag.h
--- sys/contrib/pf/net/pf_mtag.h	12 Sep 2009 12:28:04 -0000	1.4
+++ sys/contrib/pf/net/pf_mtag.h	12 Sep 2009 12:56:24 -0000
@@ -29,6 +29,7 @@
  *
  */
 
+#if 0
 #ifndef _NET_PF_MTAG_H_
 #define _NET_PF_MTAG_H_
 
@@ -84,3 +85,4 @@
 }
 #endif /* _KERNEL */
 #endif /* _NET_PF_MTAG_H_ */
+#endif
Index: sys/sys/mbuf.h
===================================================================
RCS file: /root/freebsd/8/src/sys/sys/mbuf.h,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 mbuf.h
--- sys/sys/mbuf.h	12 Sep 2009 12:19:25 -0000	1.1.1.1
+++ sys/sys/mbuf.h	12 Sep 2009 12:55:51 -0000
@@ -107,6 +107,27 @@
 	void			(*m_tag_free)(struct m_tag *);
 };
 
+#define PF_TAG_GENERATED                0x01
+#define PF_TAG_FRAGCACHE                0x02
+#define PF_TAG_TRANSLATE_LOCALHOST      0x04
+#define PF_DN_TAG_PRESENT               0x08
+#define PF_FASTFWD_OURS_PRESENT         0x10
+#define PF_DIVERT_TAG_PRESENT           0x20
+
+struct pf_mtag {
+        void            *hdr;           /* saved hdr pos in mbuf, for ECN */
+        u_int            rtableid;      /* alternate routing table id */
+        u_int32_t        qid;           /* queue id */
+        u_int32_t        qid_hash;      /* queue hashid used by WFQ like algos */
+        u_int16_t        tag;           /* tag id */
+        u_int8_t         flags;
+        u_int8_t         routed;
+        u_int8_t	 af;            /* for ECN */
+};
+
+#define pf_find_mtag(m)	&(m)->m_pkthdr.pf_tag
+#define pf_get_mtag(m)	&(m)->m_pkthdr.pf_tag
+
 /*
  * Record/packet header in first mbuf of chain; valid only if M_PKTHDR is set.
  */
@@ -126,6 +147,7 @@
 		u_int16_t vt_vtag;	/* Ethernet 802.1p+q vlan tag */
 		u_int16_t vt_nrecs;	/* # of IGMPv3 records in this chain */
 	} PH_vt;
+	struct pf_mtag	  pf_tag;
 	SLIST_HEAD(packet_tags, m_tag) tags; /* list of packet tags */
 };
 #define ether_vtag	PH_vt.vt_vtag
