Index: proc.h
===================================================================
RCS file: /home/ncvs/src/sys/sys/proc.h,v
retrieving revision 1.535.2.4
diff -u -r1.535.2.4 proc.h
--- proc.h	12 Dec 2009 20:06:25 -0000	1.535.2.4
+++ proc.h	29 Dec 2009 23:54:53 -0000
@@ -654,7 +654,7 @@
  * We use process IDs <= PID_MAX; PID_MAX + 1 must also fit in a pid_t,
  * as it is used to represent "no process group".
  */
-#define	PID_MAX		99999
+#define	PID_MAX		64000
 #define	NO_PID		100000
 
 #define	SESS_LEADER(p)	((p)->p_session->s_leader == (p))
