Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /root/freebsd/test/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.4
diff -u -r1.4 pf.c
--- sys/contrib/pf/net/pf.c	21 Apr 2010 21:51:16 -0000	1.4
+++ sys/contrib/pf/net/pf.c	21 Apr 2010 21:53:04 -0000
@@ -6363,6 +6363,13 @@
 
 	ip = mtod(m0, struct ip *);
 
+	if (IN_MULTICAST(ntohl(ip->ip_dst.s_addr)))
+                return;
+	else if (r->rt == PF_REPLYTO && in_localaddr(ip->ip_dst))
+		goto donelocal;
+	else if (r->rt == PF_ROUTETO && r->direction == dir && in_localip(ip->ip_dst))
+		goto donelocal;
+
 	ro = &iproute;
 	bzero((caddr_t)ro, sizeof(*ro));
 	dst = satosin(&ro->ro_dst);
@@ -6404,6 +6411,8 @@
 	}
 	if (ifp == NULL)
 		goto bad;
+	else if (in_broadcast(ip->ip_dst, oifp)) /* XXX: LOCKING of address list?! */
+		goto donelocal;
 
 	if (oifp != ifp) {
 #ifdef __FreeBSD__
@@ -6585,6 +6594,7 @@
 done:
 	if (r->rt != PF_DUPTO)
 		*m = NULL;
+donelocal:
 	if (ro == &iproute && ro->ro_rt)
 		RTFREE(ro->ro_rt);
 	return;
