Index: sys/netinet/tcp_input.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/netinet/tcp_input.c,v
retrieving revision 1.2
diff -u -r1.2 tcp_input.c
--- sys/netinet/tcp_input.c	4 Apr 2011 21:09:43 -0000	1.2
+++ sys/netinet/tcp_input.c	4 Apr 2011 21:48:16 -0000
@@ -1559,7 +1559,8 @@
 				(TF_RCVD_SCALE|TF_REQ_SCALE)) {
 				tp->rcv_scale = tp->request_r_scale;
 			}
-			tp->rcv_adv += tp->rcv_wnd;
+			tp->rcv_adv += imin(tp->rcv_wnd,
+			    TCP_MAXWIN << tp->rcv_scale);
 			tp->snd_una++;		/* SYN is acked */
 			/*
 			 * If there's data, delay ACK; if there's also a FIN
