Index: sys/netinet/tcp_lro.c
===================================================================
RCS file: /home/ncvs/src/sys/netinet/tcp_lro.c,v
retrieving revision 1.2.2.1.4.1
diff -u -r1.2.2.1.4.1 tcp_lro.c
--- sys/netinet/tcp_lro.c	14 Jun 2010 02:09:06 -0000	1.2.2.1.4.1
+++ sys/netinet/tcp_lro.c	8 Apr 2011 08:33:28 -0000
@@ -279,7 +279,9 @@
 		    lro->dest_ip == ip->ip_dst.s_addr) {
 			/* Try to append it */
 
-			if (__predict_false(seq != lro->next_seq)) {
+			if (__predict_false(seq != lro->next_seq ||
+				(tcp_data_len == 0 &&
+				lro->ack_seq == tcp->th_ack))) {
 				/* out of order packet */
 				SLIST_REMOVE(&cntl->lro_active, lro,
 					     lro_entry, next);
