Index: usr.sbin/wpa/hostapd/Makefile
===================================================================
RCS file: /home/ncvs/src/usr.sbin/wpa/hostapd/Makefile,v
retrieving revision 1.7.2.1.4.1
diff -u -r1.7.2.1.4.1 Makefile
--- usr.sbin/wpa/hostapd/Makefile	14 Jun 2010 02:09:06 -0000	1.7.2.1.4.1
+++ usr.sbin/wpa/hostapd/Makefile	11 Aug 2010 17:31:56 -0000
@@ -13,7 +13,7 @@
 	hw_features.c ieee802_11.c ieee802_11_common.c ieee802_11_auth.c \
 	ieee802_1x.c ip_addr.c md5.c mlme.c pmksa_cache.c radius.c \
 	radius_client.c rc4.c sha1.c sta_info.c vlan_init.c wme.c \
-	wpa.c wpa_auth_ie.c wpa_common.c wpa_debug.c wpabuf.c
+	wpa.c wpa_auth_ie.c wpa_common.c wpa_debug.c wpabuf.c preauth.c
 SRCS+=	l2_packet.c driver_freebsd.c os_unix.c
 
 MAN=	hostapd.8 hostapd.conf.5
@@ -26,7 +26,7 @@
 
 CFLAGS+= -I${HOSTAPD_DISTDIR}
 
-CFLAGS+= -DCONFIG_DRIVER_BSD
+CFLAGS+= -DCONFIG_DRIVER_BSD -DCONFIG_RSN_PREAUTH
 CFLAGS+= -DCONFIG_DRIVER_RADIUS_ACL
 .if ${MK_INET6} != "no"
 CFLAGS+= -DCONFIG_IPV6
