Index: secure/lib/libssh/Makefile
===================================================================
RCS file: /home/ncvs/src/secure/lib/libssh/Makefile,v
retrieving revision 1.34.2.4
diff -u -p -r1.34.2.4 Makefile
--- secure/lib/libssh/Makefile	6 Oct 2006 14:07:23 -0000	1.34.2.4
+++ secure/lib/libssh/Makefile	7 Oct 2006 15:43:13 -0000
@@ -32,14 +32,14 @@ LDADD=	-lz
 
 .if !defined(NO_KERBEROS)
 CFLAGS+= -DGSSAPI -DHAVE_GSSAPI_H=1 -DKRB5 -DHEIMDAL
-DPADD+=	${LIBGSSAPI} ${LIBKRB5} ${LIBASN1} ${LIBCOM_ERR} ${LIBMD} ${LIBROKEN}
-LDADD+=	-lgssapi -lkrb5 -lasn1 -lcom_err -lmd -lroken
+DPADD+=	${LIBGSSAPI} ${LIBKRB5} ${LIBASN1} ${LIBCOM_ERR} ${LIBROKEN}
+LDADD+=	-lgssapi -lkrb5 -lasn1 -lcom_err -lroken
 .endif
 
 NO_LINT=
 
-DPADD+=	${LIBCRYPTO} ${LIBCRYPT}
-LDADD+=	-lcrypto -lcrypt
+DPADD+=	${LIBCRYPTO} ${LIBCRYPT} ${LIBMD}
+LDADD+=	-lcrypto -lcrypt -lmd
 
 .include <bsd.lib.mk>
 
Index: secure/libexec/ssh-keysign/Makefile
===================================================================
RCS file: /home/ncvs/src/secure/libexec/ssh-keysign/Makefile,v
retrieving revision 1.6.2.1
diff -u -p -r1.6.2.1 Makefile
--- secure/libexec/ssh-keysign/Makefile	8 Jul 2006 14:37:18 -0000	1.6.2.1
+++ secure/libexec/ssh-keysign/Makefile	8 Oct 2006 00:42:40 -0000
@@ -8,8 +8,8 @@ CFLAGS+=-I${SSHDIR} -include ssh_namespa
 BINMODE=4511
 .endif
 
-DPADD=	${LIBSSH} ${LIBCRYPT} ${LIBCRYPTO} ${LIBZ}
-LDADD=	-lssh -lcrypt -lcrypto -lz
+DPADD=	${LIBSSH} ${LIBCRYPT} ${LIBCRYPTO} ${LIBMD} ${LIBZ}
+LDADD=	-lssh -lcrypt -lcrypto -lmd -lz
 
 .include <bsd.prog.mk>
 
Index: secure/usr.bin/sftp/Makefile
===================================================================
RCS file: /home/ncvs/src/secure/usr.bin/sftp/Makefile,v
retrieving revision 1.12.2.1
diff -u -p -r1.12.2.1 Makefile
--- secure/usr.bin/sftp/Makefile	8 Jul 2006 14:37:18 -0000	1.12.2.1
+++ secure/usr.bin/sftp/Makefile	7 Oct 2006 15:29:30 -0000
@@ -4,8 +4,8 @@ PROG=	sftp
 SRCS=   sftp.c sftp-client.c sftp-common.c sftp-glob.c progressmeter.c
 CFLAGS+=-I${SSHDIR} -include ssh_namespace.h
 
-DPADD=	${LIBSSH} ${LIBCRYPT} ${LIBCRYPTO} ${LIBZ} ${LIBEDIT}
-LDADD=	-lssh -lcrypt -lcrypto -lz -ledit
+DPADD=	${LIBSSH} ${LIBCRYPT} ${LIBCRYPTO} ${LIBZ} ${LIBEDIT} ${LIBNCURSES}
+LDADD=	-lssh -lcrypt -lcrypto -lz -ledit -lncurses
 
 .include <bsd.prog.mk>
 
Index: secure/usr.bin/ssh/Makefile
===================================================================
RCS file: /home/ncvs/src/secure/usr.bin/ssh/Makefile,v
retrieving revision 1.31.2.1
diff -u -p -r1.31.2.1 Makefile
--- secure/usr.bin/ssh/Makefile	8 Jul 2006 14:37:18 -0000	1.31.2.1
+++ secure/usr.bin/ssh/Makefile	7 Oct 2006 17:26:38 -0000
@@ -11,13 +11,15 @@ SRCS=	ssh.c readconf.c clientloop.c ssht
 	sshconnect.c sshconnect1.c sshconnect2.c \
 	gss-genr.c
 
+DPADD=	${LIBSSH} ${LIBMD} ${LIBUTIL} ${LIBZ}
+LDADD=	-lssh -lmd -lutil -lz
 DPADD=	${LIBSSH} ${LIBUTIL} ${LIBZ}
 LDADD=	-lssh -lutil -lz
 
 .if !defined(NO_KERBEROS)
 CFLAGS+= -DGSSAPI -DHAVE_GSSAPI_H=1 -DKRB5 -DHEIMDAL
-DPADD+=	 ${LIBGSSAPI} ${LIBKRB5} ${LIBASN1} ${LIBCOM_ERR} ${LIBMD} ${LIBROKEN}
-LDADD+=	 -lgssapi -lkrb5 -lasn1 -lcom_err -lmd -lroken
+DPADD+=	 ${LIBGSSAPI} ${LIBKRB5} ${LIBASN1} ${LIBCOM_ERR} ${LIBROKEN}
+LDADD+=	 -lgssapi -lkrb5 -lasn1 -lcom_err -lroken
 .endif
 
 .if defined(X11BASE)
Index: secure/usr.bin/ssh-keyscan/Makefile
===================================================================
RCS file: /home/ncvs/src/secure/usr.bin/ssh-keyscan/Makefile,v
retrieving revision 1.9.2.1
diff -u -p -r1.9.2.1 Makefile
--- secure/usr.bin/ssh-keyscan/Makefile	8 Jul 2006 14:37:19 -0000	1.9.2.1
+++ secure/usr.bin/ssh-keyscan/Makefile	7 Oct 2006 15:29:58 -0000
@@ -3,8 +3,8 @@
 PROG=	ssh-keyscan
 CFLAGS+=-I${SSHDIR} -include ssh_namespace.h
 
-DPADD=	${LIBSSH} ${LIBCRYPT} ${LIBCRYPTO} ${LIBZ}
-LDADD=	-lssh -lcrypt -lcrypto -lz
+DPADD=	${LIBSSH} ${LIBCRYPT} ${LIBCRYPTO} ${LIBMD} ${LIBZ}
+LDADD=	-lssh -lcrypt -lcrypto -lmd -lz
 
 .include <bsd.prog.mk>
 

