diff -urN apinger-0.6.1/src/apinger.h apinger-0.6.1-pf/src/apinger.h
--- src/apinger.h	2008-09-20 21:11:37.000000000 +0200
+++ src/apinger.h	2008-09-20 21:43:31.000000000 +0200
@@ -100,14 +100,14 @@
 
 struct trace_info {
 	struct timeval timestamp;
-	int seq;
+	uint16_t seq;
 	void *target_id;
 };
 
 #ifdef FORKED_RECEIVER
 struct piped_info {
 	struct trace_info ti;
-	int icmp_seq;
+	uint16_t icmp_seq;
 	struct timeval recv_timestamp;
 };
 #endif
diff -urN apinger-0.6.1/src/icmp.c apinger-0.6.1-pf/src/icmp.c
--- src/icmp.c	2008-09-20 21:11:37.000000000 +0200
+++ src/icmp.c	2008-09-20 21:39:52.000000000 +0200
@@ -136,7 +136,7 @@
 	p->icmp_type=ICMP_ECHO;
 	p->icmp_code=0;
 	p->icmp_cksum=0;
-	p->icmp_seq=seq%65536;
+	p->icmp_seq=seq;
 	p->icmp_id=ident;
 
 #ifdef HAVE_SCHED_YIELD
@@ -171,6 +171,7 @@
 struct iovec iov;
 struct msghdr msg;
 struct cmsghdr *c;
+reloophack:
 
         iov.iov_base=buf;
         iov.iov_len=1000;
@@ -183,6 +184,7 @@
         len=recvmsg(icmp_sock, &msg, MSG_DONTWAIT);
 #else
 socklen_t sl;
+reloophack:
 
         sl=sizeof(from);
         len=recvfrom(icmp_sock,buf,1024,MSG_DONTWAIT,(struct sockaddr *)&from,&sl);
@@ -197,7 +199,7 @@
         debug("checking CMSG...");
         for (c = CMSG_FIRSTHDR(&msg); c; c = CMSG_NXTHDR(&msg, c)) {
                 debug("CMSG level: %i type: %i",c->cmsg_level,c->cmsg_type);
-                if (c->cmsg_level != SOL_SOCKET || c->cmsg_type != SO_TIMESTAMP)
+                if (c->cmsg_level != SOL_SOCKET || c->cmsg_type != SCM_TIMESTAMP)
                         continue;
                 if (c->cmsg_len < CMSG_LEN(sizeof(struct timeval)))
                         continue;
@@ -227,7 +227,8 @@
 		return;
 	}
 	if (icmp->icmp_id != ident){
-		debug("Alien echo-reply received");
+		debug("Alien echo-reply received from %s. Expected %i, received %i",inet_ntoa(from.sin_addr), ident, icmp->icmp_id);
+		goto reloophack;        
 		return;
 	}
 	debug("Ping reply from %s",inet_ntoa(from.sin_addr));
diff -urN apinger-0.6.1/src/main.c apinger-0.6.1-pf/src/main.c
--- src/main.c	2008-09-20 21:11:37.000000000 +0200
+++ src/main.c	2008-09-20 21:38:34.000000000 +0200
@@ -277,7 +277,7 @@
 		return 1;
 	}
 
-	ident=getpid() & 0xFFFF;
+	ident=getpid();
 	signal(SIGTERM,signal_handler);
 	signal(SIGINT,signal_handler);
 	signal(SIGHUP,signal_handler);
