--- src/apinger.c	2009-06-04 13:06:46.000000000 +0200
+++ src/apinger.c	2009-06-04 14:00:55.000000000 +0200
@@ -853,9 +853,9 @@
 		fprintf(f,"Received packets buffer: %s %s\n",buf2,buf1);
 		if (t->recently_lost!=really_lost){
 			fprintf(f,"   lost packet count mismatch (%i!=%i)!\n",t->recently_lost,really_lost);
-			logit("%s: Lost packet count mismatch (%i!=%i)!",t->name,t->recently_lost,really_lost);
-			logit("%s: Received packets buffer: %s %s\n",t->name,buf2,buf1);
+			logit("Target \"%s\": Lost packet count mismatch (%i(recently_lost) != %i(really_lost))!",t->name,t->recently_lost,really_lost);
+			logit("Target \"%s\": Received packets buffer: %s %s\n",t->name,buf2,buf1);
 			err=1;
 		}
 		free(buf1);
 		free(buf2);
--- src/icmp.c	2008-09-20 21:11:37.000000000 +0200
+++ src/icmp.c	2008-09-20 21:43:31.000000000 +0200
@@ -227,7 +227,7 @@
 		return;
 	}
 	if (icmp->icmp_id != ident){
-		debug("Alien echo-reply received");
+		debug("Alien echo-reply received from %s. Expected %i, received %i",inet_ntoa(from.sin_addr), ident, icmp->icmp_id);
 		return;
 	}
 	debug("Ping reply from %s",inet_ntoa(from.sin_addr));
--- src/main.c.old	2008-07-10 22:36:21.000000000 +0200
+++ src/main.c	2008-07-10 22:36:53.000000000 +0200
@@ -286,6 +286,8 @@
 #ifdef FORKED_RECEIVER
 	signal(SIGCHLD,sigchld_handler);
 #endif
+	logit("Starting Alarm Pinger, apinger(%i)", ident);
+
 	main_loop();
 	if (icmp_sock>=0) close(icmp_sock);
 	if (icmp6_sock>=0) close(icmp6_sock);
