--- src/icmp.c.old	2008-09-22 19:13:52.000000000 +0200
+++ src/icmp.c	2008-09-22 19:13:58.000000000 +0200
@@ -171,6 +171,7 @@
 struct iovec iov;
 struct msghdr msg;
 struct cmsghdr *c;
+reloophack:
 
 	iov.iov_base=buf;
 	iov.iov_len=1000;
@@ -183,6 +184,7 @@
 	len=recvmsg(icmp_sock, &msg, MSG_DONTWAIT);
 #else
 socklen_t sl;
+reloophack:
 
 	sl=sizeof(from);
 	len=recvfrom(icmp_sock,buf,1024,MSG_DONTWAIT,(struct sockaddr *)&from,&sl);
@@ -228,6 +230,7 @@
 	}
 	if (icmp->icmp_id != ident){
 		debug("Alien echo-reply received from %s. Expected %i, received %i",inet_ntoa(from.sin_addr), ident, icmp->icmp_id);
+		goto reloophack;	
 		return;
 	}
 	debug("Ping reply from %s",inet_ntoa(from.sin_addr));

