--- src/apinger.c.old	2010-05-06 22:45:33.000000000 +0000
+++ src/apinger.c	2010-05-06 22:45:38.000000000 +0000
@@ -271,23 +271,15 @@
 time_t tm;
 	
 	tm=time(NULL);
-	fprintf(f,"%s",ctime(&tm));
-	if (on)
-		fprintf(f,"ALARM went off: %s\n",a->name);
-	else
-		fprintf(f,"alarm canceled: %s\n",a->name);
-	fprintf(f,"Target: %s\n",t->name);
-	fprintf(f,"Description: %s\n",t->description);
-	fprintf(f,"Probes sent: %i\n",t->last_sent+1);
-	fprintf(f,"Replies received: %i\n",t->received);
-	fprintf(f,"Last reply received: #%i %s",t->last_received,
-			ctime(&t->last_received_tv.tv_sec));
+	fprintf(f,"%s|%s|%i|%i|%u|",t->name, t->description, t->last_sent+1,
+		t->received, t->last_received_tv.tv_sec);
 	if (AVG_DELAY_KNOWN(t)){
-		fprintf(f,"Recent avg. delay: %4.3fms\n",AVG_DELAY(t));
+		fprintf(f,"%4.3fms|",AVG_DELAY(t));
 	}
 	if (AVG_LOSS_KNOWN(t)){
-		fprintf(f,"Recent avg. packet loss: %5.1f%%\n",AVG_LOSS(t));
+		fprintf(f,"%5.1f%%",AVG_LOSS(t));
 	}
+	fprintf(f, "\n");
 }
 
 void make_reports(struct target *t,struct alarm_cfg *a,int on,char* thisid,char* lastid){
@@ -798,25 +790,21 @@
 		return;
 	}
 	tm=time(NULL);
-	fprintf(f,"%s\n",ctime(&tm));
 	for(t=targets;t;t=t->next){
-		fprintf(f,"Target: %s\n",t->name);
-		fprintf(f,"Description: %s\n",t->description);
-		fprintf(f,"Last reply received: #%i %s",t->last_received,
-			ctime(&t->last_received_tv.tv_sec));
-		fprintf(f,"Average delay: %0.3fms\n",AVG_DELAY(t));
+		fprintf(f,"%s|%s|%i|%i|%u|",t->name, t->description, t->last_sent+1,
+                	t->received, t->last_received_tv.tv_sec);
+		fprintf(f,"%0.3fms|", AVG_DELAY(t));
 		if (AVG_LOSS_KNOWN(t)){
-			fprintf(f,"Average packet loss: %0.1f%%\n",AVG_LOSS(t));
+			fprintf(f,"%0.1f%%",AVG_LOSS(t));
 		}
-		fprintf(f,"Active alarms:");
+		fprintf(f, "|");
 		if (t->active_alarms){
 			for(al=t->active_alarms;al;al=al->next){
 				a=al->alarm;
 				fprintf(f," \"%s\"",a->name);
 			}
-			fprintf(f,"\n");
 		}
-		else fprintf(f," None\n");
+		else fprintf(f," None");
 
 		buf1=NEW(char,t->config->avg_loss_delay_samples+1);
 		buf2=NEW(char,t->config->avg_loss_samples+1);
@@ -850,9 +838,7 @@
 			}
 		}
 		buf2[i]=0;
-		fprintf(f,"Received packets buffer: %s %s\n",buf2,buf1);
 		if (t->recently_lost!=really_lost){
-			fprintf(f,"   lost packet count mismatch (%i!=%i)!\n",t->recently_lost,really_lost);
 			logit("Target \"%s\": Lost packet count mismatch (%i(recently_lost) != %i(really_lost))!",t->name,t->recently_lost,really_lost);
 			logit("Target \"%s\": Received packets buffer: %s %s\n",t->name,buf2,buf1);
 			err=1;
