diff -ur ../ipsec-tools-0.8.0.old/src/racoon/isakmp.c ./src/racoon/isakmp.c
--- ../ipsec-tools-0.8.0.old/src/racoon/isakmp.c	2012-01-27 23:02:07.000000000 +0000
+++ ./src/racoon/isakmp.c	2012-01-27 23:06:58.000000000 +0000
@@ -3252,11 +3252,13 @@
 	vchar_t *buf = NULL;
 	struct sadb_msg *msg, *next, *end;
 	struct sadb_sa *sa;
+	struct sadb_lifetime *lt;
 	struct sockaddr *src, *dst;
 	caddr_t mhp[SADB_EXT_MAX + 1];
 	u_int proto_id;
 	struct ph2handle *iph2;
 	struct ph1handle *new_iph1;
+	u_int64_t created;
 
 	plog(LLV_INFO, LOCATION, NULL,
 		 "purging ISAKMP-SA spi=%s.\n",
@@ -3308,6 +3310,11 @@
 		pk_fixup_sa_addresses(mhp);
 		src = PFKEY_ADDR_SADDR(mhp[SADB_EXT_ADDRESS_SRC]);
 		dst = PFKEY_ADDR_SADDR(mhp[SADB_EXT_ADDRESS_DST]);
+		lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_CURRENT];
+		if (lt != NULL)
+			created = lt->sadb_lifetime_addtime;
+		else
+			created = 0;
 
 		if (sa->sadb_sa_state != SADB_SASTATE_LARVAL &&
 		    sa->sadb_sa_state != SADB_SASTATE_MATURE &&
@@ -3373,7 +3380,7 @@
 
 		/* delete a relative phase 2 handle. */
 		if (iph2 != NULL) {
-			delete_spd(iph2, 0);
+			delete_spd(iph2, created);
 			remph2(iph2);
 			delph2(iph2);
 		}
diff -ur ../ipsec-tools-0.8.0.old/src/racoon/isakmp_inf.c ./src/racoon/isakmp_inf.c
--- ../ipsec-tools-0.8.0.old/src/racoon/isakmp_inf.c	2012-01-27 23:02:08.000000000 +0000
+++ ./src/racoon/isakmp_inf.c	2012-01-27 23:07:36.000000000 +0000
@@ -1158,7 +1158,7 @@
 		pk_fixup_sa_addresses(mhp);
 		src = PFKEY_ADDR_SADDR(mhp[SADB_EXT_ADDRESS_SRC]);
 		dst = PFKEY_ADDR_SADDR(mhp[SADB_EXT_ADDRESS_DST]);
-		lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_HARD];
+		lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_CURRENT];
 		if(lt != NULL)
 			created = lt->sadb_lifetime_addtime;
 		else
diff -ur ../ipsec-tools-0.8.0.old/src/racoon/pfkey.c ./src/racoon/pfkey.c
--- ../ipsec-tools-0.8.0.old/src/racoon/pfkey.c	2012-01-27 23:02:07.000000000 +0000
+++ ./src/racoon/pfkey.c	2012-01-27 23:05:18.000000000 +0000
@@ -2303,7 +2303,7 @@
 	saddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_SRC];
 	daddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_DST];
 	xpl = (struct sadb_x_policy *)mhp[SADB_X_EXT_POLICY];
-	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_HARD];
+	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_CURRENT];
 	if(lt != NULL)
 		created = lt->sadb_lifetime_addtime;
 	else
@@ -2441,7 +2441,7 @@
 	saddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_SRC];
 	daddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_DST];
 	xpl = (struct sadb_x_policy *)mhp[SADB_X_EXT_POLICY];
-	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_HARD];
+	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_CURRENT];
 	if(lt != NULL)
 		created = lt->sadb_lifetime_addtime;
 	else
@@ -2573,7 +2573,7 @@
 	saddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_SRC];
 	daddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_DST];
 	xpl = (struct sadb_x_policy *)mhp[SADB_X_EXT_POLICY];
-	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_HARD];
+	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_CURRENT];
 	if(lt != NULL)
 		created = lt->sadb_lifetime_addtime;
 	else
@@ -2649,7 +2649,7 @@
 	saddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_SRC];
 	daddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_DST];
 	xpl = (struct sadb_x_policy *)mhp[SADB_X_EXT_POLICY];
-	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_HARD];
+	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_CURRENT];
 	if(lt != NULL)
 		created = lt->sadb_lifetime_addtime;
 	else
@@ -2740,7 +2740,7 @@
 	saddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_SRC];
 	daddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_DST];
 	xpl = (struct sadb_x_policy *)mhp[SADB_X_EXT_POLICY];
-	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_HARD];
+	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_CURRENT];
 	if(lt != NULL)
 		created = lt->sadb_lifetime_addtime;
 	else
@@ -3386,7 +3386,7 @@
 	saddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_SRC];
 	daddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_DST];
 	xpl = (struct sadb_x_policy *)mhp[SADB_X_EXT_POLICY];
-	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_HARD];
+	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_CURRENT];
 	if (lt != NULL)
 		created = lt->sadb_lifetime_addtime;
 	else
@@ -3731,12 +3731,7 @@
 	saddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_SRC];
 	daddr = (struct sadb_address *)mhp[SADB_EXT_ADDRESS_DST];
 	xpl = (struct sadb_x_policy *)mhp[SADB_X_EXT_POLICY];
-	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_HARD];
-	if(lt != NULL)
-		created = lt->sadb_lifetime_addtime;
-	else
-		created = 0;
-	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_HARD];
+	lt = (struct sadb_lifetime*)mhp[SADB_EXT_LIFETIME_CURRENT];
 	if(lt != NULL)
 		created = lt->sadb_lifetime_addtime;
 	else
