From c37eeb7718c2ff0943b6a608ff76f3f55c9bcc90 Mon Sep 17 00:00:00 2001
From: yongari <yongari@FreeBSD.org>
Date: Thu, 7 Jun 2012 05:37:46 +0000
Subject: [PATCH 06/12] MFC r236371:   Remove unnecessary device_printfs.

---
 sys/dev/bge/if_bge.c |    8 --------
 1 file changed, 8 deletions(-)

diff --git a/sys/dev/bge/if_bge.c b/sys/dev/bge/if_bge.c
index f3d5ca9..e868d12 100644
--- a/sys/dev/bge/if_bge.c
+++ b/sys/dev/bge/if_bge.c
@@ -2800,17 +2800,9 @@ bge_mbox_reorder(struct bge_softc *sc)
 	for (;;) {
 		dev = device_get_parent(bus);
 		bus = device_get_parent(dev);
-		device_printf(sc->bge_dev, "dev : %s%d, bus : %s%d\n",
-		    device_get_name(dev), device_get_unit(dev),
-		    device_get_name(bus), device_get_unit(bus));
 		if (device_get_devclass(dev) != pcib)
 			break;
 		for (i = 0; i < nitems(mbox_reorder_lists); i++) {
-			device_printf(sc->bge_dev,
-			    "probing dev : %s%d, vendor : 0x%04x "
-			    "device : 0x%04x\n",
-			    device_get_name(dev), device_get_unit(dev),
-			    pci_get_vendor(dev), pci_get_device(dev));
 			if (pci_get_vendor(dev) ==
 			    mbox_reorder_lists[i].vendor &&
 			    pci_get_device(dev) ==
-- 
1.7.9.5

