Index: sys/dev/re/if_re.c
===================================================================
--- sys/dev/re/if_re.c	(revision 192617)
+++ sys/dev/re/if_re.c	(working copy)
@@ -2209,6 +2209,12 @@
 	    RL_ISR_TX_ERR|RL_ISR_TX_DESC_UNAVAIL))
 		re_txeof(sc);
 
+	if (sc->rl_type == RL_8139CPLUS && (status & RL_ISR_FIFO_OFLOW) != 0) {
+		device_printf(sc->rl_dev, "Rx FIFO overflow, resetting\n");
+		re_init_locked(sc);
+		RL_UNLOCK(sc);
+		return;
+	}
 	if (status & RL_ISR_SYSTEM_ERR)
 		re_init_locked(sc);
 
