--- sys/netinet/ip_carp.c.orig	Fri Jun  5 13:30:09 2009
+++ sys/netinet/ip_carp.c	Fri Jun  5 13:33:13 2009
@@ -35,6 +35,7 @@
 #include <sys/types.h>
 #include <sys/param.h>
 #include <sys/systm.h>
+#include <sys/bus.h>
 #include <sys/conf.h>
 #include <sys/kernel.h>
 #include <sys/limits.h>
@@ -2143,6 +2144,7 @@
 carp_set_state(struct carp_softc *sc, int state)
 {

+	char vhid_buf[15], *dev_msg;
 	if (sc->sc_carpdev)
 		CARP_SCLOCK_ASSERT(sc);

@@ -2153,15 +2155,20 @@
 	switch (state) {
 	case BACKUP:
 		SC2IFP(sc)->if_link_state = LINK_STATE_DOWN;
+		dev_msg = "CARP_SLAVE";
 		break;
 	case MASTER:
 		SC2IFP(sc)->if_link_state = LINK_STATE_UP;
+		dev_msg = "CARP_MASTER";
 		break;
 	default:
 		SC2IFP(sc)->if_link_state = LINK_STATE_UNKNOWN;
+		dev_msg = "CARP_UNKNOWN";
 		break;
 	}
 	rt_ifmsg(SC2IFP(sc));
+	snprintf(vhid_buf, sizeof(vhid_buf), "vhid=%d", sc->sc_vhid);
+	devctl_notify("IFNET", SC2IFP(sc)->if_xname, dev_msg, vhid_buf);
 }

 void
