Index: contrib/pf/pfctl/parse.y
===================================================================
RCS file: /home/ermal/tmprepo/RELENG_7/contrib/pf/pfctl/parse.y,v
retrieving revision 1.2
diff -u -r1.2 parse.y
--- contrib/pf/pfctl/parse.y	11 Feb 2009 23:20:04 -0000	1.2
+++ contrib/pf/pfctl/parse.y	13 Feb 2009 20:55:59 -0000
@@ -132,6 +132,7 @@
 enum	{ PF_STATE_OPT_MAX, PF_STATE_OPT_NOSYNC, PF_STATE_OPT_SRCTRACK,
 	    PF_STATE_OPT_MAX_SRC_STATES, PF_STATE_OPT_MAX_SRC_CONN,
 	    PF_STATE_OPT_MAX_SRC_CONN_RATE, PF_STATE_OPT_MAX_SRC_NODES,
+	    PF_STATE_OPT_MAX_PACKETS, 
 	    PF_STATE_OPT_OVERLOAD, PF_STATE_OPT_STATELOCK,
 	    PF_STATE_OPT_TIMEOUT };
 
@@ -143,6 +144,7 @@
 		u_int32_t	 max_states;
 		u_int32_t	 max_src_states;
 		u_int32_t	 max_src_conn;
+		u_int32_t	 max_packets;
 		struct {
 			u_int32_t	limit;
 			u_int32_t	seconds;
@@ -210,6 +212,7 @@
 	int			 allowopts;
 	char			*label;
 	struct node_qassign	 queues;
+	u_int16_t		 divert;
 	char			*tag;
 	char			*match_tag;
 	u_int8_t		 match_tag_not;
@@ -441,8 +444,8 @@
 %token  SRCIP DSTIP SRCPORT DSTPORT SRCIP6 DSTIP6 FLOWID NOERROR
 %token	LOAD RULESET_OPTIMIZATION
 %token	STICKYADDRESS MAXSRCSTATES MAXSRCNODES SOURCETRACK GLOBAL RULE
-%token	MAXSRCCONN MAXSRCCONNRATE OVERLOAD FLUSH
-%token	TAGGED TAG IFBOUND FLOATING STATEPOLICY ROUTE
+%token	MAXSRCCONN MAXSRCCONNRATE MAXPCKT OVERLOAD FLUSH
+%token	TAGGED TAG IFBOUND FLOATING DIVERT STATEPOLICY ROUTE
 %token	<v.string>		STRING
 %token	<v.i>			PORTBINARY
 %type	<v.interface>		interface if_list if_item_not if_item
@@ -470,6 +473,7 @@
 %type	<v.route>		route
 %type	<v.redirection>		redirection redirpool
 %type	<v.string>		label string tag anchorname
+%type	<v.number>		divert
 %type	<v.keep_state>		keep
 %type	<v.state_opt>		state_opt_spec state_opt_list state_opt_item
 %type	<v.logquick>		logquick quick log logopts logopt
@@ -2045,6 +2049,8 @@
 			r.rtableid = $9.rtableid;
 
 			r.af = $6;
+			if ($9.divert)
+				r.divert = $9.divert;
 			if ($9.tag)
 				if (strlcpy(r.tagname, $9.tag,
 				    PF_TAG_NAME_SIZE) >= PF_TAG_NAME_SIZE) {
@@ -2130,6 +2136,14 @@
 					}
 					r.rule_flag |= PFRULE_NOSYNC;
 					break;
+				case PF_STATE_OPT_MAX_PACKETS:
+					if (o->data.max_packets == 0) {
+						yyerror("max_packets must be"
+							"greater than 0");
+						YYERROR;
+					}
+					r.spare1 = o->data.max_packets;
+					break;
 				case PF_STATE_OPT_SRCTRACK:
 					if (srctrack) {
 						yyerror("state option "
@@ -2511,6 +2525,9 @@
 			filter_opts.dnpipe = $3;
 			filter_opts.free_flags |= PFRULE_DN_IS_QUEUE;
 		}
+		| DIVERT number				{
+			filter_opts.divert = $2;
+		}
 		| TAG string				{
 			filter_opts.tag = $2;
 		}
@@ -3586,6 +3603,15 @@
 			$$->next = NULL;
 			$$->tail = $$;
 		}
+		| MAXPCKT number                    	{
+                        $$ = calloc(1, sizeof(struct node_state_opt));
+                        if ($$ == NULL)
+                                err(1, "state_opt_item: calloc");
+                        $$->type = PF_STATE_OPT_MAX_PACKETS;
+                        $$->data.max_src_states = $2;
+                        $$->next = NULL;
+                        $$->tail = $$;
+                }
 		| MAXSRCCONN number			{
 			$$ = calloc(1, sizeof(struct node_state_opt));
 			if ($$ == NULL)
@@ -4250,6 +4276,10 @@
 		}
 		;
 
+divert		: /* empty */		{ $$ = 0; }
+		| DIVERT number		{ $$ = $2;   }
+		;
+
 tag		: /* empty */		{ $$ = NULL; }
 		| TAG STRING		{ $$ = $2; }
 		;
@@ -5341,6 +5371,7 @@
 		{ "crop",		FRAGCROP},
 		{ "debug",		DEBUG},
 		{ "delay",              DELAY},
+		{ "divert",		DIVERT},
 		{ "dngred",             DNGRED},
 		{ "dnpipe",             DNPIPE},
                 { "dnqueue",            DNQUEUE},
@@ -5351,6 +5382,7 @@
 		{ "dst-ip",             DSTIP},
                 { "dst-ip6",            DSTIP6},
                 { "dst-port",           DSTPORT},
+		{ "dummynet",		DUMMYNET},
 		{ "dup-to",		DUPTO},
 		{ "fairq",		FAIRQ},
 		{ "fastroute",		FASTROUTE},
@@ -5384,6 +5416,7 @@
 		{ "mask",		MASK},
 		{ "max",		MAXIMUM},
 		{ "max-mss",		MAXMSS},
+		{ "max-packets",	MAXPCKT},
 		{ "max-src-conn",	MAXSRCCONN},
 		{ "max-src-conn-rate",	MAXSRCCONNRATE},
 		{ "max-src-nodes",	MAXSRCNODES},
Index: contrib/pf/pfctl/pfctl_parser.c
===================================================================
RCS file: /home/ermal/tmprepo/RELENG_7/contrib/pf/pfctl/pfctl_parser.c,v
retrieving revision 1.2
diff -u -r1.2 pfctl_parser.c
--- contrib/pf/pfctl/pfctl_parser.c	11 Feb 2009 23:20:04 -0000	1.2
+++ contrib/pf/pfctl/pfctl_parser.c	13 Feb 2009 20:55:59 -0000
@@ -915,6 +915,12 @@
 			printf("max-src-conn %u", r->max_src_conn);
 			opts = 0;
 		}
+		if (r->spare1) {
+                        if (!opts)
+                                printf(", ");
+                        printf("max-packets %u", r->spare1);
+                        opts = 0;
+		}
 		if (r->max_src_conn_rate.limit) {
 			if (!opts)
 				printf(", ");
@@ -998,6 +1004,8 @@
 		printf(" queue(%s, %s)", r->qname, r->pqname);
 	else if (r->qname[0])
 		printf(" queue %s", r->qname);
+	if (r->divert)
+		printf(" divert %u", r->divert);
 	if (r->tagname[0])
 		printf(" tag %s", r->tagname);
 	if (r->match_tagname[0]) {
Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /home/ermal/tmprepo/RELENG_7/sys/contrib/pf/net/pf.c,v
retrieving revision 1.2
diff -u -r1.2 pf.c
--- sys/contrib/pf/net/pf.c	11 Feb 2009 23:20:04 -0000	1.2
+++ sys/contrib/pf/net/pf.c	7 Mar 2009 18:07:55 -0000
@@ -138,6 +138,7 @@
 #ifdef __FreeBSD__
 #include <netinet/ip_fw.h>
 #include <netinet/ip_dummynet.h>
+#include <netinet/ip_divert.h>
 
 #include <netinet6/ip6_var.h>
 #include <netinet6/in6_pcb.h>
@@ -383,6 +384,8 @@
 		r = (*state)->rule.ptr;						\
         	if (r->dnpipe && !(pd->pf_mtag->flags & PF_DN_TAG_PRESENT))	\
                 	return (PF_PASS);					\
+		else if (r->divert && !(pd->pf_mtag->flags & PF_DIVERT_TAG_PRESENT)) \
+			return (PF_PASS);					\
 	} while (0)
 
 struct pf_src_tree tree_src_tracking;
@@ -4710,7 +4713,8 @@
          */
         if ((th->th_flags & TH_SYN) && src->state == TCPS_SYN_SENT &&
 		dst->state == TCPS_CLOSED) {
-		if (pd->pf_mtag->flags & PF_DN_TAG_PRESENT)
+		if ((pd->pf_mtag->flags & PF_DN_TAG_PRESENT) ||
+			(pd->pf_mtag->flags & PF_DIVERT_TAG_PRESENT))
                 	return (PF_PASS);
 	}
 
@@ -6821,6 +6825,8 @@
 
 #ifdef __FreeBSD__
 	struct m_tag *dn_tag;
+	u_int16_t divertcookie = 0;
+	u_int8_t divflags = 0;
 	struct ip_fw_args	 dnflow;
 
 	PF_LOCK();
@@ -6893,18 +6899,29 @@
 		goto done;
 	}
 
-	dn_tag = m_tag_find(m, PACKET_TAG_DUMMYNET, NULL);
-        if (dn_tag != NULL &&
-                ((struct dn_pkt_tag *)(dn_tag+1))->dn_client == DN_CLIENT_PF) {
+	if (DUMMYNET_LOADED && 
+        	((dn_tag = m_tag_find(m, PACKET_TAG_DUMMYNET, NULL)) != NULL &&
+                ((struct dn_pkt_tag *)(dn_tag+1))->dn_client == DN_CLIENT_PF)) {
 		pd.pf_mtag->flags |= PF_DN_TAG_PRESENT;
                 m_tag_delete(m, dn_tag);
+		goto nonormalize;
         }
 
+	if (ip_divert_ptr != NULL &&
+		((dn_tag = m_tag_find(m, PACKET_TAG_DIVERT, NULL)) != NULL)) {
+		pd.pf_mtag->flags |= PF_DIVERT_TAG_PRESENT;
+		divertcookie = divert_cookie(dn_tag);
+		divflags |= divertcookie & PFSTATE_DIVERT_MASK;
+		divertcookie &= ~PFSTATE_DIVERT_MASK; 
+		m_tag_delete(m, dn_tag);
+	}
+
 	/* We do IP header normalization and packet reassembly here */
 	else if (pf_normalize_ip(m0, dir, kif, &reason, &pd) != PF_PASS) {
 		action = PF_DROP;
 		goto done;
-	}
+	} 
+nonormalize:
 	m = *m0;
 	h = mtod(m, struct ip *);
 
@@ -7091,6 +7108,17 @@
 		    ("pf: dropping packet with ip options\n"));
 	}
 
+#ifdef __FreeBSD__
+	if (s) {
+		if (divertcookie) {
+			s->divert_cookie = divertcookie;
+			s->local_flags |= divflags;
+		} else if (s->divert_cookie) {
+			divertcookie = s->divert_cookie;
+			divflags = s->local_flags;
+		}
+	}
+#endif
 	if ((s && s->tag) || r->rtableid)
 		pf_tag_packet(m, pd.pf_mtag, s ? s->tag : 0, r->rtableid);
 
@@ -7110,7 +7138,57 @@
 #endif /* ALTQ */
 
 #ifdef __FreeBSD__
-        if (r->dnpipe && DUMMYNET_LOADED &&
+#if 0
+	if (divflags & PFSTATE_DIVERT_TAG)
+		pd.pf_mtag->tag = divertcookie;
+	else 
+#endif
+	if (divflags & PFSTATE_DIVERT_ALTQ)
+		pd.pf_mtag->qid = divertcookie;
+	else if (divflags & PFSTATE_DIVERT_ACTION) {
+		struct pf_rule *dlr;
+		action = PF_DROP;
+		REASON_SET(&reason, PFRES_DIVERT);
+               	log = 1;
+               	DPFPRINTF(PF_DEBUG_MISC,
+                  		("pf: changing action to with overload from divert.\n"));
+		dlr = r;
+                PFLOG_PACKET(kif, h, m, AF_INET, dir, reason, dlr, a, 
+			ruleset, &pd);
+		PF_UNLOCK();
+		return (action);
+	}
+	
+	if (r->divert && ip_divert_ptr != NULL &&
+		!(pd.pf_mtag->flags & PF_DIVERT_TAG_PRESENT) && 
+		!divflags) {
+			if ((h->ip_p == IPPROTO_TCP && s && 
+				(dir == PF_IN ? s->src.state == TCPS_ESTABLISHED : 
+					s->dst.state == TCPS_ESTABLISHED) &&
+				(s->packets[dir == PF_OUT] - 2) <= r->spare1) || !r->spare1 ||
+				(h->ip_p != IPPROTO_TCP && s && s->packets[dir == PF_OUT] <= r->spare1)) {
+				
+				struct divert_tag *dt;
+                        	dn_tag = m_tag_get(PACKET_TAG_DIVERT, 
+					sizeof(struct divert_tag), M_NOWAIT);
+				if (dn_tag != NULL) {
+				
+                        		dt = (struct divert_tag *)(dn_tag+1);
+                        		dt->cookie = 0;
+                        		dt->info = r->divert;
+                        		m_tag_prepend(m, dn_tag);
+
+					PF_UNLOCK();
+
+					ip_divert_ptr(*m0, 
+						dir == 
+						PF_IN ? DIV_DIR_IN : DIV_DIR_OUT);
+
+					*m0 = NULL;	
+					return (action);
+				}
+		 	}
+        } else if (r->dnpipe && DUMMYNET_LOADED &&
                 !(pd.pf_mtag->flags & PF_DN_TAG_PRESENT)) {
                         struct ip_fw dummyrule;
 
@@ -7119,9 +7197,15 @@
                         dummyrule.act_ofs = 0;
                         dummyrule.cmd_len = 1;
                         if (dir != r->direction && r->pdnpipe) {
-                                dnflow.cookie = r->pdnpipe;
+				if (divflags & PFSTATE_DIVERT_DNCOOKIE && divertcookie)
+					dnflow.cookie = divertcookie;
+				else
+                                	dnflow.cookie = r->pdnpipe;
                         } else if (dir == r->direction) {
-                                dnflow.cookie = r->dnpipe;
+				if (divflags & PFSTATE_DIVERT_DNCOOKIE && divertcookie)
+					dnflow.cookie = divertcookie;
+				else
+                                	dnflow.cookie = r->dnpipe;
                         } else
                                 goto continueprocessing;
 
Index: sys/contrib/pf/net/pf_mtag.h
===================================================================
RCS file: /home/ermal/tmprepo/RELENG_7/sys/contrib/pf/net/pf_mtag.h,v
retrieving revision 1.2
diff -u -r1.2 pf_mtag.h
--- sys/contrib/pf/net/pf_mtag.h	11 Feb 2009 23:20:04 -0000	1.2
+++ sys/contrib/pf/net/pf_mtag.h	13 Feb 2009 20:55:59 -0000
@@ -38,6 +38,7 @@
 #define	PF_TAG_FRAGCACHE		0x02
 #define	PF_TAG_TRANSLATE_LOCALHOST	0x04
 #define PF_DN_TAG_PRESENT	        0x08
+#define PF_DIVERT_TAG_PRESENT		0x10
 
 struct pf_mtag {
 	void		*hdr;		/* saved hdr pos in mbuf, for ECN */
Index: sys/contrib/pf/net/pfvar.h
===================================================================
RCS file: /home/ermal/tmprepo/RELENG_7/sys/contrib/pf/net/pfvar.h,v
retrieving revision 1.2
diff -u -r1.2 pfvar.h
--- sys/contrib/pf/net/pfvar.h	11 Feb 2009 23:20:04 -0000	1.2
+++ sys/contrib/pf/net/pfvar.h	13 Feb 2009 20:55:59 -0000
@@ -649,6 +649,7 @@
 	u_int16_t		 tag;
 	u_int16_t		 match_tag;
 	u_int16_t		 spare2;		/* netgraph */
+	u_int16_t		 divert;		/* divert */
 
 	struct pf_rule_uid	 uid;
 	struct pf_rule_gid	 gid;
@@ -829,7 +830,12 @@
 	u_int8_t	 direction;
 #ifdef __FreeBSD__
 	u_int8_t	 local_flags;
-#define	PFSTATE_EXPIRING 0x01
+#define	PFSTATE_EXPIRING 	0x01
+#define PFSTATE_DIVERT_ALTQ     0x10
+#define PFSTATE_DIVERT_DNCOOKIE 0x20
+#define PFSTATE_DIVERT_ACTION   0x40
+#define PFSTATE_DIVERT_TAG      0x80
+#define PFSTATE_DIVERT_MASK	0xF000
 #else
 	u_int8_t	 pad;
 #endif
@@ -865,6 +871,7 @@
 	u_int32_t	 expire;
 	u_int32_t	 pfsync_time;
 	u_int16_t	 tag;
+	u_int16_t	 divert_cookie;
 };
 
 TAILQ_HEAD(pf_rulequeue, pf_rule);
@@ -1120,7 +1127,8 @@
 #define PFRES_MAXSTATES	12		/* State limit */
 #define PFRES_SRCLIMIT	13		/* Source node/conn limit */
 #define PFRES_SYNPROXY	14		/* SYN proxy */
-#define PFRES_MAX	15		/* total+1 */
+#define	PFRES_DIVERT	15		/* Divert override */
+#define PFRES_MAX	16		/* total+1 */
 
 #define PFRES_NAMES { \
 	"match", \
@@ -1138,6 +1146,7 @@
 	"state-limit", \
 	"src-limit", \
 	"synproxy", \
+	"divert", \
 	NULL \
 }
 
Index: sys/netinet/ip_divert.c
===================================================================
RCS file: /home/ermal/tmprepo/RELENG_7/sys/netinet/ip_divert.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 ip_divert.c
--- sys/netinet/ip_divert.c	11 Feb 2009 23:18:51 -0000	1.1.1.1
+++ sys/netinet/ip_divert.c	13 Feb 2009 20:55:59 -0000
@@ -752,5 +752,4 @@
 };
 
 DECLARE_MODULE(ipdivert, ipdivertmod, SI_SUB_PROTO_IFATTACHDOMAIN, SI_ORDER_ANY);
-MODULE_DEPEND(dummynet, ipfw, 2, 2, 2);
 MODULE_VERSION(ipdivert, 1);
Index: sys/netinet/ip_divert.h
===================================================================
RCS file: /home/ermal/tmprepo/RELENG_7/sys/netinet/ip_divert.h,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 ip_divert.h
--- sys/netinet/ip_divert.h	11 Feb 2009 23:18:51 -0000	1.1.1.1
+++ sys/netinet/ip_divert.h	13 Feb 2009 20:55:59 -0000
@@ -50,6 +50,9 @@
 	u_int16_t	cookie;		/* ipfw rule number */
 };
 
+#define DIV_DIR_IN      1
+#define DIV_DIR_OUT     0
+
 /*
  * Return the divert cookie associated with the mbuf; if any.
  */
Index: sys/netinet/ip_fw_pfil.c
===================================================================
RCS file: /home/ermal/tmprepo/RELENG_7/sys/netinet/ip_fw_pfil.c,v
retrieving revision 1.2
diff -u -r1.2 ip_fw_pfil.c
--- sys/netinet/ip_fw_pfil.c	11 Feb 2009 23:20:05 -0000	1.2
+++ sys/netinet/ip_fw_pfil.c	13 Feb 2009 20:55:59 -0000
@@ -72,16 +72,11 @@
 
 int ipfw_chg_hook(SYSCTL_HANDLER_ARGS);
 
-/* Divert hooks. */
-ip_divert_packet_t *ip_divert_ptr = NULL;
-
 /* ng_ipfw hooks. */
 ng_ipfw_input_t *ng_ipfw_input_p = NULL;
 
 /* Forward declarations. */
 static int	ipfw_divert(struct mbuf **, int, int);
-#define	DIV_DIR_IN	1
-#define	DIV_DIR_OUT	0
 
 int
 ipfw_check_in(void *arg, struct mbuf **m0, struct ifnet *ifp, int dir,
Index: sys/netinet/ip_input.c
===================================================================
RCS file: /home/ermal/tmprepo/RELENG_7/sys/netinet/ip_input.c,v
retrieving revision 1.2
diff -u -r1.2 ip_input.c
--- sys/netinet/ip_input.c	11 Feb 2009 23:20:05 -0000	1.2
+++ sys/netinet/ip_input.c	13 Feb 2009 20:55:59 -0000
@@ -81,6 +81,7 @@
 /* XXX: Temporary until ipfw_ether and ipfw_bridge are converted. */
 #include <netinet/ip_fw.h>
 #include <netinet/ip_dummynet.h>
+#include <netinet/ip_divert.h>
 
 #include <security/mac/mac_framework.h>
 
@@ -203,6 +204,10 @@
 ip_fw_chk_t *ip_fw_chk_ptr = NULL;
 ip_dn_io_t *ip_dn_io_ptr = NULL;
 ip_dn_ruledel_t *ip_dn_ruledel_ptr = NULL;
+
+/* Divert hooks. */
+ip_divert_packet_t *ip_divert_ptr = NULL;
+
 int fw_one_pass = 1;
 
 static void	ip_freef(struct ipqhead *, struct ipq *);
