Index: Makefile
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/Makefile,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 Makefile
--- Makefile	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ Makefile	17 Mar 2009 21:56:13 -0000
@@ -1,10 +1,5 @@
-# $FreeBSD: src/usr.sbin/fifolog/Makefile,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+# $FreeBSD: src/usr.sbin/fifolog/Makefile,v 1.2 2008/04/10 14:01:59 ru Exp $
 
-VERSION_MAJOR = 1
-VERSION_MINOR = 1
-
-SUBDIR	=	lib fifolog_create fifolog_writer fifolog_reader
+SUBDIR=	lib fifolog_create fifolog_writer fifolog_reader
 
 .include <bsd.subdir.mk>
-
-test:	_SUBDIR
Index: Makefile.inc
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/Makefile.inc,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 Makefile.inc
--- Makefile.inc	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ Makefile.inc	17 Mar 2009 21:56:13 -0000
@@ -1,13 +1,10 @@
-# $FreeBSD: src/usr.sbin/fifolog/Makefile.inc,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+# $FreeBSD: src/usr.sbin/fifolog/Makefile.inc,v 1.2 2008/04/10 14:01:59 ru Exp $
 
-LIBFIFOLOG =	${.OBJDIR}/../lib/libfifolog.a
+LIBFIFOLOG= ${.OBJDIR}/../lib/libfifolog.a
 
+WARNS?=	6
 
-WARNS ?=	6
-
-# CFLAGS	+=	-O0 -g
-
-# LINT =		flint
-# LINTFLAGS =	${.CURDIR}/../flint.lnt -I/usr/include 
+#LINT=	flint
+#LINTFLAGS= ${.CURDIR}/../flint.lnt -I/usr/include
 
 .include "../Makefile.inc"
Index: flint.lnt
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/flint.lnt,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 flint.lnt
--- flint.lnt	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ flint.lnt	17 Mar 2009 21:56:13 -0000
@@ -1,4 +1,4 @@
-// $FreeBSD: src/usr.sbin/fifolog/flint.lnt,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+// $FreeBSD: src/usr.sbin/fifolog/flint.lnt,v 1.1 2008/03/09 19:14:36 phk Exp $
 // FlexeLint file for fifolog tools
 //
 
Index: fifolog_create/Makefile
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/fifolog_create/Makefile,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 Makefile
--- fifolog_create/Makefile	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ fifolog_create/Makefile	17 Mar 2009 21:56:13 -0000
@@ -1,22 +1,22 @@
-# $FreeBSD: src/usr.sbin/fifolog/fifolog_create/Makefile,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+# $FreeBSD: src/usr.sbin/fifolog/fifolog_create/Makefile,v 1.2 2008/04/10 14:01:59 ru Exp $
 
-PROG =		fifolog_create
+PROG=	fifolog_create
 
-CFLAGS +=	-I${.CURDIR}/../lib
+CFLAGS+= -I${.CURDIR}/../lib
 
-DPADD =		${LIBFIFOLOG} ${LIBUTIL}
-LDADD =		${LIBFIFOLOG} -lutil
+DPADD=	${LIBFIFOLOG} ${LIBUTIL}
+LDADD=	${LIBFIFOLOG} -lutil
 
-MAN	=	fifolog.1
-MLINKS	+=	fifolog.1 fifolog_create.1
-MLINKS	+=	fifolog.1 fifolog_reader.1
-MLINKS	+=	fifolog.1 fifolog_writer.1
+MAN=	fifolog.1
+MLINKS=	fifolog.1 fifolog_create.1 \
+	fifolog.1 fifolog_reader.1 \
+	fifolog.1 fifolog_writer.1
 
-.include <bsd.prog.mk>
-
-test:	${PROG}
+regress:
 	rm -f /tmp/fifolog.?
 	./${PROG} /tmp/fifolog.0
 	./${PROG} -s 10m /tmp/fifolog.1
 	./${PROG} -l 1k /tmp/fifolog.2
 	./${PROG} -r 1k /tmp/fifolog.3
+
+.include <bsd.prog.mk>
Index: fifolog_create/fifolog.1
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/fifolog_create/fifolog.1,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 fifolog.1
--- fifolog_create/fifolog.1	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ fifolog_create/fifolog.1	17 Mar 2009 21:56:13 -0000
@@ -22,20 +22,14 @@
 .\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 .\" SUCH DAMAGE.
 .\"
-.\" $FreeBSD: src/usr.sbin/fifolog/fifolog_create/fifolog.1,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+.\" $FreeBSD: src/usr.sbin/fifolog/fifolog_create/fifolog.1,v 1.2 2008/04/10 14:01:59 ru Exp $
 .\"
-.Dd Feb 9, 2008
-.Os FreeBSD 8.0
-.Dt FIFOLOG 1 
+.Dd February 9, 2008
+.Os
+.Dt FIFOLOG 1
 .Sh NAME
-.Nm fifolog_create
-.Nd Initialize storage for fifolog
-.br
-.Nm fifolog_write
-.Nd Write data to fifolog
-.br
-.Nm fifolog_read
-.Nd Seek and extract data from fifolog
+.Nm fifolog_create , fifolog_write , fifolog_read
+.Nd "initialize, write, seek and extract data from a fifolog"
 .Sh SYNOPSIS
 .Nm fifolog_create
 .Op Fl l Ar record-size
@@ -43,7 +37,7 @@
 .Op Fl s Ar size
 .Ar file
 .Nm fifolog_reader
-.Op Fl t 
+.Op Fl t
 .Op Fl b Ar tstart
 .Op Fl B Ar Tstart
 .Op Fl e Ar tend
@@ -70,35 +64,40 @@
 to seek out a particular time interval in the stored data, without
 having to decompress the entire logfile.
 .Pp
-.Nm Fifolog_create
+The
+.Nm fifolog_create
+utility
 is used to initialize the first sector of a disk device
-or filesystem file to make it a fifolog and should be called only
+or file system file to make it a fifolog and should be called only
 once.
 .Pp
 Running
-.Nm
+.Nm fifolog_create
 on an existing fifolog will reset it so that
 .Nm fifolog_reader
 and
 .Nm fifolog_writer
 will not see the previous contents.
-(The previos contents is not physically erased, and with a bit
-of hand-work, all but the first record can be easily recovered).
+(The previous contents are not physically erased, and with a bit
+of hand-work all but the first record can be easily recovered.)
 .Pp
-If the file does not already exist
+If the
+.Ar file
+does not already exist,
 .Nm
 will attempt to create and
-.Xr ftruncate 3
+.Xr ftruncate 2
 it to the specified size, defaulting to 86400 records of 512 bytes
 if the
-.Fl r ,
-.Fl l 
+.Fl r , l
 or
-.Fl s 
-arguments do not specify otherwise.
+.Fl s
+options do not specify otherwise.
 .Pp
-.Nm Fifolog_writer
-will read standard input and write it to the end of the fifolog 
+The
+.Nm fifolog_writer
+utility
+will read standard input and write it to the end of the fifolog
 according to the parameters given.
 .Pp
 Writes happen whenever the output buffer is filled with compressed
@@ -106,7 +105,7 @@
 buffer to be written.
 .Pp
 The first and faster timer,
-.Fl w write-rate ,
+.Fl w Ar write-rate ,
 forces available data to be written
 but does not flush and reset the compression dictionary.
 This timer is intended to minimize the amount of logdata lost in RAM
@@ -114,27 +113,29 @@
 the previous write.
 .Pp
 The second and slower timer,
-.Fl s sync-rate ,
+.Fl s Ar sync-rate ,
 forces a full flush and reset of the compression
 engine and causes the next record written to be a synchronization
 point with an uncompressed timestamp, making it possible to start
 reading the logfile from that record.
 By default this timer fires a minute after the previous sync.
 .Pp
-The 
-.Fl z compression
-argument controls the
+The
+.Fl z Ar compression
+option controls the
 .Xr zlib 3
-compression level, legal values are zero to nine which is the default.
+compression level; legal values are zero to nine which is the default.
 .Pp
-.Nm Fifolog_reader
+The
+.Nm fifolog_reader
+utility
 will retrieve records from the fifolog according to the specified
-parameters and write them either to stdout or the file specified
+parameters and write them either to standard output or the file specified
 with
 .Fl o .
 .Pp
 It is possible to specify a start and end time to limit the amount
-of data 
+of data
 .Nm fifolog_reader
 will report.
 The lower-case variants
@@ -142,69 +143,76 @@
 and
 .Fl e
 take a
-.Xr time_t
+.Vt time_t
 value, whereas the upper-case variants
 .Fl B
 and
 .Fl E
-take human redable specifications such as "1 hour ago".
+take human-readable specifications such as
+.Dq Li "1 hour ago" .
 .Pp
 The
 .Fl t
-argument forces timestamps to be formatted as "YYYYMMDDhhmmss" instead
-of as time_t, and
+option forces timestamps to be formatted as
+.Dq Li "YYYYMMDDhhmmss"
+instead of as
+.Vt time_t ,
+and
 .Fl T
-allows the specification of a
+allows the specification of an
 .Xr strftime 3
 formatting string.
 .Pp
 Finally, records can be filtered such that only records matching the
-(REG_BASIC) regular expression specified with
+.Pq Dv REG_BASIC
+regular expression specified with
 .Fl R
-is output.
+are output.
 .Sh IMPLEMENTATION NOTES
-The data stored in the fifolog consists of three layers, an outher
+The data stored in the fifolog consists of three layers, an outer
 layer that allows searches to synchronization points based on timestamps
 without having to decompress and decode the actual contents, a
 compression layer implemented with
-.Xr zlib 3
+.Xr zlib 3 ,
 and an inner serialization and timestamping layer.
 .Pp
-The exact encoding is described in the fifolog.h file.
+The exact encoding is described in the
+.Pa fifolog.h
+file.
 .Pp
 Fifolog is particularly well suited for use on Flash based media, where
-it results in much lower write-wear, than a filesystem with regular
-logfiles rotated with 
+it results in much lower write-wear, than a file system with regular
+log files rotated with
 .Xr newsyslog 8
 etc.
 .Sh EXAMPLES
 Create a fifolog with 1024*1024 records of 512 bytes:
-.Bd -literal
-fifolog_create -r 10m /tmp/fifolog
-.Ed
+.Pp
+.Dl "fifolog_create -r 10m /tmp/fifolog"
 .Pp
 Write a single record to this file:
-.Bd -literal
-date | fifolog_writer /tmp/fifolog
-.Ed
+.Pp
+.Dl "date | fifolog_writer /tmp/fifolog"
 .Pp
 Read it back with human readable timestamps:
-.Bd -literal
-fifolog_reader -t /tmp/fifolog
-.Ed
+.Pp
+.Dl "fifolog_reader -t /tmp/fifolog"
 .Pp
 One particular useful use of
 .Nm fifolog_writer
 is with
-.Xr syslogd 8 
+.Xr syslogd 8
 using a line such as this in
 .Xr /etc/syslog.conf 5 :
-.Bd -literal
-*.* |fifolog_writer /var/log/syslog_fifolog
-.Ed
+.Pp
+.Dl "*.* |fifolog_writer /var/log/syslog_fifolog"
 .Sh HISTORY
-The fifolog tools have been liberated from an open source SCADA applications
-called "measured", which monitors and controls remote radio navigation 
+The fifolog tools have been liberated from an open source
+.Tn SCADA
+applications called
+.Dq measured ,
+which monitors and controls remote radio navigation
 transmitters for the Danish Air Traffic Control system.
 .Sh AUTHORS
-The fifolog tools were written by Poul-Henning Kamp
+The fifolog tools were written by
+.An Poul-Henning Kamp .
Index: fifolog_create/fifolog_create.c
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/fifolog_create/fifolog_create.c,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 fifolog_create.c
--- fifolog_create/fifolog_create.c	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ fifolog_create/fifolog_create.c	17 Mar 2009 21:56:13 -0000
@@ -23,10 +23,12 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/fifolog_create/fifolog_create.c,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/fifolog_create/fifolog_create.c,v 1.2 2008/05/14 23:29:01 phk Exp $
  */
 
 #include <stdio.h>
+#include <stdlib.h>
+#include <sysexits.h>
 #include <unistd.h>
 #include <err.h>
 #include <libutil.h>
@@ -36,6 +38,14 @@
 #define DEF_RECSIZE	512
 #define DEF_RECCNT	(24 * 60 * 60)
 
+static void
+usage(void)
+{
+	fprintf(stderr, "Usage: fifolog_create [-l record-size] "
+	    "[-r record-count] [-s size] file\n");
+	exit(EX_USAGE);
+}
+
 int
 main(int argc, char * const *argv)
 {
@@ -63,13 +73,13 @@
 				err(1, "Couldn't parse -s argument");
 			break;
 		default:
-			errx(1, "Usage");
+			usage();
 		}
 	}
 	argc -= optind;
 	argv += optind;
 	if (argc != 1) 
-		errx(1, "Usage");
+		usage();
 
 	if (size != 0 && reccnt != 0 && recsize != 0) {		/* N N N */
 		if (size !=  reccnt * recsize) 
Index: fifolog_reader/Makefile
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/fifolog_reader/Makefile,v
retrieving revision 1.2.2.1.2.1
diff -u -r1.2.2.1.2.1 Makefile
--- fifolog_reader/Makefile	25 Nov 2008 02:59:29 -0000	1.2.2.1.2.1
+++ fifolog_reader/Makefile	17 Mar 2009 21:56:13 -0000
@@ -1,21 +1,20 @@
-# $FreeBSD: src/usr.sbin/fifolog/fifolog_reader/Makefile,v 1.2.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+# $FreeBSD: src/usr.sbin/fifolog/fifolog_reader/Makefile,v 1.3 2008/04/10 14:01:59 ru Exp $
 
-PROG =		fifolog_reader
+PROG=	fifolog_reader
 
-CFLAGS +=	-I${.CURDIR}/../lib  -I${.CURDIR}
+CFLAGS+= -I${.CURDIR}/../lib
 
-NO_MAN =	see ../fifolog_create/fifolog.1
+NO_MAN=	# see ../fifolog_create/fifolog.1
 
-DPADD =		${LIBFIFOLOG} ${LIBUTIL} ${LIBZ}
-LDADD =		${LIBFIFOLOG} -lutil -lz
+DPADD=	${LIBFIFOLOG} ${LIBUTIL} ${LIBZ}
+LDADD=	${LIBFIFOLOG} -lutil -lz
 
-.include <bsd.prog.mk>
-
-test:	${PROG}
+regress:
 	./${PROG} /tmp/fifolog.0
 	./${PROG} -t /tmp/fifolog.0
 	./${PROG} /tmp/fifolog.1
 	./${PROG} -B "00:00" /tmp/fifolog.1
+	./${PROG} -T "%y%m%d-%H%M%S" /tmp/fifolog.1
+	./${PROG} -T "" /tmp/fifolog.1
 
-t2:
-	./${PROG} -t /critter/10.1.29.74.fifolog  
+.include <bsd.prog.mk>
Index: fifolog_reader/fifolog_reader.c
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/fifolog_reader/fifolog_reader.c,v
retrieving revision 1.2.2.1.2.1
diff -u -r1.2.2.1.2.1 fifolog_reader.c
--- fifolog_reader/fifolog_reader.c	25 Nov 2008 02:59:29 -0000	1.2.2.1.2.1
+++ fifolog_reader/fifolog_reader.c	17 Mar 2009 21:56:13 -0000
@@ -23,7 +23,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/fifolog_reader/fifolog_reader.c,v 1.2.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/fifolog_reader/fifolog_reader.c,v 1.3 2008/05/14 23:29:01 phk Exp $
  */
 
 #include <stdio.h>
@@ -33,6 +33,7 @@
 #include <time.h>
 #include <string.h>
 #include <stdlib.h>
+#include <sysexits.h>
 #include <regex.h>
 
 #include "libfifolog.h"
@@ -59,7 +60,9 @@
 	if (opt_R != NULL && regexec(&R, (const char *)p, 0, NULL, 0))
 		return;
 
-	if (opt_T != NULL) {
+	if (opt_T != NULL && *opt_T == '\0') {
+		fprintf(fo, "%s\n", p);
+	} else if (opt_T != NULL) {
 		(void)gmtime_r(&now, &utc);
 		i = strftime(buf, sizeof buf, opt_T, &utc);
 		assert(i > 0);
@@ -85,7 +88,7 @@
 		"\t-t # format timestamps as %%Y%%m%%d%%H%%M%%S\n"
 		"\t-T <timestamp format>\n"
 	);
-	exit (2);
+	exit (EX_USAGE);
 }
 
 int
@@ -94,9 +97,6 @@
 	int ch, i;
 	off_t o;
 	struct fifolog_reader *fl;
-	const char *progname;
-
-	progname=argv[0];
 
 	time(&opt_E);
 	opt_o = "-";
@@ -147,13 +147,14 @@
 		}
 	}
 
+	if (argv[0] == NULL)
+		Usage();
+
 	fprintf(stderr, "From\t%jd %s", (intmax_t)opt_B, ctime(&opt_B));
 	fprintf(stderr, "To\t%jd %s", (intmax_t)opt_E, ctime(&opt_E));
 	if (opt_B >= opt_E)
 		errx(1, "Begin time not before End time");
 
-	if (argv[0] == NULL)
-		errx(1, "Usage: %s [options] fifolog", progname);
 	fl = fifolog_reader_open(argv[0]);
 	
 	if (!strcmp(opt_o, "-"))
Index: fifolog_writer/Makefile
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/fifolog_writer/Makefile,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 Makefile
--- fifolog_writer/Makefile	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ fifolog_writer/Makefile	17 Mar 2009 21:56:13 -0000
@@ -1,16 +1,16 @@
-# $FreeBSD: src/usr.sbin/fifolog/fifolog_writer/Makefile,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+# $FreeBSD: src/usr.sbin/fifolog/fifolog_writer/Makefile,v 1.2 2008/04/10 14:01:59 ru Exp $
 
-PROG =		fifolog_writer
+PROG=	fifolog_writer
 
-CFLAGS +=	-I${.CURDIR}/../lib
+CFLAGS+= -I${.CURDIR}/../lib
 
-NO_MAN =	see ../fifolog_create/fifolog.1
+NO_MAN=	# see ../fifolog_create/fifolog.1
 
-DPADD =		${LIBFIFOLOG} ${LIBUTIL} ${LIBZ}
-LDADD =		${LIBFIFOLOG} -lutil -lz
+DPADD=	${LIBFIFOLOG} ${LIBUTIL} ${LIBZ}
+LDADD=	${LIBFIFOLOG} -lutil -lz
 
-.include <bsd.prog.mk>
-
-test:	${PROG}
+regress:
 	date | ./${PROG} -z 0 /tmp/fifolog.0
 	lptest 65 | ./${PROG} -z 9 /tmp/fifolog.1
+
+.include <bsd.prog.mk>
Index: fifolog_writer/fifolog_writer.c
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/fifolog_writer/fifolog_writer.c,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 fifolog_writer.c
--- fifolog_writer/fifolog_writer.c	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ fifolog_writer/fifolog_writer.c	17 Mar 2009 21:56:13 -0000
@@ -23,11 +23,12 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/fifolog_writer/fifolog_writer.c,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/fifolog_writer/fifolog_writer.c,v 1.2 2008/05/14 23:29:02 phk Exp $
  */
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <sysexits.h>
 #include <err.h>
 #include <unistd.h>
 #include <ctype.h>
@@ -38,6 +39,14 @@
 
 #include "libfifolog.h"
 
+static void
+usage(void)
+{
+	fprintf(stderr, "fifolog_writer [-w write-rate] [-s sync-rate] "
+	    "[-z compression] file\n");
+	exit(EX_USAGE);
+}
+
 int
 main(int argc, char * const *argv)
 {
@@ -62,19 +71,19 @@
 			z_opt = strtoul(optarg, NULL, 0);
 			break;
 		default:
-			errx(1, "Usage");
+			usage();
 		}
 	}
 	argc -= optind;
 	argv += optind;
 	if (argc != 1) 
-		errx(1, "Usage");
+		usage();
 
 	if (z_opt > 9)
-		errx(1, "Usage");
+		usage();
 
 	if (w_opt > s_opt)
-		errx(1, "Usage");
+		usage();
 
 	f = fifolog_write_new();
 	assert(f != NULL);
Index: lib/Makefile
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/lib/Makefile,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 Makefile
--- lib/Makefile	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ lib/Makefile	17 Mar 2009 21:56:13 -0000
@@ -1,17 +1,11 @@
-# $FreeBSD: src/usr.sbin/fifolog/lib/Makefile,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+# $FreeBSD: src/usr.sbin/fifolog/lib/Makefile,v 1.2 2008/04/10 14:02:00 ru Exp $
 
-LIB =		fifolog
-INTERNALLIB =	API not published or supported.
+LIB=	fifolog
+INTERNALLIB=	# API not published or supported.
 
-SRCS +=		fifolog_int.c
-SRCS +=		fifolog_create.c
-SRCS +=		fifolog_write_poll.c
-SRCS +=		fifolog_reader.c
-SRCS +=		getdate.y
+SRCS=	fifolog_int.c fifolog_create.c fifolog_write_poll.c fifolog_reader.c
+SRCS+=	getdate.y
 
-CFLAGS +=	-I${.CURDIR}
+CFLAGS+= -I${.CURDIR}
 
 .include <bsd.lib.mk>
-
-test:
-	echo ${HAS_EXPAND_NUMBER}
Index: lib/fifolog.h
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/lib/fifolog.h,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 fifolog.h
--- lib/fifolog.h	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ lib/fifolog.h	17 Mar 2009 21:56:13 -0000
@@ -23,7 +23,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog.h,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog.h,v 1.1 2008/03/09 19:14:36 phk Exp $
  */
 
 #ifndef __LOCAL_FIFOLOG_H_
Index: lib/fifolog_create.c
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/lib/fifolog_create.c,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 fifolog_create.c
--- lib/fifolog_create.c	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ lib/fifolog_create.c	17 Mar 2009 21:56:13 -0000
@@ -23,7 +23,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog_create.c,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog_create.c,v 1.1 2008/03/09 19:14:36 phk Exp $
  */
 
 #include <assert.h>
Index: lib/fifolog_int.c
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/lib/fifolog_int.c,v
retrieving revision 1.2.2.1.2.1
diff -u -r1.2.2.1.2.1 fifolog_int.c
--- lib/fifolog_int.c	25 Nov 2008 02:59:29 -0000	1.2.2.1.2.1
+++ lib/fifolog_int.c	17 Mar 2009 21:56:13 -0000
@@ -23,7 +23,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog_int.c,v 1.2.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog_int.c,v 1.3 2008/12/09 09:25:03 phk Exp $
  */
 
 #include <assert.h>
@@ -221,9 +221,9 @@
 	recno++;			/* label sector */
 	i = pread(ff->fd, ff->recbuf, ff->recsize, recno * ff->recsize);
 	if (i < 0)
-		return (-1);
+		return (-2);
 	if (i != (int)ff->recsize)
-		return (-1);
+		return (-3);
 	return (0);
 }
 
Index: lib/fifolog_reader.c
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/lib/fifolog_reader.c,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 fifolog_reader.c
--- lib/fifolog_reader.c	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ lib/fifolog_reader.c	17 Mar 2009 21:56:13 -0000
@@ -23,7 +23,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog_reader.c,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog_reader.c,v 1.3 2009/02/02 14:30:07 phk Exp $
  */
 
 #include <stdio.h>
@@ -98,7 +98,7 @@
 	assert(*o < ff->logsize);
 	e = fifolog_int_read(ff, *o);
 	if (e)
-		err(1, "Read error while looking for SYNC");
+		err(1, "Read error (%d) while looking for SYNC", e);
 	seq = be32dec(ff->recbuf);
 	if (*o == 0 && seq == 0)
 		return (0);
@@ -113,7 +113,7 @@
 			return (2);	/* wraparound */
 		e = fifolog_int_read(ff, *o);
 		if (e)
-			err(1, "Read error while looking for SYNC");
+			err(1, "Read error (%d) while looking for SYNC", e);
 		seqs = be32dec(ff->recbuf);
 		if (seqs != seq)
 			return (3);		/* End of log */
@@ -155,6 +155,7 @@
 		retval = fifolog_int_findend(fr->ff, &s);
 		if (retval != NULL)
 			err(1, "%s", retval);
+		s++;
 		e = fifolog_reader_findsync(fr->ff, &s);
 		if (e == 0)
 			return (0);		/* empty fifolog */
@@ -174,8 +175,10 @@
 			continue;
 		}
 		e = fifolog_int_read(fr->ff, o + st);
-		if (e)
-			err(1, "Read error, duing binary search");
+		if (e) {
+			s = st = s / 2;
+			continue;
+		}
 		/* If not in same part, sequence won't match */
 		seqs = be32dec(fr->ff->recbuf);
 		if (seqs != seq + st) {
@@ -254,7 +257,7 @@
 	while (1) {
 		e = fifolog_int_read(fr->ff, o);
 		if (e)
-			err(1, "Read error");
+			err(1, "Read error (%d)", e);
 		if (++o >= fr->ff->logsize)
 			o = 0;
 		seq = be32dec(fr->ff->recbuf);
Index: lib/fifolog_write.h
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/lib/fifolog_write.h,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 fifolog_write.h
--- lib/fifolog_write.h	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ lib/fifolog_write.h	17 Mar 2009 21:56:13 -0000
@@ -23,7 +23,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog_write.h,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog_write.h,v 1.1 2008/03/09 19:14:36 phk Exp $
  */
 
 #define FIFOLOG_PT_BYTES_PRE		0
Index: lib/fifolog_write_poll.c
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/lib/fifolog_write_poll.c,v
retrieving revision 1.2.2.1.2.1
diff -u -r1.2.2.1.2.1 fifolog_write_poll.c
--- lib/fifolog_write_poll.c	25 Nov 2008 02:59:29 -0000	1.2.2.1.2.1
+++ lib/fifolog_write_poll.c	17 Mar 2009 21:56:13 -0000
@@ -23,7 +23,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog_write_poll.c,v 1.2.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/lib/fifolog_write_poll.c,v 1.3 2009/02/02 14:29:15 phk Exp $
  */
 
 #include <assert.h>
@@ -152,15 +152,16 @@
 	es = fifolog_int_findend(f->ff, &o);
 	if (es != NULL)
 		return (es);
-	if (o == 0) {
-		f->seq = 0;
-		f->recno = 0;
+	i = fifolog_int_read(f->ff, o);
+	if (i)
+		return ("Read error, looking for seq");
+	f->seq = be32dec(f->ff->recbuf);
+	if (f->seq == 0) {
+		/* Empty fifolog */
+		f->seq = random();
 	} else {
-		i = fifolog_int_read(f->ff, o);
-		if (i)
-			return ("Read error, looking for seq");
-		f->seq = be32dec(f->ff->recbuf) + 1;
 		f->recno = o + 1;
+		f->seq++;
 	}
 
 	f->ibufsize = 32768;
Index: lib/getdate.y
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/lib/getdate.y,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 getdate.y
--- lib/getdate.y	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ lib/getdate.y	17 Mar 2009 21:56:13 -0000
@@ -12,7 +12,7 @@
 **  Picked up from CVS and slightly cleaned up by to WARNS=5 level by
 **  Poul-Henning Kamp <phk@FreeBSD.org>
 **
-** $FreeBSD: src/usr.sbin/fifolog/lib/getdate.y,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+** $FreeBSD: src/usr.sbin/fifolog/lib/getdate.y,v 1.1 2008/03/09 19:14:36 phk Exp $
 */
 
 #include <stdio.h>
Index: lib/libfifolog.h
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/lib/libfifolog.h,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 libfifolog.h
--- lib/libfifolog.h	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ lib/libfifolog.h	17 Mar 2009 21:56:13 -0000
@@ -23,7 +23,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/lib/libfifolog.h,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/lib/libfifolog.h,v 1.1 2008/03/09 19:14:36 phk Exp $
  */
 
 #include <sys/stdint.h>
Index: lib/libfifolog_int.h
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/lib/libfifolog_int.h,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 libfifolog_int.h
--- lib/libfifolog_int.h	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ lib/libfifolog_int.h	17 Mar 2009 21:56:13 -0000
@@ -23,7 +23,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/lib/libfifolog_int.h,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/lib/libfifolog_int.h,v 1.1 2008/03/09 19:14:36 phk Exp $
  */
 
 struct fifolog_file {
Index: lib/miniobj.h
===================================================================
RCS file: /home/ncvs/src/usr.sbin/fifolog/lib/miniobj.h,v
retrieving revision 1.2.2.1.2.1
diff -u -r1.2.2.1.2.1 miniobj.h
--- lib/miniobj.h	25 Nov 2008 02:59:29 -0000	1.2.2.1.2.1
+++ lib/miniobj.h	17 Mar 2009 21:56:13 -0000
@@ -23,7 +23,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/usr.sbin/fifolog/lib/miniobj.h,v 1.2.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/usr.sbin/fifolog/lib/miniobj.h,v 1.2 2008/03/16 23:00:43 phk Exp $
  */
 
 #define ALLOC_OBJ(to, type_magic) 					\
