Index: if_sis.c
===================================================================
RCS file: /home/ncvs/src/sys/pci/Attic/if_sis.c,v
retrieving revision 1.150.2.1.2.1
diff -u -r1.150.2.1.2.1 if_sis.c
--- if_sis.c	25 Nov 2008 02:59:29 -0000	1.150.2.1.2.1
+++ if_sis.c	19 Mar 2009 02:24:38 -0000
@@ -1431,7 +1431,11 @@
 		 * it should simply get re-used next time this descriptor
 	 	 * comes up in the ring.
 		 */
-		if (!(rxstat & SIS_CMDSTS_PKT_OK)) {
+		if ((ifp->if_capenable & IFCAP_VLAN_MTU) != 0 &&
+		 total_len <= (ETHER_MAX_LEN + ETHER_VLAN_ENCAP_LEN -
+		 ETHER_CRC_LEN))
+			rxstat &= ~SIS_RXSTAT_GIANT;
+               if (SIS_RXSTAT_ERROR(rxstat) != 0) {
 			ifp->if_ierrors++;
 			if (rxstat & SIS_RXSTAT_COLL)
 				ifp->if_collisions++;
Index: if_sisreg.h
===================================================================
RCS file: /home/ncvs/src/sys/pci/Attic/if_sisreg.h,v
retrieving revision 1.38.6.1
diff -u -r1.38.6.1 if_sisreg.h
--- if_sisreg.h	25 Nov 2008 02:59:29 -0000	1.38.6.1
+++ if_sisreg.h	19 Mar 2009 02:25:26 -0000
@@ -348,6 +348,11 @@
 #define SIS_RXSTAT_OVERRUN	0x02000000
 #define SIS_RXSTAT_RX_ABORT	0x04000000
 
+#define SIS_RXSTAT_ERROR(x)                                      \
+	((x) & (SIS_RXSTAT_RX_ABORT | SIS_RXSTAT_OVERRUN |              \
+	SIS_RXSTAT_GIANT | SIS_RXSTAT_SYMBOLERR | SIS_RXSTAT_RUNT |     \
+	SIS_RXSTAT_CRCERR | SIS_RXSTAT_ALIGNERR))
+
 #define SIS_DSTCLASS_REJECT	0x00000000
 #define SIS_DSTCLASS_UNICAST	0x00800000
 #define SIS_DSTCLASS_MULTICAST	0x01000000
