--- Makefile.orig	2011-06-28 09:02:38.060403307 +1200
+++ Makefile	2011-06-28 09:03:45.077440584 +1200
@@ -3,10 +3,10 @@
 
 .include <bsd.own.mk>
 
-.PATH: ${.CURDIR}/../../usr.bin/wall
+BINDIR=	%%PREFIX%%/sbin
 
 PROG=	syslogd
-MAN=	syslog.conf.5 syslogd.8
+MAN=
 SRCS=	syslogd.c ttymsg.c
 
 DPADD=	${LIBUTIL}
@@ -19,8 +19,8 @@
 .endif
 
 CFLAGS+= -DWITH_ZMQ
-LDADD+= -lczmq
+LDADD+= -lczmq -lzmq -lpthread -lstdc++
 
-CFLAGS+= -I${.CURDIR}/../../usr.bin/wall
+CFLAGS+= -I%%LOCALBASE%%/include -L%%LOCALBASE%%/lib
 
 .include <bsd.prog.mk>
