--- Makefile.orig	2011-06-28 08:51:14.858425643 +1200
+++ Makefile	2011-06-28 08:55:40.997517039 +1200
@@ -3,10 +3,10 @@
 
 .include <bsd.own.mk>
 
-.PATH: ${.CURDIR}/../../usr.bin/wall
+BINDIR=	%%PREFIX%%/sbin
 
 PROG=	syslogd
-MAN=	syslog.conf.5 syslogd.8
+MAN=
 SRCS=	syslogd.c ttymsg.c
 
 DPADD=	${LIBUTIL}
@@ -21,6 +21,6 @@
 CFLAGS+= -DWITH_ZMQ
 LDADD+= -lczmq
 
-CFLAGS+= -I${.CURDIR}/../../usr.bin/wall
+CFLAGS+= -I%%LOCALBASE%%/include -L%%LOCALBASE%%/lib
 
 .include <bsd.prog.mk>
