--- src/racoon/isakmp.c	2008-04-23 11:39:29.000000000 +0200
+++ src/racoon/isakmp.c	2008-04-23 12:43:50.000000000 +0200
@@ -3358,7 +3358,7 @@
 	struct ph2handle *iph2;
 	struct ph1handle *new_iph1;
 
-	plog(LLV_INFO, LOCATION, NULL,
+	plog(LLV_DEBUG, LOCATION, NULL,
 		 "purging ISAKMP-SA spi=%s.\n",
 		 isakmp_pindex(&(iph1->index), iph1->msgid));
 
@@ -3452,7 +3452,7 @@
 				if (iph2->status == PHASE2ST_ESTABLISHED ||
 					iph2->status == PHASE2ST_EXPIRED) {
 					
-					plog(LLV_INFO, LOCATION, NULL,
+					plog(LLV_DEBUG, LOCATION, NULL,
 						 "keeping IPsec-SA spi=%u - found valid ISAKMP-SA spi=%s.\n",
 						 ntohl(sa->sadb_sa_spi),
 						 isakmp_pindex(&(new_iph1->index), new_iph1->msgid));
@@ -3476,7 +3476,7 @@
 			delph2(iph2);
 		}
 
-		plog(LLV_INFO, LOCATION, NULL,
+		plog(LLV_DEBUG, LOCATION, NULL,
 			 "purged IPsec-SA spi=%u.\n",
 			 ntohl(sa->sadb_sa_spi));
 
@@ -3487,7 +3487,7 @@
 		vfree(buf);
 
 	/* Mark the phase1 handler as EXPIRED */
-	plog(LLV_INFO, LOCATION, NULL,
+	plog(LLV_DEBUG, LOCATION, NULL,
 		 "purged ISAKMP-SA spi=%s.\n",
 		 isakmp_pindex(&(iph1->index), iph1->msgid));
 
