Index: biosdisk.c
===================================================================
RCS file: /home/ncvs/src/sys/boot/i386/libi386/biosdisk.c,v
retrieving revision 1.48.2.5.2.1
diff -u -r1.48.2.5.2.1 biosdisk.c
--- biosdisk.c	25 Nov 2008 02:59:29 -0000	1.48.2.5.2.1
+++ biosdisk.c	17 Apr 2009 13:54:08 -0000
@@ -81,7 +81,7 @@
     int			od_cyl;			/* BIOS geometry */
     int			od_hds;
     int			od_sec;
-    int			od_boff;		/* block offset from beginning of BIOS disk */
+    daddr_t		od_boff;		/* block offset from beginning of BIOS disk */
     int			od_flags;
 #define BD_MODEINT13		0x0000
 #define BD_MODEEDD1		0x0001
