Index: bioscd.c
===================================================================
RCS file: /home/ncvs/src/sys/boot/i386/libi386/bioscd.c,v
retrieving revision 1.9.2.1.4.1
diff -u -r1.9.2.1.4.1 bioscd.c
--- bioscd.c	25 Nov 2008 02:59:29 -0000	1.9.2.1.4.1
+++ bioscd.c	19 Mar 2009 16:12:18 -0000
@@ -26,7 +26,7 @@
  */
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/sys/boot/i386/libi386/bioscd.c,v 1.9.2.1.4.1 2008/11/25 02:59:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/sys/boot/i386/libi386/bioscd.c,v 1.9.2.1 2007/11/01 18:19:18 jhb Exp $");
 
 /*
  * BIOS CD device handling for CD's that have been booted off of via no
Index: biosdisk.c
===================================================================
RCS file: /home/ncvs/src/sys/boot/i386/libi386/biosdisk.c,v
retrieving revision 1.48.2.5.2.1
diff -u -r1.48.2.5.2.1 biosdisk.c
--- biosdisk.c	25 Nov 2008 02:59:29 -0000	1.48.2.5.2.1
+++ biosdisk.c	19 Mar 2009 16:12:18 -0000
@@ -25,7 +25,7 @@
  */
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/sys/boot/i386/libi386/biosdisk.c,v 1.48.2.5.2.1 2008/11/25 02:59:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/sys/boot/i386/libi386/biosdisk.c,v 1.48.2.5 2008/03/08 23:36:25 bz Exp $");
 
 /*
  * BIOS disk device handling.
Index: libi386.h
===================================================================
RCS file: /home/ncvs/src/sys/boot/i386/libi386/libi386.h,v
retrieving revision 1.27.6.1
diff -u -r1.27.6.1 libi386.h
--- libi386.h	25 Nov 2008 02:59:29 -0000	1.27.6.1
+++ libi386.h	19 Mar 2009 16:12:18 -0000
@@ -23,7 +23,7 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  *
- * $FreeBSD: src/sys/boot/i386/libi386/libi386.h,v 1.27.6.1 2008/11/25 02:59:29 kensmith Exp $
+ * $FreeBSD: src/sys/boot/i386/libi386/libi386.h,v 1.27 2006/11/02 01:23:17 marcel Exp $
  */
 
 
