Index: bootinfo.c
===================================================================
RCS file: /home/ncvs/src/sys/boot/i386/libi386/bootinfo.c,v
retrieving revision 1.39
diff -u -r1.39 bootinfo.c
--- bootinfo.c	22 Sep 2005 15:14:12 -0000	1.39
+++ bootinfo.c	29 Aug 2008 21:27:11 -0000
@@ -144,9 +144,9 @@
     /*
      * XXX: Note that until the kernel is ready to respect multiple consoles
      * for the boot messages, the first named console is the primary console
+     * if (!strcmp(string, "vidconsole"))
+     *    howto &= ~RB_SERIAL;
      */
-    if (!strcmp(string, "vidconsole"))
-	howto &= ~RB_SERIAL;
 
     free(string);
 
