Index: cardbus_cis.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/cardbus/cardbus_cis.c,v
retrieving revision 1.62
diff -u -r1.62 cardbus_cis.c
--- cardbus_cis.c	8 Jun 2007 04:03:57 -0000	1.62
+++ cardbus_cis.c	3 Oct 2008 01:24:29 -0000
@@ -435,10 +435,12 @@
     struct resource *res)
 {
 	if (res != CIS_CONFIG_SPACE) {
-		bus_release_resource(child, SYS_RES_MEMORY, rid, res);
+		//bus_release_resource(child, SYS_RES_MEMORY, rid, res);
+		bus_release_resource(cbdev, SYS_RES_MEMORY, rid, res);
 		if (rid == PCIM_CIS_ASI_ROM)
 			pci_write_config(child, rid, pci_read_config(child,
 			    rid, 4) & ~PCIR_BIOS, 4);
+		PCI_DISABLE_IO(cbdev, child, SYS_RES_MEMORY);
 	}
 }
 
@@ -448,6 +450,8 @@
 {
 	struct resource *res;
 	uint32_t space;
+	uint32_t testval;
+	uint32_t size;
 
 	space = *start & PCIM_CIS_ASI_MASK;
 	switch (space) {
@@ -477,9 +481,15 @@
 		return (NULL);
 	}
 
+	pci_write_config(child, *rid, 0xffffffff, 4);
+	testval = pci_read_config(child, *rid, 4);
+	size = CARDBUS_MAPREG_MEM_SIZE(testval);
+
 	/* allocate the memory space to read CIS */
-	res = bus_alloc_resource(child, SYS_RES_MEMORY, rid, 0, ~0, 1,
-	    rman_make_alignment_flags(4096) | RF_ACTIVE);
+	//res = bus_alloc_resource(child, SYS_RES_MEMORY, rid, 0, ~0, 1,
+	//rman_make_alignment_flags(4096) | RF_ACTIVE);
+	res = bus_alloc_resource(cbdev, SYS_RES_MEMORY, rid, 0, ~0, size,
+	rman_make_alignment_flags(size) | RF_ACTIVE);
 	if (res == NULL) {
 		device_printf(cbdev, "Unable to allocate resource "
 		    "to read CIS.\n");
@@ -489,6 +499,8 @@
 		pci_write_config(child, *rid,
 		    rman_get_start(res) | PCIM_BIOS_ENABLE, 4);
 
+	PCI_ENABLE_IO(cbdev, child, SYS_RES_MEMORY);
+
 	/* Flip to the right ROM image if CIS is in ROM */
 	if (space == PCIM_CIS_ASI_ROM) {
 		bus_space_tag_t bt;
