Index: sys/netinet/ip_carp.c
===================================================================
RCS file: /home/eri/mine/FreeBSD/src/sys/netinet/ip_carp.c,v
retrieving revision 1.52
diff -u -r1.52 ip_carp.c
--- sys/netinet/ip_carp.c	7 Oct 2007 20:44:22 -0000	1.52
+++ sys/netinet/ip_carp.c	13 Jul 2008 20:16:07 -0000
@@ -1438,7 +1438,11 @@
 			carp_set_state(sc, INIT);
 		if (sc->sc_naddrs)
 			SC2IFP(sc)->if_flags |= IFF_UP;
+		if (sc->sc_carpdev)
+			CARP_SCLOCK(sc);
 		carp_setrun(sc, 0);
+		if (sc->sc_carpdev)
+			CARP_SCUNLOCK(sc);
 		return (0);
 	}
 
