Index: sys/fs/fifofs/fifo_vnops.c
===================================================================
RCS file: /home/ncvs/src/sys/fs/fifofs/fifo_vnops.c,v
retrieving revision 1.138.2.2.2.1
diff -u -r1.138.2.2.2.1 fifo_vnops.c
--- sys/fs/fifofs/fifo_vnops.c	15 Apr 2009 03:14:26 -0000	1.138.2.2.2.1
+++ sys/fs/fifofs/fifo_vnops.c	7 Nov 2009 00:47:56 -0000
@@ -237,6 +237,8 @@
 	if (ap->a_mode & FWRITE) {
 		if ((ap->a_mode & O_NONBLOCK) && fip->fi_readers == 0) {
 			mtx_unlock(&fifo_mtx);
+			if (fip->fi_writers == 0)
+           	fifo_cleanup(vp);
 			return (ENXIO);
 		}
 		fip->fi_writers++;
