Index: ieee80211_power.c
===================================================================
RCS file: /home/ncvs/src/sys/net80211/ieee80211_power.c,v
retrieving revision 1.2.6.1
diff -u -r1.2.6.1 ieee80211_power.c
--- ieee80211_power.c	25 Nov 2008 02:59:29 -0000	1.2.6.1
+++ ieee80211_power.c	25 Jan 2009 20:20:08 -0000
@@ -48,30 +48,30 @@
 void
 ieee80211_power_attach(struct ieee80211com *ic)
 {
-	if (ic->ic_opmode == IEEE80211_M_HOSTAP ||
-	    ic->ic_opmode == IEEE80211_M_IBSS) {
-		/* NB: driver should override */
-		ic->ic_set_tim = ieee80211_set_tim;
-	}
+       if (ic->ic_caps & IEEE80211_C_HOSTAP ||
+           ic->ic_caps & IEEE80211_C_IBSS) {
+               /* NB: driver should override */
+               ic->ic_set_tim = ieee80211_set_tim;
+       }
 }
 
 void
 ieee80211_power_lateattach(struct ieee80211com *ic)
 {
-	/*
-	 * Allocate these only if needed.  Beware that we
-	 * know adhoc mode doesn't support ATIM yet...
-	 */
-	if (ic->ic_opmode == IEEE80211_M_HOSTAP) {
-		ic->ic_tim_len = howmany(ic->ic_max_aid,8) * sizeof(uint8_t);
-		MALLOC(ic->ic_tim_bitmap, uint8_t *, ic->ic_tim_len,
-			M_DEVBUF, M_NOWAIT | M_ZERO);
-		if (ic->ic_tim_bitmap == NULL) {
-			printf("%s: no memory for TIM bitmap!\n", __func__);
-			/* XXX good enough to keep from crashing? */
-			ic->ic_tim_len = 0;
-		}
-	}
+       /*
+        * Allocate these only if needed.  Beware that we
+        * know adhoc mode doesn't support ATIM yet...
+        */
+       if (ic->ic_caps & IEEE80211_C_HOSTAP) {
+               ic->ic_tim_len = howmany(ic->ic_max_aid,8) * sizeof(uint8_t);
+               MALLOC(ic->ic_tim_bitmap, uint8_t *, ic->ic_tim_len,
+                       M_DEVBUF, M_NOWAIT | M_ZERO);
+               if (ic->ic_tim_bitmap == NULL) {
+                       printf("%s: no memory for TIM bitmap!\n", __func__);
+                       /* XXX good enough to keep from crashing? */
+                       ic->ic_tim_len = 0;
+               }
+       }
 }
 
 void
