Index: if_enc.c
===================================================================
RCS file: /home/ncvs/src/sys/net/if_enc.c,v
retrieving revision 1.6.4.1
diff -u -r1.6.4.1 if_enc.c
--- if_enc.c	29 Dec 2007 17:28:38 -0000	1.6.4.1
+++ if_enc.c	8 Sep 2008 20:38:06 -0000
@@ -204,6 +204,14 @@
 	if ((encif->if_drv_flags & IFF_DRV_RUNNING) == 0)
 		return (0);
 
+	if (dir == PFIL_IN) {
+		encif->if_ipackets++;
+		encif->if_ibytes += (*mp)->m_pkthdr.len;
+	} else {
+		encif->if_opackets++;
+		encif->if_obytes += (*mp)->m_pkthdr.len;
+	}
+
 	/* Skip pfil(9) if no filters are loaded */
 	if (!(PFIL_HOOKED(&inet_pfil_hook)
 #ifdef INET6
