Index: LICENSE
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/LICENSE,v
retrieving revision 1.1.2.1.2.1
diff -u -r1.1.2.1.2.1 LICENSE
--- LICENSE	25 Nov 2008 02:59:29 -0000	1.1.2.1.2.1
+++ LICENSE	21 Mar 2009 18:58:27 -0000
@@ -30,4 +30,4 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/LICENSE,v 1.1.2.1.2.1 2008/11/25 02:59:29 kensmith Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/LICENSE,v 1.1.2.1 2008/08/11 21:02:22 jfv Exp $*/
Index: ixgbe.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/ixgbe.c,v
retrieving revision 1.4.2.2.2.2
diff -u -r1.4.2.2.2.2 ixgbe.c
--- ixgbe.c	30 Nov 2008 04:44:46 -0000	1.4.2.2.2.2
+++ ixgbe.c	21 Mar 2009 18:58:27 -0000
@@ -30,7 +30,7 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/ixgbe.c,v 1.4.2.2.2.2 2008/11/30 04:44:46 jfv Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/ixgbe.c,v 1.4.2.3 2008/12/01 04:03:17 jfv Exp $*/
 
 #ifdef HAVE_KERNEL_OPTION_HEADERS
 #include "opt_device_polling.h"
Index: ixgbe.h
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/ixgbe.h,v
retrieving revision 1.3.2.2.2.2
diff -u -r1.3.2.2.2.2 ixgbe.h
--- ixgbe.h	30 Nov 2008 04:44:46 -0000	1.3.2.2.2.2
+++ ixgbe.h	21 Mar 2009 18:58:27 -0000
@@ -30,7 +30,7 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/ixgbe.h,v 1.3.2.2.2.2 2008/11/30 04:44:46 jfv Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/ixgbe.h,v 1.3.2.3 2008/12/01 04:03:17 jfv Exp $*/
 
 
 #ifndef _IXGBE_H_
Index: ixgbe_82598.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/ixgbe_82598.c,v
retrieving revision 1.3.2.2.2.2
diff -u -r1.3.2.2.2.2 ixgbe_82598.c
--- ixgbe_82598.c	30 Nov 2008 04:44:46 -0000	1.3.2.2.2.2
+++ ixgbe_82598.c	21 Mar 2009 18:58:27 -0000
@@ -30,7 +30,7 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_82598.c,v 1.3.2.2.2.2 2008/11/30 04:44:46 jfv Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_82598.c,v 1.3.2.3 2008/12/01 04:03:17 jfv Exp $*/
 
 #include "ixgbe_type.h"
 #include "ixgbe_api.h"
Index: ixgbe_api.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/ixgbe_api.c,v
retrieving revision 1.3.2.2.2.2
diff -u -r1.3.2.2.2.2 ixgbe_api.c
--- ixgbe_api.c	30 Nov 2008 04:44:46 -0000	1.3.2.2.2.2
+++ ixgbe_api.c	21 Mar 2009 18:58:27 -0000
@@ -30,7 +30,7 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_api.c,v 1.3.2.2.2.2 2008/11/30 04:44:46 jfv Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_api.c,v 1.3.2.3 2008/12/01 04:03:17 jfv Exp $*/
 
 #include "ixgbe_api.h"
 #include "ixgbe_common.h"
Index: ixgbe_api.h
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/ixgbe_api.h,v
retrieving revision 1.3.2.2.2.2
diff -u -r1.3.2.2.2.2 ixgbe_api.h
--- ixgbe_api.h	30 Nov 2008 04:44:46 -0000	1.3.2.2.2.2
+++ ixgbe_api.h	21 Mar 2009 18:58:27 -0000
@@ -30,7 +30,7 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_api.h,v 1.3.2.2.2.2 2008/11/30 04:44:46 jfv Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_api.h,v 1.3.2.3 2008/12/01 04:03:17 jfv Exp $*/
 
 #ifndef _IXGBE_API_H_
 #define _IXGBE_API_H_
Index: ixgbe_common.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/ixgbe_common.c,v
retrieving revision 1.3.2.2.2.2
diff -u -r1.3.2.2.2.2 ixgbe_common.c
--- ixgbe_common.c	30 Nov 2008 04:44:46 -0000	1.3.2.2.2.2
+++ ixgbe_common.c	21 Mar 2009 18:58:27 -0000
@@ -30,7 +30,7 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_common.c,v 1.3.2.2.2.2 2008/11/30 04:44:46 jfv Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_common.c,v 1.3.2.3 2008/12/01 04:03:17 jfv Exp $*/
 
 #include "ixgbe_common.h"
 #include "ixgbe_api.h"
Index: ixgbe_common.h
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/ixgbe_common.h,v
retrieving revision 1.3.2.2.2.2
diff -u -r1.3.2.2.2.2 ixgbe_common.h
--- ixgbe_common.h	30 Nov 2008 04:44:46 -0000	1.3.2.2.2.2
+++ ixgbe_common.h	21 Mar 2009 18:58:27 -0000
@@ -30,7 +30,7 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_common.h,v 1.3.2.2.2.2 2008/11/30 04:44:46 jfv Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_common.h,v 1.3.2.3 2008/12/01 04:03:17 jfv Exp $*/
 
 #ifndef _IXGBE_COMMON_H_
 #define _IXGBE_COMMON_H_
Index: ixgbe_osdep.h
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/ixgbe_osdep.h,v
retrieving revision 1.3.2.2.2.2
diff -u -r1.3.2.2.2.2 ixgbe_osdep.h
--- ixgbe_osdep.h	30 Nov 2008 04:44:46 -0000	1.3.2.2.2.2
+++ ixgbe_osdep.h	21 Mar 2009 18:58:27 -0000
@@ -30,7 +30,7 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_osdep.h,v 1.3.2.2.2.2 2008/11/30 04:44:46 jfv Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_osdep.h,v 1.3.2.3 2008/12/01 04:03:17 jfv Exp $*/
 
 #ifndef _IXGBE_OS_H_
 #define _IXGBE_OS_H_
Index: ixgbe_phy.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/ixgbe_phy.c,v
retrieving revision 1.3.2.2.2.2
diff -u -r1.3.2.2.2.2 ixgbe_phy.c
--- ixgbe_phy.c	30 Nov 2008 04:44:46 -0000	1.3.2.2.2.2
+++ ixgbe_phy.c	21 Mar 2009 18:58:27 -0000
@@ -30,7 +30,7 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_phy.c,v 1.3.2.2.2.2 2008/11/30 04:44:46 jfv Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_phy.c,v 1.3.2.3 2008/12/01 04:03:17 jfv Exp $*/
 
 #include "ixgbe_api.h"
 #include "ixgbe_common.h"
Index: ixgbe_phy.h
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/ixgbe_phy.h,v
retrieving revision 1.3.2.2.2.2
diff -u -r1.3.2.2.2.2 ixgbe_phy.h
--- ixgbe_phy.h	30 Nov 2008 04:44:46 -0000	1.3.2.2.2.2
+++ ixgbe_phy.h	21 Mar 2009 18:58:27 -0000
@@ -30,7 +30,7 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_phy.h,v 1.3.2.2.2.2 2008/11/30 04:44:46 jfv Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_phy.h,v 1.3.2.3 2008/12/01 04:03:17 jfv Exp $*/
 
 #ifndef _IXGBE_PHY_H_
 #define _IXGBE_PHY_H_
Index: ixgbe_type.h
===================================================================
RCS file: /home/ncvs/src/sys/dev/ixgbe/ixgbe_type.h,v
retrieving revision 1.3.2.2.2.2
diff -u -r1.3.2.2.2.2 ixgbe_type.h
--- ixgbe_type.h	30 Nov 2008 04:44:46 -0000	1.3.2.2.2.2
+++ ixgbe_type.h	21 Mar 2009 18:58:27 -0000
@@ -30,7 +30,7 @@
   POSSIBILITY OF SUCH DAMAGE.
 
 ******************************************************************************/
-/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_type.h,v 1.3.2.2.2.2 2008/11/30 04:44:46 jfv Exp $*/
+/*$FreeBSD: src/sys/dev/ixgbe/ixgbe_type.h,v 1.3.2.3 2008/12/01 04:03:17 jfv Exp $*/
 
 #ifndef _IXGBE_TYPE_H_
 #define _IXGBE_TYPE_H_
