Index: sh.c
===================================================================
RCS file: /home/ncvs/src/contrib/tcsh/sh.c,v
retrieving revision 1.1.1.9.2.1
diff -u -r1.1.1.9.2.1 sh.c
--- sh.c	16 Oct 2007 16:18:39 -0000	1.1.1.9.2.1
+++ sh.c	6 Oct 2009 20:33:45 -0000
@@ -1306,6 +1306,7 @@
     /*
      * Mop-up.
      */
+	handle_pending_signals();
     if (intty) {
 	if (loginsh) {
 	    xprintf("logout\n");
