Index: usr.bin/sockstat/sockstat.c
===================================================================
RCS file: /home/eri/mine/FreeBSD/src/usr.bin/sockstat/sockstat.c,v
retrieving revision 1.17
diff -u -r1.17 sockstat.c
--- usr.bin/sockstat/sockstat.c	16 Jun 2007 20:24:55 -0000	1.17
+++ usr.bin/sockstat/sockstat.c	23 Jul 2008 18:12:23 -0000
@@ -69,7 +69,7 @@
 /*
  * Default protocols to use if no -P was defined.
  */
-static const char *default_protos[] = {"tcp", "udp", "divert" };
+static const char *default_protos[] = {"tcp", "udp" };
 static size_t	   default_numprotos =
     sizeof(default_protos) / sizeof(default_protos[0]);
 
