Index: src/racoon/pfkey.c
===================================================================
--- src/racoon/pfkey.c	2008-01-22 13:11:11.000000000 +0200
+++ src/racoon/pfkey.c	2008-01-22 13:12:06.000000000 +0200
@@ -448,6 +448,21 @@
 	return 0;
 }
 
+int
+pfkey_reload()
+{
+	flushsp();
+
+	if (pfkey_send_spddump(lcconf->sock_pfkey) < 0) {
+		plog(LLV_ERROR, LOCATION, NULL,
+			"libipsec sending spddump failed: %s\n",
+			ipsec_strerror());
+		return -1;
+	}
+
+	return 0;
+}
+
 /* %%% for conversion */
 /* IPSECDOI_ATTR_AUTH -> SADB_AALG */
 static u_int
Index: src/racoon/pfkey.h
===================================================================
--- src/racoon/pfkey.h	2008-01-22 13:11:11.000000000 +0200
+++ src/racoon/pfkey.h	2008-01-22 13:12:24.000000000 +0200
@@ -46,6 +46,7 @@
 extern vchar_t *pfkey_dump_sadb __P((int));
 extern void pfkey_flush_sadb __P((u_int));
 extern int pfkey_init __P((void));
+extern int pfkey_reload __P((void));
 
 extern struct pfkey_st *pfkey_getpst __P((caddr_t *, int, int));
 
Index: src/racoon/session.c
===================================================================
--- src/racoon/session.c	2008-01-22 13:12:36.000000000 +0200
+++ src/racoon/session.c	2008-01-22 13:14:33.000000000 +0200
@@ -361,10 +361,8 @@
 	save_rmconf();
 	initrmconf();
 
-	/* Do a part of pfkey_init() ?
-	 * SPD reload ?
-	 */
-	
+	pfkey_reload();
+
 	save_params();
 	error = cfparse();
 	if (error != 0){
