--- ipsec-tools/src/racoon/isakmp_inf.c	2008-04-23 11:39:30.000000000 +0200
+++ ipsec-tools-0.7-cvs2/src/racoon/isakmp_inf.c	2008-04-23 12:57:53.000000000 +0200
@@ -1129,7 +1129,7 @@
 		if (!iph1)
 			continue;
 
-		plog(LLV_INFO, LOCATION, NULL,
+		plog(LLV_DEBUG, LOCATION, NULL,
 			"purged ISAKMP-SA proto_id=%s spi=%s.\n",
 			s_ipsecdoi_proto(proto),
 			isakmp_pindex(&spi[i], 0));
@@ -1249,7 +1249,7 @@
 				delph2(iph2);
 			}
 
-			plog(LLV_INFO, LOCATION, NULL,
+			plog(LLV_DEBUG, LOCATION, NULL,
 				"purged IPsec-SA proto_id=%s spi=%u.\n",
 				s_ipsecdoi_proto(proto),
 				ntohl(spi[i]));
@@ -1301,7 +1301,7 @@
 	for (i = 0; i < pfkey_nsatypes; i++) {
 		proto_id = pfkey2ipsecdoi_proto(pfkey_satypes[i].ps_satype);
 
-		plog(LLV_INFO, LOCATION, NULL,
+		plog(LLV_DEBUG, LOCATION, NULL,
 		    "purging %s SAs for %s -> %s\n",
 		    pfkey_satypes[i].ps_name, loc, rem);
 		if (pfkey_send_delete_all(lcconf->sock_pfkey,
@@ -1316,7 +1316,7 @@
 
 		deleteallph2(iph1->local, iph1->remote, proto_id);
 
-		plog(LLV_INFO, LOCATION, NULL,
+		plog(LLV_DEBUG, LOCATION, NULL,
 		    "purging %s SAs for %s -> %s\n",
 		    pfkey_satypes[i].ps_name, rem, loc);
 		if (pfkey_send_delete_all(lcconf->sock_pfkey,
@@ -1440,7 +1440,7 @@
 			continue;
 		}
 
-		plog(LLV_INFO, LOCATION, NULL,
+		plog(LLV_DEBUG, LOCATION, NULL,
 			"purging spi=%u.\n", ntohl(sa->sadb_sa_spi));
 		pfkey_send_delete(lcconf->sock_pfkey,
 			msg->sadb_msg_satype,
