Index: ata-disk.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/ata/ata-disk.c,v
retrieving revision 1.216.2.2.2.1
diff -u -r1.216.2.2.2.1 ata-disk.c
--- ata-disk.c	25 Oct 2009 01:10:29 -0000	1.216.2.2.2.1
+++ ata-disk.c	11 Dec 2009 19:07:08 -0000
@@ -132,10 +132,10 @@
     adp->disk->d_fwheads = adp->heads;
     adp->disk->d_unit = device_get_unit(dev);
     if (atadev->param.support.command2 & ATA_SUPPORT_FLUSHCACHE)
-	adp->disk->d_flags = DISKFLAG_CANFLUSHCACHE;
+	adp->disk->d_flags |= DISKFLAG_CANFLUSHCACHE;
     if ((atadev->param.support.command2 & ATA_SUPPORT_CFA) ||
 	atadev->param.config == ATA_PROTO_CFA)
-	adp->disk->d_flags = DISKFLAG_CANDELETE;
+	adp->disk->d_flags |= DISKFLAG_CANDELETE;
     strlcpy(adp->disk->d_ident, atadev->param.serial,
 	sizeof(adp->disk->d_ident));
     disk_create(adp->disk, DISK_VERSION);
