Index: g_label.c
===================================================================
RCS file: /home/ncvs/src/sys/geom/label/g_label.c,v
retrieving revision 1.22
diff -u -r1.22 g_label.c
--- g_label.c	25 Mar 2009 20:38:57 -0000	1.22
+++ g_label.c	7 Apr 2009 23:19:17 -0000
@@ -104,7 +104,7 @@
 g_label_orphan(struct g_consumer *cp)
 {
 
-	G_LABEL_DEBUG(0, "Label %s removed.",
+	G_LABEL_DEBUG(1, "Label %s removed.",
 	    LIST_FIRST(&cp->geom->provider)->name);
 	g_slice_orphan(cp);
 }
@@ -113,7 +113,7 @@
 g_label_spoiled(struct g_consumer *cp)
 {
 
-	G_LABEL_DEBUG(0, "Label %s removed.",
+	G_LABEL_DEBUG(1, "Label %s removed.",
 	    LIST_FIRST(&cp->geom->provider)->name);
 	g_slice_spoiled(cp);
 }
@@ -181,7 +181,7 @@
 	g_access(cp, -1, 0, 0);
 	g_slice_config(gp, 0, G_SLICE_CONFIG_SET, (off_t)0, mediasize,
 	    pp->sectorsize, name);
-	G_LABEL_DEBUG(0, "Label for provider %s is %s.", pp->name, name);
+	G_LABEL_DEBUG(1, "Label for provider %s is %s.", pp->name, name);
 	return (gp);
 }
 
@@ -203,7 +203,7 @@
 			return (EBUSY);
 		}
 	} else {
-		G_LABEL_DEBUG(0, "Label %s removed.",
+		G_LABEL_DEBUG(1, "Label %s removed.",
 		    LIST_FIRST(&gp->provider)->name);
 	}
 	g_slice_spoiled(LIST_FIRST(&gp->consumer));
