Index: sys/contrib/pf/net/pf_ioctl.c
===================================================================
RCS file: /root/freebsd/src/repo/src/sys/contrib/pf/net/pf_ioctl.c,v
retrieving revision 1.2
diff -u -r1.2 pf_ioctl.c
--- sys/contrib/pf/net/pf_ioctl.c	11 Jun 2009 21:20:01 -0000	1.2
+++ sys/contrib/pf/net/pf_ioctl.c	11 Jun 2009 21:20:40 -0000
@@ -2497,6 +2497,15 @@
 		break;
 	}
 
+	case DIOCGETNAMEDALTQ: {
+		struct pfioc_ruleset	*pa = (struct pfioc_ruleset *)addr;
+
+		if (pa->name[0])
+			pa->nr = pf_qname2qid(pa->name);
+
+		break;
+	}
+	
 	case DIOCGETALTQS: {
 		struct pfioc_altq	*pa = (struct pfioc_altq *)addr;
 		struct pf_altq		*altq;
@@ -2576,6 +2585,16 @@
 	}
 #endif /* ALTQ */
 
+	case DIOCGETNAMEDTAG: {
+                /* Little abuse. */
+                struct pfioc_ruleset    *pa = (struct pfioc_ruleset *)addr;
+
+                if (pa->name[0])
+                        pa->nr = pf_tagname2tag(pa->name);
+
+                break;
+        }
+
 	case DIOCBEGINADDRS: {
 		struct pfioc_pooladdr	*pp = (struct pfioc_pooladdr *)addr;
 
Index: sys/contrib/pf/net/pfvar.h
===================================================================
RCS file: /root/freebsd/src/repo/src/sys/contrib/pf/net/pfvar.h,v
retrieving revision 1.4
diff -u -r1.4 pfvar.h
--- sys/contrib/pf/net/pfvar.h	11 Jun 2009 21:20:33 -0000	1.4
+++ sys/contrib/pf/net/pfvar.h	11 Jun 2009 21:20:40 -0000
@@ -1605,6 +1605,8 @@
 #define DIOCGIFSPEED	_IOWR('D', 92, struct pf_ifspeed)
 #endif
 #define DIOCKILLIFSTATES  	_IOWR('D', 93, struct pfioc_state_kill)
+#define DIOCGETNAMEDALTQ        _IOR('D', 94, u_int32_t)
+#define DIOCGETNAMEDTAG         _IOR('D', 95, u_int32_t)
 
 #ifdef _KERNEL
 RB_HEAD(pf_src_tree, pf_src_node);
