Index: in_mcast.c
===================================================================
RCS file: /home/ncvs/src/sys/netinet/in_mcast.c,v
retrieving revision 1.27.2.1
diff -u -r1.27.2.1 in_mcast.c
--- in_mcast.c	3 Aug 2009 08:13:06 -0000	1.27.2.1
+++ in_mcast.c	10 Sep 2009 23:21:11 -0000
@@ -2024,7 +2024,10 @@
 			error = ENOMEM;
 			goto out_imo_free;
 		}
-	}
+	}  else if (is_new) {
+		/* Old style ASM filter mode is always exclude */
+		imf_init(imf, MCAST_UNDEFINED, MCAST_EXCLUDE);
+ 	}
 
 	/*
 	 * Begin state merge transaction at IGMP layer.
