Index: ip_carp.c
===================================================================
RCS file: /home/ncvs/src/sys/netinet/ip_carp.c,v
retrieving revision 1.65
diff -u -r1.65 ip_carp.c
--- ip_carp.c	29 Apr 2009 19:19:13 -0000	1.65
+++ ip_carp.c	4 May 2009 18:31:52 -0000
@@ -503,7 +503,8 @@
 			ifpromisc(sc->sc_carpdev, 0);
 			sc->sc_carpdev->if_carp = NULL;
 			CARP_LOCK_DESTROY(cif);
-			free(cif, M_IFADDR);
+			//free(cif, M_IFADDR);
+			free(cif, M_CARP);
 		} else if (unlock)
 			CARP_UNLOCK(cif);
 		sc->sc_carpdev = NULL;
