Index: pf.c
===================================================================
RCS file: /root/freebsd/src/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.6
diff -u -r1.6 pf.c
--- pf.c	11 Jun 2009 21:28:23 -0000	1.6
+++ pf.c	11 Jun 2009 21:29:28 -0000
@@ -7442,7 +7442,9 @@
 		m_freem(*m0);
 		*m0 = NULL;
 		action = PF_PASS;
-	} else if (r->rt)
+	} else if (r->rt &&
+		/* Check to avoid breaking ipfw forward rules needed on CP and proxies listening on loclhost*/
+		!((m->m_flags & M_FASTFWD_OURS) || (pd.pf_mtag->flags & PF_TAG_TRANSLATE_LOCALHOST)))
 		/* pf_route can free the mbuf causing *m0 to become NULL */
 		pf_route(m0, r, dir, ifp, s, &pd);
 
