Index: kern_timeout.c
===================================================================
RCS file: /home/ncvs/src/sys/kern/kern_timeout.c,v
retrieving revision 1.118.2.1
diff -u -r1.118.2.1 kern_timeout.c
--- kern_timeout.c	3 Aug 2009 08:13:06 -0000	1.118.2.1
+++ kern_timeout.c	9 Sep 2009 01:07:31 -0000
@@ -323,7 +323,7 @@
 	steps = 0;
 	cc = (struct callout_cpu *)arg;
 	CC_LOCK(cc);
-	while (cc->cc_softticks != ticks) {
+	while (cc->cc_softticks-1 != ticks) {
 		/*
 		 * cc_softticks may be modified by hard clock, so cache
 		 * it while we work on a given bucket.
