Index: Makefile
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/Makefile,v
retrieving revision 1.42.2.1.2.1
diff -u -r1.42.2.1.2.1 Makefile
--- Makefile	25 Oct 2009 01:10:29 -0000	1.42.2.1.2.1
+++ Makefile	15 Feb 2009 16:16:38 -0000
@@ -1,5 +1,5 @@
 #	@(#)Makefile	8.1 (Berkeley) 6/12/93
-# $FreeBSD: src/usr.bin/netstat/Makefile,v 1.42.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $
+# $FreeBSD: src/usr.bin/netstat/Makefile,v 1.42 2009/02/15 16:16:38 bms Exp $
 
 .include <bsd.own.mk>
 
Index: atalk.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/atalk.c,v
retrieving revision 1.27.2.1.2.1
diff -u -r1.27.2.1.2.1 atalk.c
--- atalk.c	25 Oct 2009 01:10:29 -0000	1.27.2.1.2.1
+++ atalk.c	2 Jan 2008 23:26:11 -0000
@@ -38,7 +38,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/atalk.c,v 1.27.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/atalk.c,v 1.27 2008/01/02 23:26:11 obrien Exp $");
 
 #include <sys/param.h>
 #include <sys/queue.h>
Index: bpf.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/bpf.c,v
retrieving revision 1.12.2.1.2.1
diff -u -r1.12.2.1.2.1 bpf.c
--- bpf.c	25 Oct 2009 01:10:29 -0000	1.12.2.1.2.1
+++ bpf.c	9 Jul 2009 22:13:22 -0000
@@ -25,7 +25,7 @@
  */
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/bpf.c,v 1.12.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/bpf.c,v 1.12 2009/06/19 20:31:44 csjp Exp $");
 
 #include <sys/types.h>
 #include <sys/protosw.h>
Index: if.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/if.c,v
retrieving revision 1.70.2.1.2.1
diff -u -r1.70.2.1.2.1 if.c
--- if.c	25 Oct 2009 01:10:29 -0000	1.70.2.1.2.1
+++ if.c	10 Jan 2010 19:38:34 -0000
@@ -38,7 +38,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/if.c,v 1.70.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/if.c,v 1.70 2008/01/02 23:26:11 obrien Exp $");
 
 #include <sys/types.h>
 #include <sys/protosw.h>
@@ -683,6 +683,8 @@
 	sigsetmask(oldmask);
 	line++;
 	first = 0;
+	if (qflag && (qval-- == 0))
+		exit(0);
 	if (line == 21)
 		goto banner;
 	else
Index: inet.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/inet.c,v
retrieving revision 1.90.2.1.2.1
diff -u -r1.90.2.1.2.1 inet.c
--- inet.c	25 Oct 2009 01:10:29 -0000	1.90.2.1.2.1
+++ inet.c	15 Mar 2009 09:58:31 -0000
@@ -38,7 +38,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/inet.c,v 1.90.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/inet.c,v 1.90 2009/03/15 09:58:31 rwatson Exp $");
 
 #include <sys/param.h>
 #include <sys/queue.h>
Index: inet6.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/inet6.c,v
retrieving revision 1.35.2.1.2.1
diff -u -r1.35.2.1.2.1 inet6.c
--- inet6.c	25 Oct 2009 01:10:29 -0000	1.35.2.1.2.1
+++ inet6.c	29 Apr 2009 09:52:04 -0000
@@ -39,7 +39,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/inet6.c,v 1.35.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/inet6.c,v 1.35 2009/04/29 09:52:04 bms Exp $");
 
 #ifdef INET6
 #include <sys/param.h>
Index: ipsec.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/ipsec.c,v
retrieving revision 1.18.2.1.2.1
diff -u -r1.18.2.1.2.1 ipsec.c
--- ipsec.c	25 Oct 2009 01:10:29 -0000	1.18.2.1.2.1
+++ ipsec.c	2 Jan 2008 23:26:11 -0000
@@ -93,7 +93,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/ipsec.c,v 1.18.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/ipsec.c,v 1.18 2008/01/02 23:26:11 obrien Exp $");
 
 #include <sys/param.h>
 #include <sys/queue.h>
Index: ipx.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/ipx.c,v
retrieving revision 1.28.2.1.2.1
diff -u -r1.28.2.1.2.1 ipx.c
--- ipx.c	25 Oct 2009 01:10:29 -0000	1.28.2.1.2.1
+++ ipx.c	2 Jan 2008 23:26:11 -0000
@@ -39,7 +39,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/ipx.c,v 1.28.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/ipx.c,v 1.28 2008/01/02 23:26:11 obrien Exp $");
 
 #include <sys/param.h>
 #include <sys/queue.h>
Index: main.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/main.c,v
retrieving revision 1.96.2.1.2.1
diff -u -r1.96.2.1.2.1 main.c
--- main.c	25 Oct 2009 01:10:29 -0000	1.96.2.1.2.1
+++ main.c	10 Jan 2010 19:39:20 -0000
@@ -44,7 +44,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/main.c,v 1.96.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/main.c,v 1.96 2009/03/19 01:43:03 bms Exp $");
 
 #include <sys/param.h>
 #include <sys/file.h>
@@ -331,6 +331,8 @@
 int	numeric_addr;	/* show addresses numerically */
 int	numeric_port;	/* show ports numerically */
 static int pflag;	/* show given protocol */
+int	qflag;		/* whether to quit after specified outputs */
+int	qval;		/* how much outputs before we exit */
 int	rflag;		/* show routing tables (or routing stats) */
 int	sflag;		/* show protocol statistics */
 int	tflag;		/* show i/f watchdog timers */
@@ -354,7 +356,7 @@
 
 	af = AF_UNSPEC;
 
-	while ((ch = getopt(argc, argv, "AaBbdf:ghI:iLlM:mN:np:rSstuWw:xz")) != -1)
+	while ((ch = getopt(argc, argv, "AaBbdf:ghI:iLlM:mN:np:q:rSstuWw:xz")) != -1)
 		switch(ch) {
 		case 'A':
 			Aflag = 1;
@@ -440,6 +442,10 @@
 			}
 			pflag = 1;
 			break;
+		case 'q':
+			qval = atoi(optarg);
+			qflag = 1;
+			break;
 		case 'r':
 			rflag = 1;
 			break;
Index: mbuf.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/mbuf.c,v
retrieving revision 1.54.2.1.2.1
diff -u -r1.54.2.1.2.1 mbuf.c
--- mbuf.c	25 Oct 2009 01:10:29 -0000	1.54.2.1.2.1
+++ mbuf.c	2 Jan 2008 23:26:11 -0000
@@ -40,7 +40,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/mbuf.c,v 1.54.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/mbuf.c,v 1.54 2008/01/02 23:26:11 obrien Exp $");
 
 #include <sys/param.h>
 #include <sys/mbuf.h>
Index: mroute.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/mroute.c,v
retrieving revision 1.34.2.1.2.1
diff -u -r1.34.2.1.2.1 mroute.c
--- mroute.c	25 Oct 2009 01:10:29 -0000	1.34.2.1.2.1
+++ mroute.c	21 Apr 2009 12:47:09 -0000
@@ -38,7 +38,7 @@
  */
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/mroute.c,v 1.34.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/mroute.c,v 1.34 2009/04/21 12:47:09 bms Exp $");
 
 /*
  * Print multicast routing structures and statistics.
Index: mroute6.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/mroute6.c,v
retrieving revision 1.22.2.1.2.1
diff -u -r1.22.2.1.2.1 mroute6.c
--- mroute6.c	25 Oct 2009 01:10:29 -0000	1.22.2.1.2.1
+++ mroute6.c	2 Jan 2008 23:26:11 -0000
@@ -66,7 +66,7 @@
  */
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/mroute6.c,v 1.22.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/mroute6.c,v 1.22 2008/01/02 23:26:11 obrien Exp $");
 
 #ifdef INET6
 #include <sys/param.h>
Index: netgraph.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/netgraph.c,v
retrieving revision 1.14.2.1.2.1
diff -u -r1.14.2.1.2.1 netgraph.c
--- netgraph.c	25 Oct 2009 01:10:29 -0000	1.14.2.1.2.1
+++ netgraph.c	2 Jan 2008 23:26:11 -0000
@@ -33,7 +33,7 @@
  */
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/netgraph.c,v 1.14.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/netgraph.c,v 1.14 2008/01/02 23:26:11 obrien Exp $");
 
 #include <sys/param.h>
 #include <sys/queue.h>
Index: netstat.1
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/netstat.1,v
retrieving revision 1.63.2.1.2.1
diff -u -r1.63.2.1.2.1 netstat.1
--- netstat.1	25 Oct 2009 01:10:29 -0000	1.63.2.1.2.1
+++ netstat.1	10 Jan 2010 19:38:34 -0000
@@ -30,7 +30,7 @@
 .\" SUCH DAMAGE.
 .\"
 .\"	@(#)netstat.1	8.8 (Berkeley) 4/18/94
-.\" $FreeBSD: src/usr.bin/netstat/netstat.1,v 1.63.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $
+.\" $FreeBSD: src/usr.bin/netstat/netstat.1,v 1.63 2009/07/09 16:40:00 trasz Exp $
 .\"
 .Dd July 9, 2009
 .Dt NETSTAT 1
@@ -133,6 +133,7 @@
 .Nm
 .Fl w Ar wait
 .Op Fl I Ar interface
+.Op Fl q Ar howmany
 .Op Fl d
 .Op Fl M Ar core
 .Op Fl N Ar system
@@ -146,6 +147,11 @@
 or a single
 .Ar interface .
 If
+.Fl q
+is also present, exit after
+.Ar howmany
+outputs.
+If
 .Fl d
 is also present, show the number of dropped packets.
 .It Xo
Index: netstat.h
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/netstat.h,v
retrieving revision 1.58.2.1.2.1
diff -u -r1.58.2.1.2.1 netstat.h
--- netstat.h	25 Oct 2009 01:10:29 -0000	1.58.2.1.2.1
+++ netstat.h	10 Jan 2010 19:38:34 -0000
@@ -31,7 +31,7 @@
  * SUCH DAMAGE.
  *
  *	@(#)netstat.h	8.2 (Berkeley) 1/4/94
- * $FreeBSD: src/usr.bin/netstat/netstat.h,v 1.58.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $
+ * $FreeBSD: src/usr.bin/netstat/netstat.h,v 1.58 2009/03/19 01:43:03 bms Exp $
  */
 
 #include <sys/cdefs.h>
@@ -47,6 +47,8 @@
 extern int	mflag;	/* show memory stats */
 extern int	numeric_addr;	/* show addresses numerically */
 extern int	numeric_port;	/* show ports numerically */
+extern int	qflag;	/* whether to quit after specified outputs */
+extern int	qval;	/* how much outputs before we exit */
 extern int	rflag;	/* show routing tables (or routing stats) */
 extern int	sflag;	/* show protocol statistics */
 extern int	tflag;	/* show i/f watchdog timers */
Index: pfkey.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/pfkey.c,v
retrieving revision 1.6.2.1.2.1
diff -u -r1.6.2.1.2.1 pfkey.c
--- pfkey.c	25 Oct 2009 01:10:29 -0000	1.6.2.1.2.1
+++ pfkey.c	2 Jan 2008 23:26:11 -0000
@@ -68,7 +68,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/pfkey.c,v 1.6.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/pfkey.c,v 1.6 2008/01/02 23:26:11 obrien Exp $");
 
 #include <sys/param.h>
 #include <sys/queue.h>
Index: route.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/route.c,v
retrieving revision 1.94.2.2.2.1
diff -u -r1.94.2.2.2.1 route.c
--- route.c	25 Oct 2009 01:10:29 -0000	1.94.2.2.2.1
+++ route.c	1 Jun 2009 15:49:42 -0000
@@ -38,7 +38,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/route.c,v 1.94.2.2.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/route.c,v 1.94 2009/06/01 15:49:42 bz Exp $");
 
 #include <sys/param.h>
 #include <sys/protosw.h>
@@ -1008,11 +1008,11 @@
 #define	p(f, m) if (rtstat.f || sflag <= 1) \
 	printf(m, rtstat.f, plural(rtstat.f))
 
-	p(rts_badredirect, "\t%hu bad routing redirect%s\n");
-	p(rts_dynamic, "\t%hu dynamically created route%s\n");
-	p(rts_newgateway, "\t%hu new gateway%s due to redirects\n");
-	p(rts_unreach, "\t%hu destination%s found unreachable\n");
-	p(rts_wildcard, "\t%hu use%s of a wildcard route\n");
+	p(rts_badredirect, "\t%u bad routing redirect%s\n");
+	p(rts_dynamic, "\t%u dynamically created route%s\n");
+	p(rts_newgateway, "\t%u new gateway%s due to redirects\n");
+	p(rts_unreach, "\t%u destination%s found unreachable\n");
+	p(rts_wildcard, "\t%u use%s of a wildcard route\n");
 #undef p
 
 	if (rttrash || sflag <= 1)
Index: sctp.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/sctp.c,v
retrieving revision 1.10.2.1.2.1
diff -u -r1.10.2.1.2.1 sctp.c
--- sctp.c	25 Oct 2009 01:10:29 -0000	1.10.2.1.2.1
+++ sctp.c	16 Apr 2008 17:40:30 -0000
@@ -35,7 +35,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/sctp.c,v 1.10.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/sctp.c,v 1.10 2008/04/16 17:40:30 rrs Exp $");
 
 #include <sys/param.h>
 #include <sys/queue.h>
Index: unix.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/netstat/unix.c,v
retrieving revision 1.22.2.1.2.1
diff -u -r1.22.2.1.2.1 unix.c
--- unix.c	25 Oct 2009 01:10:29 -0000	1.22.2.1.2.1
+++ unix.c	31 Dec 2008 08:56:49 -0000
@@ -38,7 +38,7 @@
 #endif
 
 #include <sys/cdefs.h>
-__FBSDID("$FreeBSD: src/usr.bin/netstat/unix.c,v 1.22.2.1.2.1 2009/10/25 01:10:29 kensmith Exp $");
+__FBSDID("$FreeBSD: src/usr.bin/netstat/unix.c,v 1.22 2008/12/31 08:56:49 maxim Exp $");
 
 /*
  * Display protocol blocks in the unix domain.
