Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /root/repo/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.4
diff -u -r1.4 pf.c
--- sys/contrib/pf/net/pf.c	8 Jun 2009 21:34:11 -0000	1.4
+++ sys/contrib/pf/net/pf.c	8 Jun 2009 21:47:18 -0000
@@ -6206,6 +6206,9 @@
 
 	ip = mtod(m0, struct ip *);
 
+	if (IN_MULTICAST(ip->ip_dst.s_addr))
+                return;
+
 	ro = &iproute;
 	bzero((caddr_t)ro, sizeof(*ro));
 	dst = satosin(&ro->ro_dst);
@@ -6247,6 +6250,12 @@
 	}
 	if (ifp == NULL)
 		goto bad;
+	else if (r->rt == PF_REPLYTO && in_localaddr(ip->ip_dst))
+		goto donelocal;
+	else if (r->rt == PF_ROUTETO && r->direction == dir && in_localip(ip->ip_dst))
+		goto donelocal;
+	else if (in_broadcast(ip->ip_dst, ifp))
+		goto donelocal;
 
 	if (oifp != ifp) {
 #ifdef __FreeBSD__
@@ -6428,6 +6437,7 @@
 done:
 	if (r->rt != PF_DUPTO)
 		*m = NULL;
+donelocal:
 	if (ro == &iproute && ro->ro_rt)
 		RTFREE(ro->ro_rt);
 	return;
