Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /home/ermal/replytorepo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.2
diff -u -r1.2 pf.c
--- sys/contrib/pf/net/pf.c	27 Apr 2009 18:08:00 -0000	1.2
+++ sys/contrib/pf/net/pf.c	27 Apr 2009 18:13:47 -0000
@@ -6206,6 +6206,8 @@
 	}
 
 	ip = mtod(m0, struct ip *);
+	if (in_localaddr(ip->ip_dst))
+		return;
 
 	ro = &iproute;
 	bzero((caddr_t)ro, sizeof(*ro));
