Index: dev/e1000/if_em.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/e1000/if_em.c,v
retrieving revision 1.19
diff -u -r1.19 if_em.c
--- dev/e1000/if_em.c	29 Jun 2009 18:17:10 -0000	1.19
+++ dev/e1000/if_em.c	1 Jul 2009 01:17:02 -0000
@@ -1051,7 +1051,11 @@
 			** listener and set the watchdog on.
 			*/
 			ETHER_BPF_MTAP(ifp, m);
+<<<<<<< if_em.c
+			addapter->watchdog_timer = EM_TX_TIMEOUT;
+=======
 			adapter->watchdog_timer = EM_TX_TIMEOUT;
+>>>>>>> 1.14
 		}
 	} else if ((error = drbr_enqueue(ifp, adapter->br, m)) != 0)
 		return (error);
@@ -1140,8 +1144,21 @@
 		if (em_xmit(adapter, &m_head)) {
 			if (m_head == NULL)
 				break;
+<<<<<<< if_em.c
+<<<<<<< if_em.c
 			ifp->if_drv_flags |= IFF_DRV_OACTIVE;
 			IFQ_DRV_PREPEND(&ifp->if_snd, m_head);
+=======
+#ifndef IFNET_BUF_RING
+=======
+>>>>>>> 1.19
+			ifp->if_drv_flags |= IFF_DRV_OACTIVE;
+			IFQ_DRV_PREPEND(&ifp->if_snd, m_head);
+<<<<<<< if_em.c
+#endif
+>>>>>>> 1.14
+=======
+>>>>>>> 1.19
 			break;
 		}
 
Index: netinet/ip_carp.c
===================================================================
RCS file: /home/ncvs/src/sys/netinet/ip_carp.c,v
retrieving revision 1.70
diff -u -r1.70 ip_carp.c
--- netinet/ip_carp.c	25 Jun 2009 16:35:28 -0000	1.70
+++ netinet/ip_carp.c	1 Jul 2009 01:17:36 -0000
@@ -503,7 +503,8 @@
 			ifpromisc(sc->sc_carpdev, 0);
 			sc->sc_carpdev->if_carp = NULL;
 			CARP_LOCK_DESTROY(cif);
-			free(cif, M_IFADDR);
+			//free(cif, M_IFADDR);
+			free(cif, M_CARP);
 		} else if (unlock)
 			CARP_UNLOCK(cif);
 		sc->sc_carpdev = NULL;
