Index: sys/kern/vfs_mount.c
===================================================================
--- sys/kern/vfs_mount.c	(revision 198966)
+++ sys/kern/vfs_mount.c	(working copy)
@@ -1644,7 +1644,7 @@ void
 vfs_mountroot(void)
 {
 	char *cp, *options;
-	int error, i, asked = 0;
+	int error, i, asked = 0, tries = 10;
 
 	options = NULL;
 
@@ -1693,6 +1693,7 @@ vfs_mountroot(void)
 	 * supplied via some other means.  This is the preferred
 	 * mechanism.
 	 */
+again:
 	cp = getenv("vfs.root.mountfrom");
 	if (cp != NULL) {
 		error = vfs_mountroot_try(cp, options);
@@ -1715,6 +1716,13 @@ vfs_mountroot(void)
 	if (ctrootdevname != NULL)
 		if (!vfs_mountroot_try(ctrootdevname, options))
 			goto mounted;
+
+	/* Hack to let usb probe */
+	if (tries--) {
+		pause("mountroot", hz);
+		goto again;
+	}
+
 	/*
 	 * Everything so far has failed, prompt on the console if we haven't
 	 * already tried that.
