Index: sys/dev/bce/if_bcereg.h
===================================================================
RCS file: /home/ncvs/src/sys/dev/bce/if_bcereg.h,v
retrieving revision 1.26.2.5.2.1
diff -u -r1.26.2.5.2.1 if_bcereg.h
--- sys/dev/bce/if_bcereg.h	14 Jun 2010 02:09:06 -0000	1.26.2.5.2.1
+++ sys/dev/bce/if_bcereg.h	9 Dec 2010 13:15:59 -0000
@@ -6170,7 +6170,7 @@
  * Page count must remain a power of 2 for all
  * of the math to work correctly.
  */
-#define RX_PAGES	2
+#define RX_PAGES	32
 #define TOTAL_RX_BD_PER_PAGE  (BCM_PAGE_SIZE / sizeof(struct rx_bd))
 #define USABLE_RX_BD_PER_PAGE (TOTAL_RX_BD_PER_PAGE - 1)
 #define TOTAL_RX_BD (TOTAL_RX_BD_PER_PAGE * RX_PAGES)
