Index: sys/netinet/ip_carp.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/netinet/ip_carp.c,v
retrieving revision 1.4
diff -u -r1.4 ip_carp.c
--- sys/netinet/ip_carp.c	27 Sep 2011 19:19:07 -0000	1.4
+++ sys/netinet/ip_carp.c	27 Sep 2011 19:21:18 -0000
@@ -940,6 +940,9 @@
 		tv.tv_usec = advskew * 1000000 / 256;
 	}
 
+	if (!carp_opts[CARPCTL_ALLOW])
+                goto resettimer;
+
 	ch.carp_version = CARP_VERSION;
 	ch.carp_type = CARP_ADVERTISEMENT;
 	ch.carp_vhid = sc->sc_vhid;
@@ -1113,6 +1116,7 @@
 	}
 #endif /* INET6 */
 
+resettimer:
 	if (advbase != 255 || advskew != 255)
 		callout_reset(&sc->sc_ad_tmo, tvtohz(&tv),
 		    carp_send_ad, sc);
