Index: sys/netinet/ip_carp.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/netinet/ip_carp.c,v
retrieving revision 1.4
diff -u -r1.4 ip_carp.c
--- sys/netinet/ip_carp.c	21 Jun 2011 21:32:07 -0000	1.4
+++ sys/netinet/ip_carp.c	21 Jun 2011 21:33:27 -0000
@@ -891,6 +891,9 @@
 {
 	struct carp_softc *sc;
 
+	if (!carp_opts[CARPCTL_ALLOW])
+                return;
+
 	mtx_lock(&carp_mtx);
 	LIST_FOREACH(sc, &carpif_list, sc_next) {
 		if (sc->sc_carpdev == NULL)
@@ -924,6 +927,9 @@
 	struct mbuf *m;
 	int len, advbase, advskew;
 
+	if (!carp_opts[CARPCTL_ALLOW])
+                return;
+
 	if (sc->sc_carpdev)
 		CARP_SCLOCK_ASSERT(sc);
 
