? hostapd.diff
Index: usr.sbin/wpa/Makefile.inc
===================================================================
RCS file: /home/ncvs/src/usr.sbin/wpa/Makefile.inc,v
retrieving revision 1.4.4.1.2.1
diff -u -r1.4.4.1.2.1 Makefile.inc
--- usr.sbin/wpa/Makefile.inc	11 Nov 2011 04:20:22 -0000	1.4.4.1.2.1
+++ usr.sbin/wpa/Makefile.inc	6 Jan 2012 17:59:53 -0000
@@ -10,6 +10,7 @@
 	${WPA_DISTDIR}/src/common \
 	${WPA_DISTDIR}/src/crypto \
 	${WPA_DISTDIR}/src/eap_common \
+	${WPA_DISTDIR}/src/eap_peer \
 	${WPA_DISTDIR}/src/eapol_supp \
 	${WPA_DISTDIR}/src/l2_packet \
 	${WPA_DISTDIR}/src/utils
Index: usr.sbin/wpa/hostapd/Makefile
===================================================================
RCS file: /home/ncvs/src/usr.sbin/wpa/hostapd/Makefile,v
retrieving revision 1.9.2.1.2.1
diff -u -r1.9.2.1.2.1 Makefile
--- usr.sbin/wpa/hostapd/Makefile	11 Nov 2011 04:20:22 -0000	1.9.2.1.2.1
+++ usr.sbin/wpa/hostapd/Makefile	6 Jan 2012 17:59:54 -0000
@@ -6,6 +6,7 @@
 	${WPA_DISTDIR}/src/ap \
 	${WPA_DISTDIR}/src/eap_server \
 	${WPA_DISTDIR}/src/eap_common \
+	${WPA_DISTDIR}/src/eap_peer \
 	${WPA_DISTDIR}/src/eapol_auth \
 	${WPA_DISTDIR}/src/drivers \
 	${WPA_DISTDIR}/src/radius \
