Index: sys/net/if_bridge.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/net/if_bridge.c,v
retrieving revision 1.3
diff -u -r1.3 if_bridge.c
--- sys/net/if_bridge.c	27 Sep 2011 19:20:58 -0000	1.3
+++ sys/net/if_bridge.c	27 Sep 2011 19:40:50 -0000
@@ -848,6 +848,8 @@
 		}
 		BRIDGE_LOCK(sc);
 		LIST_FOREACH(bif, &sc->sc_iflist, bif_next) {
+			if (bif->bif_ifp->if_type == IFT_GIF)
+				continue;
 			if (bif->bif_ifp->if_mtu != ifr->ifr_mtu) {
 				log(LOG_NOTICE, "%s: invalid MTU: %lu(%s)"
 				    " != %d\n", sc->sc_ifp->if_xname,
@@ -1099,6 +1101,7 @@
 		goto out;
 	}
 
+	if (ifs->if_type != IFT_GIF) {
 	/* Allow the first Ethernet member to define the MTU */
 	if (LIST_EMPTY(&sc->sc_iflist))
 		sc->sc_ifp->if_mtu = ifs->if_mtu;
@@ -1108,6 +1111,7 @@
 		error = EINVAL;
 		goto out;
 	}
+	}
 
 	/*
 	 * Assign the interface's MAC address to the bridge if it's the first
