Index: net/if_ethersubr.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/net/if_ethersubr.c,v
retrieving revision 1.2
diff -u -r1.2 if_ethersubr.c
--- net/if_ethersubr.c	27 Sep 2011 14:19:30 -0000	1.2
+++ net/if_ethersubr.c	27 Sep 2011 14:37:22 -0000
@@ -1016,8 +1016,6 @@
 	for (i = 0; i < ifp->if_addrlen; i++)
 		if (lla[i] != 0)
 			break; 
-	if (i != ifp->if_addrlen)
-		if_printf(ifp, "Ethernet address: %6D\n", lla, ":");
 }
 
 /*
