Index: if.h
===================================================================
RCS file: /root/freebsd/repo/src/sys/net/if.h,v
retrieving revision 1.2
diff -u -r1.2 if.h
--- if.h	27 Sep 2011 14:19:30 -0000	1.2
+++ if.h	27 Sep 2011 19:52:07 -0000
@@ -237,7 +237,7 @@
 
 #define	IFCAP_CANTCHANGE	(IFCAP_NETMAP)
 
-#define	IFQ_MAXLEN	50
+#define	IFQ_MAXLEN	128
 #define	IFNET_SLOWHZ	1		/* granularity is 1 second */
 
 /*
