Index: sys/netinet/ip_input.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/netinet/ip_input.c,v
retrieving revision 1.4
diff -u -r1.4 ip_input.c
--- sys/netinet/ip_input.c	27 Sep 2011 19:59:02 -0000	1.4
+++ sys/netinet/ip_input.c	27 Sep 2011 20:12:05 -0000
@@ -529,6 +529,7 @@
 		m->m_flags &= ~M_FASTFWD_OURS;
 		goto ours;
 	}
+#if 0
 	if (m_tag_find(m, PACKET_TAG_IPFORWARD, NULL) != NULL) {
 		/*
 		 * Directly ship the packet on.  This allows forwarding
@@ -538,6 +539,7 @@
 		ip_forward(m, 1);
 		return;
 	}
+#endif
 #endif /* IPFIREWALL_FORWARD */
 
 passin:
Index: sys/netinet/ip_input.c.orig
===================================================================
RCS file: /root/freebsd/repo/src/sys/netinet/ip_input.c.orig,v
retrieving revision 1.1
diff -u -r1.1 ip_input.c.orig
--- sys/netinet/ip_input.c.orig	27 Sep 2011 20:10:54 -0000	1.1
+++ sys/netinet/ip_input.c.orig	27 Sep 2011 19:57:17 -0000
@@ -529,13 +529,13 @@
 		m->m_flags &= ~M_FASTFWD_OURS;
 		goto ours;
 	}
-	if ((dchg = (m_tag_find(m, PACKET_TAG_IPFORWARD, NULL) != NULL)) != 0) {
+	if (m_tag_find(m, PACKET_TAG_IPFORWARD, NULL) != NULL) {
 		/*
 		 * Directly ship the packet on.  This allows forwarding
 		 * packets originally destined to us to some other directly
 		 * connected host.
 		 */
-		ip_forward(m, dchg);
+		ip_forward(m, 1);
 		return;
 	}
 #endif /* IPFIREWALL_FORWARD */
Index: sys/netinet/ip_output.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/netinet/ip_output.c,v
retrieving revision 1.3
diff -u -r1.3 ip_output.c
--- sys/netinet/ip_output.c	27 Sep 2011 17:20:03 -0000	1.3
+++ sys/netinet/ip_output.c	27 Sep 2011 20:12:05 -0000
@@ -563,6 +563,7 @@
 		error = netisr_queue(NETISR_IP, m);
 		goto done;
 	}
+#if 0
 	/* Or forward to some other address? */
 	fwd_tag = m_tag_find(m, PACKET_TAG_IPFORWARD, NULL);
 	if (fwd_tag) {
@@ -572,6 +573,7 @@
 		m_tag_delete(m, fwd_tag);
 		goto again;
 	}
+#endif
 #endif /* IPFIREWALL_FORWARD */
 
 passout:
Index: sys/netinet/ip_output.c.orig
===================================================================
RCS file: /root/freebsd/repo/src/sys/netinet/ip_output.c.orig,v
retrieving revision 1.1
diff -u -r1.1 ip_output.c.orig
--- sys/netinet/ip_output.c.orig	27 Sep 2011 20:10:54 -0000	1.1
+++ sys/netinet/ip_output.c.orig	27 Sep 2011 17:19:56 -0000
@@ -476,18 +476,20 @@
 
 sendit:
 #ifdef IPSEC
-	switch(ip_ipsec_output(&m, inp, &flags, &error)) {
-	case 1:
-		goto bad;
-	case -1:
-		goto done;
-	case 0:
-	default:
-		break;	/* Continue with packet processing. */
+	if (V_ipipsec_in_use) {
+		switch(ip_ipsec_output(&m, inp, &flags, &error)) {
+		case 1:
+			goto bad;
+		case -1:
+			goto done;
+		case 0:
+		default:
+			break;	/* Continue with packet processing. */
+		}
+		/* Update variables that are affected by ipsec4_output(). */
+		ip = mtod(m, struct ip *);
+		hlen = ip->ip_hl << 2;
 	}
-	/* Update variables that are affected by ipsec4_output(). */
-	ip = mtod(m, struct ip *);
-	hlen = ip->ip_hl << 2;
 #endif /* IPSEC */
 
 	/* Jump over all PFIL processing if hooks are not active. */
@@ -1084,7 +1086,7 @@
 		{
 			caddr_t req;
 			struct mbuf *m;
-
+			
 			if ((error = soopt_getm(sopt, &m)) != 0) /* XXX */
 				break;
 			if ((error = soopt_mcopyin(sopt, m)) != 0) /* XXX */
@@ -1210,7 +1212,7 @@
 			struct mbuf *m = NULL;
 			caddr_t req = NULL;
 			size_t len = 0;
-
+			
 			if (m != 0) {
 				req = mtod(m, caddr_t);
 				len = m->m_len;
