Index: sys/netinet/ip_input.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/netinet/ip_input.c,v
retrieving revision 1.3
diff -u -r1.3 ip_input.c
--- sys/netinet/ip_input.c	27 Sep 2011 17:20:03 -0000	1.3
+++ sys/netinet/ip_input.c	27 Sep 2011 19:57:17 -0000
@@ -529,13 +529,13 @@
 		m->m_flags &= ~M_FASTFWD_OURS;
 		goto ours;
 	}
-	if ((dchg = (m_tag_find(m, PACKET_TAG_IPFORWARD, NULL) != NULL)) != 0) {
+	if (m_tag_find(m, PACKET_TAG_IPFORWARD, NULL) != NULL) {
 		/*
 		 * Directly ship the packet on.  This allows forwarding
 		 * packets originally destined to us to some other directly
 		 * connected host.
 		 */
-		ip_forward(m, dchg);
+		ip_forward(m, 1);
 		return;
 	}
 #endif /* IPFIREWALL_FORWARD */
