Index: pf.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.6
diff -u -r1.6 pf.c
--- pf.c	27 Sep 2011 07:33:13 -0000	1.6
+++ pf.c	27 Sep 2011 14:01:55 -0000
@@ -7429,7 +7429,9 @@
 		break;
 	default:
 		/* pf_route can free the mbuf causing *m0 to become NULL */
-		if (r->rt)
+		if (r->rt &&
+		    /* Check to avoid breaking ipfw forward rules needed on CP and proxies listening on loclhost*/
+		    !((m->m_flags & M_FASTFWD_OURS) || (pd.pf_mtag->flags & PF_TAG_TRANSLATE_LOCALHOST)))
 			pf_route(m0, r, dir, kif->pfik_ifp, s, &pd);
 		break;
 	}
@@ -7992,7 +7994,9 @@
 		break;
 	default:
 		/* pf_route6 can free the mbuf causing *m0 to become NULL */
-		if (r->rt)
+		if (r->rt &&
+		    /* Check to avoid breaking ipfw forward rules needed on CP and proxies listening on loclhost*/
+		    !((m->m_flags & M_FASTFWD_OURS) || (pd.pf_mtag->flags & PF_TAG_TRANSLATE_LOCALHOST)))
 			pf_route6(m0, r, dir, kif->pfik_ifp, s, &pd);
 		break;
 	}
