Index: Makefile
===================================================================
RCS file: /root/freebsd/repo/src/sys/modules/ipfw/Makefile,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 Makefile
--- Makefile	21 Jun 2011 19:47:46 -0000	1.1.1.1
+++ Makefile	21 Jun 2011 20:38:37 -0000
@@ -18,7 +18,8 @@
 #CFLAGS+= -DIPFIREWALL_VERBOSE_LIMIT=100
 #
 #If you want it to pass all packets by default
-#CFLAGS+= -DIPFIREWALL_DEFAULT_TO_ACCEPT
+CFLAGS+= -DIPFIREWALL_DEFAULT_TO_ACCEPT
+CFLAGS+= -DIPFIREWALL_FORWARD
 #
 
 .if !defined(KERNBUILDDIR)
