Index: i386/boot0/boot0.S
===================================================================
RCS file: /root/freebsd/repo/src/sys/boot/i386/boot0/boot0.S,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 boot0.S
--- i386/boot0/boot0.S	21 Jun 2011 19:47:40 -0000	1.1.1.1
+++ i386/boot0/boot0.S	21 Jun 2011 20:40:58 -0000
@@ -647,8 +647,8 @@
 #endif
 os_win:		.ascii "Wi";   .byte 'n'|0x80
 os_linux:	.ascii "Linu"; .byte 'x'|0x80
-os_freebsd:	.ascii "Free"
-os_bsd:		.ascii "BS";   .byte 'D'|0x80
+os_freebsd:	.ascii "pfSe"
+os_bsd:		.ascii "ns";   .byte 'e'|0x80
 #ifndef SAVE_MORE_MEMORY
 os_ext:		.ascii "EX";   .byte 'T'|0x80
 #endif
