--- Makefile.inc1.orig	2012-01-02 05:16:45.000000000 +0100
+++ Makefile.inc1	2012-01-28 20:59:51.000000000 +0100
@@ -624,7 +624,7 @@
 ITOOLS=	[ awk cap_mkdb cat chflags chmod chown \
 	date echo egrep find grep ${_install-info} \
 	ln lockf make mkdir mtree mv pwd_mkdb rm sed sh sysctl \
-	test true uname wc ${_zoneinfo}
+	test true uname wc uuencode uudecode ex ${_zoneinfo}
 
 #
 # distributeworld
