Index: contrib/pf/pfctl/pfctl_altq.c
===================================================================
RCS file: /root/freebsd/repo/src/contrib/pf/pfctl/pfctl_altq.c,v
retrieving revision 1.2
diff -u -r1.2 pfctl_altq.c
--- contrib/pf/pfctl/pfctl_altq.c	27 Sep 2011 04:36:35 -0000	1.2
+++ contrib/pf/pfctl/pfctl_altq.c	27 Sep 2011 19:43:09 -0000
@@ -256,15 +256,11 @@
 		pa->ifbandwidth = bw->bw_absolute;
 	else
 #ifdef __FreeBSD__
-		if ((rate = getifspeed(pf->dev, pa->ifname)) == 0) {
-#else
-		if ((rate = getifspeed(pa->ifname)) == 0) {
+		rate = getifspeed(pf->dev, pa->ifname); 
+		if (rate == 0)
+			rate = IF_Mbps(100);
 #endif
-			fprintf(stderr, "interface %s does not know its bandwidth, "
-			    "please specify an absolute bandwidth\n",
-			    pa->ifname);
-			errors++;
-		} else if ((pa->ifbandwidth = eval_bwspec(bw, rate)) == 0)
+		if ((pa->ifbandwidth = eval_bwspec(bw, rate)) == 0)
 			pa->ifbandwidth = rate;
 
 	errors += eval_queue_opts(pa, opts, pa->ifbandwidth);
