Index: contrib/pf/pfctl/pf_print_state.c
===================================================================
RCS file: /root/freebsd/repo/src/contrib/pf/pfctl/pf_print_state.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 pf_print_state.c
--- contrib/pf/pfctl/pf_print_state.c	17 Jan 2012 16:27:02 -0000	1.1.1.1
+++ contrib/pf/pfctl/pf_print_state.c	17 Jan 2012 16:37:25 -0000
@@ -337,6 +337,10 @@
 			printf(", anchor %u", ntohl(s->anchor));
 		if (ntohl(s->rule) != -1)
 			printf(", rule %u", ntohl(s->rule));
+		if (s->pqid)
+			printf(", priqueue %u", s->pqid);
+		if (s->qid)
+			printf(", queue %u", s->qid);
 		if (s->state_flags & PFSTATE_SLOPPY)
 			printf(", sloppy");
 		if (s->state_flags & PFSTATE_PFLOW)
Index: sys/contrib/pf/net/if_pfsync.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/contrib/pf/net/if_pfsync.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 if_pfsync.c
--- sys/contrib/pf/net/if_pfsync.c	17 Jan 2012 16:27:02 -0000	1.1.1.1
+++ sys/contrib/pf/net/if_pfsync.c	17 Jan 2012 16:40:24 -0000
@@ -673,6 +673,8 @@
 	sp->log = st->log;
 	sp->timeout = st->timeout;
 	sp->state_flags = st->state_flags;
+	sp->qid = st->qid;
+	sp->pqid = st->pqid;
 	if (st->src_node)
 		sp->sync_flags |= PFSYNC_FLAG_SRCNODE;
 	if (st->nat_src_node)
Index: sys/contrib/pf/net/pfvar.h
===================================================================
RCS file: /root/freebsd/repo/src/sys/contrib/pf/net/pfvar.h,v
retrieving revision 1.4
diff -u -r1.4 pfvar.h
--- sys/contrib/pf/net/pfvar.h	17 Jan 2012 16:37:08 -0000	1.4
+++ sys/contrib/pf/net/pfvar.h	17 Jan 2012 16:39:23 -0000
@@ -971,6 +971,8 @@
 	u_int32_t	 bytes[2][2];
 	u_int32_t	 creatorid;
 	sa_family_t	 af;
+	u_int16_t	 qid;
+	u_int16_t	 pqid;
 	u_int8_t	 proto;
 	u_int8_t	 direction;
 #ifdef __FreeBSD__
