Index: contrib/pf/pfctl/pf_print_state.c
===================================================================
RCS file: /root/freebsd/repo/src/contrib/pf/pfctl/pf_print_state.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 pf_print_state.c
--- contrib/pf/pfctl/pf_print_state.c	21 Jun 2011 19:47:47 -0000	1.1.1.1
+++ contrib/pf/pfctl/pf_print_state.c	21 Jun 2011 21:42:00 -0000
@@ -294,6 +294,10 @@
 			printf(", anchor %u", s->anchor.nr);
 		if (s->rule.nr != -1)
 			printf(", rule %u", s->rule.nr);
+		if (s->pqid)
+			printf(", priqueue %u", s->pqid);
+		if (s->qid)
+			printf(", queue %u", s->qid);
 		if (s->state_flags & PFSTATE_SLOPPY)
 			printf(", sloppy");
 		if (s->src_node != NULL)
