Index: contrib/pf/pfctl/pf_print_state.c
===================================================================
RCS file: /root/freebsd/repo/src/contrib/pf/pfctl/pf_print_state.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 pf_print_state.c
--- contrib/pf/pfctl/pf_print_state.c	27 Sep 2011 04:25:20 -0000	1.1.1.1
+++ contrib/pf/pfctl/pf_print_state.c	27 Sep 2011 19:48:22 -0000
@@ -337,6 +337,10 @@
 			printf(", anchor %u", ntohl(s->anchor));
 		if (ntohl(s->rule) != -1)
 			printf(", rule %u", ntohl(s->rule));
+		if (s->pqid)
+			printf(", priqueue %u", s->pqid);
+		if (s->qid)
+			printf(", queue %u", s->qid);
 		if (s->state_flags & PFSTATE_SLOPPY)
 			printf(", sloppy");
 		if (s->state_flags & PFSTATE_PFLOW)
