Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.7
diff -u -r1.7 pf.c
--- sys/contrib/pf/net/pf.c	27 Sep 2011 14:03:59 -0000	1.7
+++ sys/contrib/pf/net/pf.c	27 Sep 2011 14:53:16 -0000
@@ -6236,6 +6236,7 @@
 	m0->m_pkthdr.csum_flags &= ifp->if_hwassist;
 
 	if (ntohs(ip->ip_len) <= ifp->if_mtu ||
+	    (m0->m_pkthdr.csum_flags & CSUM_TSO) ||
 	    (ifp->if_hwassist & CSUM_FRAGMENT &&
 	    ((ip->ip_off & htons(IP_DF)) == 0))) {
 		/*
@@ -6252,6 +6253,7 @@
 				ip->ip_sum = in_cksum(m0, ip->ip_hl << 2);
 			}
 		}
+		m0->m_flags &= ~(M_PROTOFLAGS);
 		PF_UNLOCK();
 		error = (*ifp->if_output)(ifp, m0, sintosa(dst), ro);
 		PF_LOCK();
@@ -6313,7 +6315,7 @@
 	 * Too large for interface; fragment if possible.
 	 * Must be able to put at least 8 bytes per fragment.
 	 */
-	if (ip->ip_off & htons(IP_DF)) {
+	if (ip->ip_off & htons(IP_DF) || (m0->m_pkthdr.csum_flags & CSUM_TSO)) {
 #ifdef __FreeBSD__
 		KMOD_IPSTAT_INC(ips_cantfrag);
 #else
@@ -6360,6 +6362,7 @@
 		m0->m_nextpkt = 0;
 #ifdef __FreeBSD__
 		if (error == 0) {
+			m0->m_flags &= ~(M_PROTOFLAGS);
 			PF_UNLOCK();
 			error = (*ifp->if_output)(ifp, m0, sintosa(dst),
 			    NULL);
