Index: sys/dev/usb/net/if_udav.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/usb/net/if_udav.c,v
retrieving revision 1.18
diff -u -r1.18 if_udav.c
--- sys/dev/usb/net/if_udav.c	19 Jun 2011 08:34:10 -0000	1.18
+++ sys/dev/usb/net/if_udav.c	22 Jun 2011 17:43:11 -0000
@@ -211,6 +211,7 @@
 	/* Kontron AG USB Ethernet */
 	{USB_VPI(USB_VENDOR_KONTRON, USB_PRODUCT_KONTRON_DM9601, 0)},
 	{USB_VPI(USB_VENDOR_KONTRON, USB_PRODUCT_KONTRON_JP1082, 0)},
+	{USB_VPI(0x0fe6, 0x8101, 0)},
 };
 
 static void
