Index: usr.sbin/wpa/hostapd/Makefile
===================================================================
RCS file: /root/freebsd/repo/src/usr.sbin/wpa/hostapd/Makefile,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 Makefile
--- usr.sbin/wpa/hostapd/Makefile	22 Jun 2011 15:43:34 -0000	1.1.1.1
+++ usr.sbin/wpa/hostapd/Makefile	22 Jun 2011 17:18:41 -0000
@@ -40,7 +40,7 @@
 
 CFLAGS+= -I${HOSTAPD_DISTDIR} -I${WPA_DISTDIR}/src/drivers
 
-CFLAGS+= -DCONFIG_DRIVER_BSD -DHOSTAPD
+CFLAGS+= -DCONFIG_DRIVER_BSD -DHOSTAPD -DCONFIG_RSN_PREAUTH
 CFLAGS+= -DCONFIG_DRIVER_RADIUS_ACL
 .if ${MK_INET6} != "no"
 CFLAGS+= -DCONFIG_IPV6
