diff --git a/sys/netinet/ip_carp.c b/sys/netinet/ip_carp.c
index 5b5fb19..d39b55a 100644
--- a/sys/netinet/ip_carp.c
+++ b/sys/netinet/ip_carp.c
@@ -1466,7 +1466,7 @@ carp_set_addr(struct carp_softc *sc, struct sockaddr_in *sin)
 	struct ip_moptions *imo = &sc->sc_imo;
 	struct in_addr addr;
 	u_long iaddr = htonl(sin->sin_addr.s_addr);
-	int own, error;
+	int error;
 
 	if (sin->sin_addr.s_addr == 0) {
 		if (sc->sc_carpdev)
@@ -1482,7 +1482,7 @@ carp_set_addr(struct carp_softc *sc, struct sockaddr_in *sin)
 	}
 
 	/* we have to do it by hands to check we won't match on us */
-	ia_if = NULL; own = 0;
+	ia_if = NULL;
 	IN_IFADDR_RLOCK();
 	TAILQ_FOREACH(ia, &V_in_ifaddrhead, ia_link) {
 		/* and, yeah, we need a multicast-capable iface too */
@@ -1491,9 +1491,6 @@ carp_set_addr(struct carp_softc *sc, struct sockaddr_in *sin)
 		    (iaddr & ia->ia_subnetmask) == ia->ia_subnet) {
 			if (!ia_if)
 				ia_if = ia;
-			if (sin->sin_addr.s_addr ==
-			    ia->ia_addr.sin_addr.s_addr)
-				own++;
 		}
 	}
 
@@ -1589,9 +1586,8 @@ carp_set_addr(struct carp_softc *sc, struct sockaddr_in *sin)
 
 	sc->sc_naddrs++;
 	SC2IFP(sc)->if_flags |= IFF_UP;
-	if (own)
-		sc->sc_advskew = 0;
-	carp_sc_state_locked(sc);
+	if (sc->sc_naddrs == 1)
+		carp_sc_state_locked(sc);
 	carp_setrun(sc);
 
 	CARP_WUNLOCK(cif);
@@ -1648,7 +1644,7 @@ carp_set_addr6(struct carp_softc *sc, struct sockaddr_in6 *sin6)
 	struct in6_ifaddr *ia, *ia_if;
 	struct ip6_moptions *im6o = &sc->sc_im6o;
 	struct in6_addr in6;
-	int own, error;
+	int error;
 
 	error = 0;
 
@@ -1666,7 +1662,7 @@ carp_set_addr6(struct carp_softc *sc, struct sockaddr_in6 *sin6)
 	}
 
 	/* we have to do it by hands to check we won't match on us */
-	ia_if = NULL; own = 0;
+	ia_if = NULL;
 	IN6_IFADDR_RLOCK();
 	TAILQ_FOREACH(ia, &V_in6_ifaddrhead, ia_link) {
 		int i;
@@ -1684,9 +1680,6 @@ carp_set_addr6(struct carp_softc *sc, struct sockaddr_in6 *sin6)
 		    (i == 4)) {
 			if (!ia_if)
 				ia_if = ia;
-			if (IN6_ARE_ADDR_EQUAL(&sin6->sin6_addr,
-			    &ia->ia_addr.sin6_addr))
-				own++;
 		}
 	}
 
@@ -1799,9 +1792,8 @@ carp_set_addr6(struct carp_softc *sc, struct sockaddr_in6 *sin6)
 
 	sc->sc_naddrs6++;
 	SC2IFP(sc)->if_flags |= IFF_UP;
-	if (own)
-		sc->sc_advskew = 0;
-	carp_sc_state_locked(sc);
+	if (sc->sc_naddrs6 == 1)
+		carp_sc_state_locked(sc);
 	carp_setrun(sc);
 
 	CARP_WUNLOCK(cif);
