diff --git a/sbin/ifconfig/ifconfig.c b/sbin/ifconfig/ifconfig.c
index 6bc02b6..851afe4 100644
--- a/sbin/ifconfig/ifconfig.c
+++ b/sbin/ifconfig/ifconfig.c
@@ -862,7 +862,7 @@ unsetifdescr(const char *val, int value, int s, const struct afswtch *afp)
 #define	IFFBITS \
 "\020\1UP\2BROADCAST\3DEBUG\4LOOPBACK\5POINTOPOINT\6SMART\7RUNNING" \
 "\10NOARP\11PROMISC\12ALLMULTI\13OACTIVE\14SIMPLEX\15LINK0\16LINK1\17LINK2" \
-"\20MULTICAST\22PPROMISC\23MONITOR\24STATICARP"
+"\20MULTICAST\22PPROMISC\23MONITOR\24STATICARP\25IPFW_FILTER"
 
 #define	IFCAPBITS \
 "\020\1RXCSUM\2TXCSUM\3NETCONS\4VLAN_MTU\5VLAN_HWTAGGING\6JUMBO_MTU\7POLLING" \
@@ -1127,6 +1127,8 @@ static struct cmd basic_cmds[] = {
 	DEF_CMD("-monitor",	-IFF_MONITOR,	setifflags),
 	DEF_CMD("staticarp",	IFF_STATICARP,	setifflags),
 	DEF_CMD("-staticarp",	-IFF_STATICARP,	setifflags),
+	DEF_CMD("ipfwfilter",	IFF_IPFW_FILTER,	setifflags),
+	DEF_CMD("-ipfwfilter",	-IFF_IPFW_FILTER,	setifflags),
 	DEF_CMD("rxcsum",	IFCAP_RXCSUM,	setifcap),
 	DEF_CMD("-rxcsum",	-IFCAP_RXCSUM,	setifcap),
 	DEF_CMD("txcsum",	IFCAP_TXCSUM,	setifcap),
diff --git a/sbin/ipfw/ipfw2.c b/sbin/ipfw/ipfw2.c
index dcf0067..39921c5 100644
--- a/sbin/ipfw/ipfw2.c
+++ b/sbin/ipfw/ipfw2.c
@@ -377,7 +377,9 @@ do_cmd(int optname, void *optval, uintptr_t optlen)
 	    optname == IP_FW_TABLE_GETSIZE ||
 	    optname == IP_FW_NAT_GET_CONFIG ||
 	    optname < 0 ||
-	    optname == IP_FW_NAT_GET_LOG) {
+	    optname == IP_FW_NAT_GET_LOG ||
+	    optname == IP_FW_TABLE_GET_ENTRY ||
+	    optname == IP_FW_TABLE_ZERO_ENTRY_STATS) {
 		if (optname < 0)
 			optname = -optname;
 		i = getsockopt(s, IPPROTO_IP, optname, optval,
@@ -3773,6 +3775,7 @@ ipfw_flush(int force)
 
 
 static void table_list(ipfw_table_entry ent, int need_header);
+static void table_list_entry(ipfw_table_entry ent);
 
 /*
  * This one handles all table-related commands
@@ -3840,6 +3843,12 @@ ipfw_table_handler(int ac, char *av[])
 		if (lookup_host(*av, (struct in_addr *)&ent.addr) != 0)
 			errx(EX_NOHOST, "hostname ``%s'' unknown", *av);
 		ac--; av++;
+		ent.mac_addr = 0;
+		if (do_add && ac >= 2 && strcmp(*av, "mac") == 0) {
+			uint8_t mask[8];
+			get_mac_addr_mask(av[1], (uint8_t*)&ent.mac_addr, mask);
+			ac-=2; av+=2;
+		}
 		if (do_add && ac) {
 			unsigned int tval;
 			/* isdigit is a bit of a hack here.. */
@@ -3883,11 +3892,61 @@ ipfw_table_handler(int ac, char *av[])
 		do {
 			table_list(ent, is_all);
 		} while (++ent.tbl < a);
+	} else if (_substrcmp(*av, "entrystats") == 0) {
+		ac--; av++;
+                if (!ac)
+                        errx(EX_USAGE, "IP address required");
+                p = strchr(*av, '/');
+                if (p) {
+                        *p++ = '\0';
+                        ent.masklen = atoi(p);
+                        if (ent.masklen > 32)
+                                errx(EX_DATAERR, "bad width ``%s''", p);
+                } else
+                        ent.masklen = 32;
+                if (lookup_host(*av, (struct in_addr *)&ent.addr) != 0)
+                        errx(EX_NOHOST, "hostname ``%s'' unknown", *av);
+                ac--; av++;
+		ent.value = 0;
+		table_list_entry(ent);	
+	} else if (_substrcmp(*av, "entryzerostats") == 0) {
+		ac--; av++;
+                if (!ac)
+                        errx(EX_USAGE, "IP address required");
+                p = strchr(*av, '/');
+                if (p) {
+                        *p++ = '\0';
+                        ent.masklen = atoi(p);
+                        if (ent.masklen > 32)
+                                errx(EX_DATAERR, "bad width ``%s''", p);
+                } else
+                        ent.masklen = 32;
+                if (lookup_host(*av, (struct in_addr *)&ent.addr) != 0)
+                        errx(EX_NOHOST, "hostname ``%s'' unknown", *av);
+                ac--; av++;
+                ent.value = 0;
+		if (do_cmd(IP_FW_TABLE_ZERO_ENTRY_STATS, &ent, sizeof(ent)) < 0)
+                	err(EX_OSERR, "getsockopt(IP_FW_TABLE_ZERO_ENTRY_STATS)");
 	} else
 		errx(EX_USAGE, "invalid table command %s", *av);
 }
 
 static void
+table_list_entry(ipfw_table_entry ent) {
+        socklen_t l;
+	char tbuf[128];
+
+        l = sizeof(ent);
+        if (do_cmd(IP_FW_TABLE_GET_ENTRY, &ent, (uintptr_t)&l) < 0)
+                err(EX_OSERR, "getsockopt(IP_FW_TABLE_GET_ENTRY)");
+
+	strncpy(tbuf, inet_ntoa(*(struct in_addr *)
+		&ent.addr), 127);
+	printf("%s/%u %u %llu %llu %u\n", tbuf, ent.masklen,
+              ent.value, align_uint64(&ent.packets), align_uint64(&ent.bytes), ent.timestamp);
+}
+
+static void
 table_list(ipfw_table_entry ent, int need_header)
 {
 	ipfw_table *tbl;
@@ -3912,20 +3971,28 @@ table_list(ipfw_table_entry ent, int need_header)
 		printf("---table(%d)---\n", tbl->tbl);
 	for (a = 0; a < tbl->cnt; a++) {
 		unsigned int tval;
+		char tval_buf[128];
+		char tmac_buf[128];
 		tval = tbl->ent[a].value;
 		if (co.do_value_as_ip) {
-			char tbuf[128];
-			strncpy(tbuf, inet_ntoa(*(struct in_addr *)
-				&tbl->ent[a].addr), 127);
 			/* inet_ntoa expects network order */
 			tval = htonl(tval);
-			printf("%s/%u %s\n", tbuf, tbl->ent[a].masklen,
-				inet_ntoa(*(struct in_addr *)&tval));
+			strlcpy(tval_buf, inet_ntoa(*(struct in_addr *)
+				&tval), sizeof(tval_buf));
+		} else {
+			snprintf(tval_buf, sizeof(tval_buf), "%u", tval);
+		}
+		if (tbl->ent[a].mac_addr) {
+			uint8_t *x = (uint8_t *)&tbl->ent[a].mac_addr;
+			snprintf(tmac_buf, sizeof(tmac_buf), "mac %02x:%02x:%02x:%02x:%02x:%02x ",
+				x[0], x[1], x[2], x[3], x[4], x[5]);
 		} else {
-			printf("%s/%u %u\n",
-				inet_ntoa(*(struct in_addr *)&tbl->ent[a].addr),
-				tbl->ent[a].masklen, tval);
+			tmac_buf[0] = 0;
 		}
+		
+		printf("%s/%u %s%s %llu %llu\n",
+			inet_ntoa(*(struct in_addr *)&tbl->ent[a].addr),
+			tbl->ent[a].masklen, tmac_buf, tval_buf, tbl->ent[a].packets, tbl->ent[a].bytes);
 	}
 	free(tbl);
 }
diff --git a/sys/net/if.h b/sys/net/if.h
index 0046233..e6e3cae 100644
--- a/sys/net/if.h
+++ b/sys/net/if.h
@@ -151,6 +151,7 @@ struct if_data {
 #define	IFF_STATICARP	0x80000		/* (n) static ARP */
 #define	IFF_DYING	0x200000	/* (n) interface is winding down */
 #define	IFF_RENAMING	0x400000	/* (n) interface is being renamed */
+#define	IFF_IPFW_FILTER 0x100000        /* pfSense hack for CP speeding up */
 
 /*
  * Old names for driver flags so that user space tools can continue to use
diff --git a/sys/net/if_ethersubr.c b/sys/net/if_ethersubr.c
index d5ed209..3b3d2ff 100644
--- a/sys/net/if_ethersubr.c
+++ b/sys/net/if_ethersubr.c
@@ -143,7 +143,7 @@ MALLOC_DEFINE(M_ARPCOM, "arpcom", "802.* interface internals");
 
 #if defined(INET) || defined(INET6)
 int
-ether_ipfw_chk(struct mbuf **m0, struct ifnet *dst, int shared);
+ether_ipfw_chk(struct mbuf **m0, struct ifnet *dst, int shared, int dir);
 static VNET_DEFINE(int, ether_ipfw);
 #define	V_ether_ipfw	VNET(ether_ipfw)
 #endif
@@ -430,7 +430,7 @@ ether_output_frame(struct ifnet *ifp, struct mbuf *m)
 #if defined(INET) || defined(INET6)
 
 	if (V_ip_fw_chk_ptr && V_ether_ipfw != 0) {
-		if (ether_ipfw_chk(&m, ifp, 0) == 0) {
+		if (ether_ipfw_chk(&m, ifp, 0, DIR_OUT) == 0) {
 			if (m) {
 				m_freem(m);
 				return EACCES;	/* pkt dropped */
@@ -454,7 +454,7 @@ ether_output_frame(struct ifnet *ifp, struct mbuf *m)
  * ether_output_frame.
  */
 int
-ether_ipfw_chk(struct mbuf **m0, struct ifnet *dst, int shared)
+ether_ipfw_chk(struct mbuf **m0, struct ifnet *dst, int shared, int dir)
 {
 	struct ether_header *eh;
 	struct ether_header save_eh;
@@ -474,9 +474,11 @@ ether_ipfw_chk(struct mbuf **m0, struct ifnet *dst, int shared)
 		/* XXX can we free it after use ? */
 		mtag->m_tag_id = PACKET_TAG_NONE;
 		r = (struct ipfw_rule_ref *)(mtag + 1);
-		if (r->info & IPFW_ONEPASS)
+		//if (r->info & IPFW_ONEPASS) {
+			m_tag_delete(*m0, mtag);
 			return (1);
-		args.rule = *r;
+		//}
+		//args.rule = *r;
 	}
 
 	/*
@@ -501,6 +503,7 @@ ether_ipfw_chk(struct mbuf **m0, struct ifnet *dst, int shared)
 	args.next_hop = NULL;	/* we do not support forward yet	*/
 	args.eh = &save_eh;	/* MAC header for bridged/MAC packets	*/
 	args.inp = NULL;	/* used by ipfw uid/gid/jail rules	*/
+	args.dir = dir;		/* pfSense addition			*/
 	i = V_ip_fw_chk_ptr(&args);
 	m = args.m;
 	if (m != NULL) {
@@ -528,7 +531,7 @@ ether_ipfw_chk(struct mbuf **m0, struct ifnet *dst, int shared)
 		return 1;
 
 	if (ip_dn_io_ptr && (i == IP_FW_DUMMYNET)) {
-		int dir;
+		//int dir;
 		/*
 		 * Pass the pkt to dummynet, which consumes it.
 		 * If shared, make a copy and keep the original.
@@ -544,7 +547,7 @@ ether_ipfw_chk(struct mbuf **m0, struct ifnet *dst, int shared)
 			 */
 			*m0 = NULL ;
 		}
-		dir = PROTO_LAYER2 | (dst ? DIR_OUT : DIR_IN);
+		dir = PROTO_LAYER2 | dir;
 		ip_dn_io_ptr(&m, dir, &args);
 		return 0;
 	}
@@ -777,7 +780,7 @@ ether_demux(struct ifnet *ifp, struct mbuf *m)
 	 * Do not do this for PROMISC frames in case we are re-entered.
 	 */
 	if (V_ip_fw_chk_ptr && V_ether_ipfw != 0 && !(m->m_flags & M_PROMISC)) {
-		if (ether_ipfw_chk(&m, NULL, 0) == 0) {
+		if (ether_ipfw_chk(&m, ifp, 0, DIR_IN) == 0) {
 			if (m)
 				m_freem(m);	/* dropped; free mbuf chain */
 			return;			/* consumed */
@@ -820,7 +823,8 @@ ether_demux(struct ifnet *ifp, struct mbuf *m)
 	 * Strip off Ethernet header.
 	 */
 	m->m_flags &= ~M_VLANTAG;
-	m->m_flags &= ~(M_PROTOFLAGS);
+	if (!(m->m_flags & M_FASTFWD_OURS))
+		m->m_flags &= ~(M_PROTOFLAGS);
 	m_adj(m, ETHER_HDR_LEN);
 
 	/*
diff --git a/sys/netinet/in.h b/sys/netinet/in.h
index 233b44f..5688e10 100644
--- a/sys/netinet/in.h
+++ b/sys/netinet/in.h
@@ -453,6 +453,8 @@ __END_DECLS
 #define	IP_FW_TABLE_FLUSH	42   /* flush table */
 #define	IP_FW_TABLE_GETSIZE	43   /* get table size */
 #define	IP_FW_TABLE_LIST	44   /* list table contents */
+#define	IP_FW_TABLE_GET_ENTRY	45   /* get statistics about a table entry */
+#define IP_FW_TABLE_ZERO_ENTRY_STATS	46	/* zero table entry stats */
 
 #define	IP_FW3			48   /* generic ipfw v.3 sockopts */
 #define	IP_DUMMYNET3		49   /* generic dummynet v.3 sockopts */
diff --git a/sys/netinet/ip_fw.h b/sys/netinet/ip_fw.h
index 6f418ec..852c2ea 100644
--- a/sys/netinet/ip_fw.h
+++ b/sys/netinet/ip_fw.h
@@ -570,9 +570,13 @@ struct _ipfw_dyn_rule {
  */
 typedef struct	_ipfw_table_entry {
 	in_addr_t	addr;		/* network address		*/
+	u_int64_t       mac_addr;       /* mac address                  */
 	u_int32_t	value;		/* value			*/
 	u_int16_t	tbl;		/* table number			*/
 	u_int8_t	masklen;	/* mask length			*/
+        u_int64_t               bytes;
+        u_int64_t               packets;
+        u_int32_t               timestamp;
 } ipfw_table_entry;
 
 typedef struct	_ipfw_table {
diff --git a/sys/netinet/ipfw/ip_fw2.c b/sys/netinet/ipfw/ip_fw2.c
index 758de4e..c7975f7 100644
--- a/sys/netinet/ipfw/ip_fw2.c
+++ b/sys/netinet/ipfw/ip_fw2.c
@@ -831,6 +831,9 @@ ipfw_chk(struct ip_fw_args *args)
 	 */
 	struct ifnet *oif = args->oif;
 
+	if (oif && !(oif->if_flags & IFF_IPFW_FILTER))
+		return (IP_FW_PASS);
+
 	int f_pos = 0;		/* index of current rule in the array */
 	int retval = 0;
 
@@ -882,6 +885,7 @@ ipfw_chk(struct ip_fw_args *args)
 	int dyn_dir = MATCH_UNKNOWN;
 	ipfw_dyn_rule *q = NULL;
 	struct ip_fw_chain *chain = &V_layer3_chain;
+	struct table_entry *tblent = NULL;
 
 	/*
 	 * We store in ulp a pointer to the upper layer protocol header.
@@ -1215,6 +1219,7 @@ do {								\
 			continue;
 
 		skip_or = 0;
+		tblent = NULL;
 		for (l = f->cmd_len, cmd = f->cmd ; l > 0 ;
 		    l -= cmdlen, cmd += cmdlen) {
 			int match;
@@ -1332,7 +1337,7 @@ do {								\
 				break;
 
 			case O_IN:	/* "out" is "not in" */
-				match = (oif == NULL);
+				match = (args->dir == DIR_IN);
 				break;
 
 			case O_LAYER2:
@@ -1368,11 +1373,17 @@ do {								\
 			case O_IP_SRC_LOOKUP:
 			case O_IP_DST_LOOKUP:
 				if (is_ipv4) {
+				    struct ether_addr *ea = NULL;
 				    uint32_t key =
 					(cmd->opcode == O_IP_DST_LOOKUP) ?
 					    dst_ip.s_addr : src_ip.s_addr;
 				    uint32_t v = 0;
 
+				    if (args->eh) {
+					ea = (struct ether_addr*)((cmd->opcode == O_IP_DST_LOOKUP) ?
+						args->eh->ether_dhost :
+						args->eh->ether_shost);
+				    }
 				    if (cmdlen > F_INSN_SIZE(ipfw_insn_u32)) {
 					/* generic lookup. The key must be
 					 * in 32bit big-endian format.
@@ -1417,10 +1428,14 @@ do {								\
 					} else
 					    break;
 				    }
-				    match = ipfw_lookup_table(chain,
-					cmd->arg1, key, &v);
-				    if (!match)
+				    tblent = ipfw_lookup_table(chain,
+					cmd->arg1, key, ea, &v);
+				    if (tblent == NULL) {
+					match = 0;
 					break;
+				    } else
+					match = 1;
+
 				    if (cmdlen == F_INSN_SIZE(ipfw_insn_u32))
 					match =
 					    ((ipfw_insn_u32 *)cmd)->d[0] == v;
@@ -1701,7 +1716,7 @@ do {								\
 
 			case O_ANTISPOOF:
 				/* Outgoing packets automatically pass/match */
-				if (oif == NULL && hlen > 0 &&
+				if (args->dir == DIR_IN && hlen > 0 &&
 				    (  (is_ipv4 && in_localaddr(src_ip))
 #ifdef INET6
 				    || (is_ipv6 &&
@@ -2195,8 +2210,7 @@ do {								\
 				break;
 
 			case O_FORWARD_IP:
-				if (args->eh)	/* not valid on layer2 pkts */
-					break;
+				if (!args->eh) {
 				if (q == NULL || q->rule != f ||
 				    dyn_dir == MATCH_FORWARD) {
 				    struct sockaddr_in *sa;
@@ -2211,6 +2225,49 @@ do {								\
 					args->next_hop = sa;
 				    }
 				}
+				} else if (args->eh) {
+					struct m_tag *fwd_tag;
+				    	struct sockaddr_in *sa;
+					u_short sum;
+
+					/*
+					 * Checksum correct? (from ip_fastfwd.c)
+					 */
+					if (m->m_pkthdr.csum_flags & CSUM_IP_CHECKED)
+						sum = !(m->m_pkthdr.csum_flags & CSUM_IP_VALID);
+					else {
+						if (hlen == sizeof(struct ip))
+							sum = in_cksum_hdr(ip);
+						else
+							sum = in_cksum(m, hlen);
+					}
+					if (sum) {
+						IPSTAT_INC(ips_badsum);
+						retval = IP_FW_DENY;
+						break;
+					}
+
+					/*
+					 * Remember that we have checked the IP header and found it valid.
+					 */
+					m->m_pkthdr.csum_flags |= (CSUM_IP_CHECKED | CSUM_IP_VALID);
+
+				    	sa = &(((ipfw_insn_sa *)cmd)->sa);
+					fwd_tag = m_tag_get(PACKET_TAG_IPFORWARD,
+                                		sizeof(struct sockaddr_in), M_NOWAIT);
+                			if (fwd_tag == NULL)
+						retval = IP_FW_DENY;
+					else {
+                				bcopy(sa, (fwd_tag+1), sizeof(struct sockaddr_in));
+                				m_tag_prepend(m, fwd_tag);
+ 
+						if (in_localip(sa->sin_addr)) {
+                        				m->m_flags |= M_FASTFWD_OURS;
+							ip->ip_len = ntohs(ip->ip_len);
+							ip->ip_off = ntohs(ip->ip_off);
+						}
+					}
+				}
 				retval = IP_FW_PASS;
 				l = 0;          /* exit inner loop */
 				done = 1;       /* exit outer loop */
@@ -2357,6 +2414,11 @@ do {								\
 		rule->pcnt++;
 		rule->bcnt += pktlen;
 		rule->timestamp = time_uptime;
+		if (tblent != NULL) {
+			tblent->packets++;
+			tblent->bytes += pktlen;
+			tblent->timestamp = time_uptime;
+		}
 	} else {
 		retval = IP_FW_DENY;
 		printf("ipfw: ouch!, skip past end of rules, denying packet\n");
diff --git a/sys/netinet/ipfw/ip_fw_pfil.c b/sys/netinet/ipfw/ip_fw_pfil.c
index ab86270..89b6f72 100644
--- a/sys/netinet/ipfw/ip_fw_pfil.c
+++ b/sys/netinet/ipfw/ip_fw_pfil.c
@@ -106,6 +106,9 @@ ipfw_check_hook(void *arg, struct mbuf **m0, struct ifnet *ifp, int dir,
 	int ipfw;
 	int ret;
 
+	if (!(ifp->if_flags & IFF_IPFW_FILTER))
+		return (0);
+	
 	/* all the processing now uses ip_len in net format */
 	if (mtod(*m0, struct ip *)->ip_v == 4)
 		SET_NET_IPLEN(mtod(*m0, struct ip *));
@@ -131,7 +134,8 @@ again:
 	}
 
 	args.m = *m0;
-	args.oif = dir == DIR_OUT ? ifp : NULL;
+	args.oif = ifp;
+	args.dir = dir;
 	args.inp = inp;
 
 	ipfw = ipfw_chk(&args);
@@ -338,7 +342,7 @@ ipfw_hook(int onoff, int pf)
 		return ENOENT;
 
 	if (onoff)
-	(void) pfil_add_hook
+	(void) pfil_add_named_hook
 	    (ipfw_check_hook, NULL, "ipfw", PFIL_IN | PFIL_OUT | PFIL_WAITOK, pfh);
 	else
 	(void) pfil_remove_hook
diff --git a/sys/netinet/ipfw/ip_fw_private.h b/sys/netinet/ipfw/ip_fw_private.h
index caa3bae..2d1e381 100644
--- a/sys/netinet/ipfw/ip_fw_private.h
+++ b/sys/netinet/ipfw/ip_fw_private.h
@@ -34,6 +34,7 @@
  */
 
 #ifdef _KERNEL
+#include <net/radix.h>
 
 /*
  * For platforms that do not have SYSCTL support, we wrap the
@@ -61,6 +62,16 @@ enum {
 	IP_FW_REASS,
 };
 
+struct table_entry {
+        struct radix_node       rn[2];
+        struct sockaddr_in      addr, mask;
+        u_int64_t               mac_addr;
+        u_int32_t               value;
+        u_int64_t               bytes;
+        u_int64_t               packets;
+        u_int32_t               timestamp;
+};
+
 /*
  * Structure for collecting parameters to dummynet for ip6_output forwarding
  */
@@ -96,6 +107,7 @@ struct ip_fw_args {
 	 */
 	struct ipfw_rule_ref rule;	/* match/restart info		*/
 
+	uint32_t        dir;            /* direction */
 	struct ether_header *eh;	/* for bridged packets		*/
 
 	struct ipfw_flow_id f_id;	/* grabbed from IP header	*/
@@ -269,18 +281,24 @@ int ipfw_check_hook(void *arg, struct mbuf **m0, struct ifnet *ifp, int dir,
      struct inpcb *inp);
 
 /* In ip_fw_table.c */
-struct radix_node;
-int ipfw_lookup_table(struct ip_fw_chain *ch, uint16_t tbl, in_addr_t addr,
-    uint32_t *val);
+int
+ipfw_lookup_table_entry(struct ip_fw_chain *ch, ipfw_table_entry *);
+struct ether_addr;
+//struct radix_node;
+struct table_entry * ipfw_lookup_table(struct ip_fw_chain *ch, uint16_t tbl, in_addr_t addr,
+    struct ether_addr *ea, uint32_t *val);
 int ipfw_init_tables(struct ip_fw_chain *ch);
 void ipfw_destroy_tables(struct ip_fw_chain *ch);
+int ipfw_flush_table_entry(struct radix_node *rn, void *arg);
 int ipfw_flush_table(struct ip_fw_chain *ch, uint16_t tbl);
+int ipfw_zero_table_entry_stats(struct ip_fw_chain *ch, uint16_t tbl, in_addr_t addr);
 int ipfw_add_table_entry(struct ip_fw_chain *ch, uint16_t tbl, in_addr_t addr,
-    uint8_t mlen, uint32_t value);
-int ipfw_dump_table_entry(struct radix_node *rn, void *arg);
+    uint8_t mlen, u_int64_t mac_addr, uint32_t value);
 int ipfw_del_table_entry(struct ip_fw_chain *ch, uint16_t tbl, in_addr_t addr,
     uint8_t mlen);
+int ipfw_count_table_entry(struct radix_node *rn, void *arg);
 int ipfw_count_table(struct ip_fw_chain *ch, uint32_t tbl, uint32_t *cnt);
+int ipfw_dump_table_entry(struct radix_node *rn, void *arg);
 int ipfw_dump_table(struct ip_fw_chain *ch, ipfw_table *tbl);
 
 /* In ip_fw_nat.c -- XXX to be moved to ip_var.h */
diff --git a/sys/netinet/ipfw/ip_fw_sockopt.c b/sys/netinet/ipfw/ip_fw_sockopt.c
index 4da2649..e9fa478 100644
--- a/sys/netinet/ipfw/ip_fw_sockopt.c
+++ b/sys/netinet/ipfw/ip_fw_sockopt.c
@@ -1099,7 +1099,7 @@ ipfw_ctl(struct sockopt *sopt)
 			if (error)
 				break;
 			error = ipfw_add_table_entry(chain, ent.tbl,
-			    ent.addr, ent.masklen, ent.value);
+			    ent.addr, ent.masklen, ent.mac_addr, ent.value);
 		}
 		break;
 
@@ -1146,6 +1146,37 @@ ipfw_ctl(struct sockopt *sopt)
 		}
 		break;
 
+	case IP_FW_TABLE_GET_ENTRY:
+		{
+			ipfw_table_entry ent;
+			time_t boot_seconds;
+
+			boot_seconds = boottime.tv_sec;
+			error = sooptcopyin(sopt, &ent,
+				sizeof(ent), sizeof(ent));
+			if (error)
+				break;
+			if (ipfw_lookup_table_entry(chain, &ent)) {
+				ent.timestamp += boot_seconds;
+				error = sooptcopyout(sopt, &ent, sizeof(ent));
+			} else
+				error = ENOENT;
+		}
+		break;
+
+	case IP_FW_TABLE_ZERO_ENTRY_STATS:
+		{
+			ipfw_table_entry ent;
+
+			error = sooptcopyin(sopt, &ent,
+				sizeof(ent), sizeof(ent));
+			if (error)
+				break;
+			error = ipfw_zero_table_entry_stats(chain, ent.tbl,
+				ent.addr);
+		}
+		break;
+
 	case IP_FW_TABLE_LIST:
 		{
 			ipfw_table *tbl;
diff --git a/sys/netinet/ipfw/ip_fw_table.c b/sys/netinet/ipfw/ip_fw_table.c
index b167aa5..14584a3 100644
--- a/sys/netinet/ipfw/ip_fw_table.c
+++ b/sys/netinet/ipfw/ip_fw_table.c
@@ -60,6 +60,7 @@ __FBSDID("$FreeBSD: src/sys/netinet/ipfw/ip_fw_table.c,v 1.9.2.3.2.1 2012/03/03
 
 #include <netinet/in.h>
 #include <netinet/ip_var.h>	/* struct ipfw_rule_ref */
+#include <net/ethernet.h>       /* struct ether_addr */
 #include <netinet/ip_fw.h>
 #include <sys/queue.h> /* LIST_HEAD */
 #include <netinet/ipfw/ip_fw_private.h>
@@ -70,11 +71,17 @@ __FBSDID("$FreeBSD: src/sys/netinet/ipfw/ip_fw_table.c,v 1.9.2.3.2.1 2012/03/03
 
 MALLOC_DEFINE(M_IPFW_TBL, "ipfw_tbl", "IpFw tables");
 
+#if 0
 struct table_entry {
 	struct radix_node	rn[2];
 	struct sockaddr_in	addr, mask;
+	u_int64_t               mac_addr;
 	u_int32_t		value;
+	u_int64_t               bytes;
+	u_int64_t               packets;
+	u_int32_t               timestamp;
 };
+#endif
 
 /*
  * The radix code expects addr and mask to be array of bytes,
@@ -90,7 +97,7 @@ struct table_entry {
 
 int
 ipfw_add_table_entry(struct ip_fw_chain *ch, uint16_t tbl, in_addr_t addr,
-    uint8_t mlen, uint32_t value)
+    uint8_t mlen, u_int64_t mac_addr, uint32_t value)
 {
 	struct radix_node_head *rnh;
 	struct table_entry *ent;
@@ -106,10 +113,22 @@ ipfw_add_table_entry(struct ip_fw_chain *ch, uint16_t tbl, in_addr_t addr,
 	KEY_LEN(ent->addr) = KEY_LEN(ent->mask) = 8;
 	ent->mask.sin_addr.s_addr = htonl(mlen ? ~((1 << (32 - mlen)) - 1) : 0);
 	ent->addr.sin_addr.s_addr = addr & ent->mask.sin_addr.s_addr;
+	ent->mac_addr = mac_addr;
 	IPFW_WLOCK(ch);
 	rn = rnh->rnh_addaddr(&ent->addr, &ent->mask, rnh, (void *)ent);
 	if (rn == NULL) {
 		IPFW_WUNLOCK(ch);
+		/* Just update if any new value needed */
+		struct table_entry *ent2;
+		ent->addr.sin_addr.s_addr = addr;
+		ent2 = (struct table_entry *)(rnh->rnh_lookup(&ent->addr, NULL, rnh));
+		if (ent2 != NULL) {
+			if (ent2->mac_addr) {
+				if (!bcmp(&mac_addr, &ent2->mac_addr, ETHER_ADDR_LEN)) {
+					ent2->value = value;
+			} else
+				ent2->value = value;
+		}
 		free(ent, M_IPFW_TBL);
 		return (EEXIST);
 	}
@@ -142,8 +161,8 @@ ipfw_del_table_entry(struct ip_fw_chain *ch, uint16_t tbl, in_addr_t addr,
 	return (0);
 }
 
-static int
-flush_table_entry(struct radix_node *rn, void *arg)
+int
+ipfw_flush_table_entry(struct radix_node *rn, void *arg)
 {
 	struct radix_node_head * const rnh = arg;
 	struct table_entry *ent;
@@ -166,7 +185,7 @@ ipfw_flush_table(struct ip_fw_chain *ch, uint16_t tbl)
 		return (EINVAL);
 	rnh = ch->tables[tbl];
 	KASSERT(rnh != NULL, ("NULL IPFW table"));
-	rnh->rnh_walktree(rnh, flush_table_entry, rnh);
+	rnh->rnh_walktree(rnh, ipfw_flush_table_entry, rnh);
 	return (0);
 }
 
@@ -202,9 +221,9 @@ ipfw_init_tables(struct ip_fw_chain *ch)
 	return (0);
 }
 
-int
+struct table_entry *
 ipfw_lookup_table(struct ip_fw_chain *ch, uint16_t tbl, in_addr_t addr,
-    uint32_t *val)
+    struct ether_addr *ea, uint32_t *val)
 {
 	struct radix_node_head *rnh;
 	struct table_entry *ent;
@@ -217,14 +236,64 @@ ipfw_lookup_table(struct ip_fw_chain *ch, uint16_t tbl, in_addr_t addr,
 	sa.sin_addr.s_addr = addr;
 	ent = (struct table_entry *)(rnh->rnh_lookup(&sa, NULL, rnh));
 	if (ent != NULL) {
+		if (ea && ent->mac_addr) {
+			u_char *test = (u_char *)&ent->mac_addr;
+			if (bcmp(test, ea->octet, ETHER_ADDR_LEN) != 0)
+				return (0);
+		}
 		*val = ent->value;
 		return (1);
 	}
 	return (0);
 }
 
-static int
-count_table_entry(struct radix_node *rn, void *arg)
+int
+ipfw_lookup_table_entry(struct ip_fw_chain *ch, ipfw_table_entry *arg)
+{
+	struct radix_node_head *rnh;
+	struct table_entry *ent;
+	struct sockaddr_in sa;
+
+	if (arg->tbl >= IPFW_TABLES_MAX)
+		return (0);
+	rnh = ch->tables[arg->tbl];
+	sa.sin_len = 8;
+	sa.sin_addr.s_addr = arg->addr;
+	ent = (struct table_entry *)(rnh->rnh_lookup(&sa, NULL, rnh));
+	if (ent != NULL) {
+		arg->bytes = ent->bytes;
+		arg->packets = ent->packets;
+		arg->value = ent->value;
+		arg->timestamp = ent->timestamp;
+		return (1);
+	}
+	return (0);
+}
+
+int
+ipfw_zero_table_entry_stats(struct ip_fw_chain *ch, uint16_t tbl, in_addr_t addr)
+{
+	struct radix_node_head *rnh;
+	struct table_entry *ent;
+	struct sockaddr_in sa;
+
+	if (tbl >= IPFW_TABLES_MAX)
+		return (EINVAL);
+	rnh = ch->tables[tbl];
+	sa.sin_len = 8;
+	sa.sin_addr.s_addr = addr;
+	ent = (struct table_entry *)(rnh->rnh_lookup(&sa, NULL, rnh));
+	if (ent != NULL) {
+		ent->packets = 0;
+		ent->bytes = 0;
+		ent->timestamp = time_uptime;
+		return (0);
+	}
+	return (EINVAL);
+}
+
+int
+ipfw_count_table_entry(struct radix_node *rn, void *arg)
 {
 	u_int32_t * const cnt = arg;
 
@@ -241,12 +310,12 @@ ipfw_count_table(struct ip_fw_chain *ch, uint32_t tbl, uint32_t *cnt)
 		return (EINVAL);
 	rnh = ch->tables[tbl];
 	*cnt = 0;
-	rnh->rnh_walktree(rnh, count_table_entry, cnt);
+	rnh->rnh_walktree(rnh, ipfw_count_table_entry, cnt);
 	return (0);
 }
 
-static int
-dump_table_entry(struct radix_node *rn, void *arg)
+int
+ipfw_dump_table_entry(struct radix_node *rn, void *arg)
 {
 	struct table_entry * const n = (struct table_entry *)rn;
 	ipfw_table * const tbl = arg;
@@ -261,7 +330,10 @@ dump_table_entry(struct radix_node *rn, void *arg)
 	else
 		ent->masklen = 33 - ffs(ntohl(n->mask.sin_addr.s_addr));
 	ent->addr = n->addr.sin_addr.s_addr;
+	ent->mac_addr = n->mac_addr;
 	ent->value = n->value;
+	ent->packets = n->packets;
+	ent->bytes = n->bytes;
 	tbl->cnt++;
 	return (0);
 }
@@ -275,7 +347,7 @@ ipfw_dump_table(struct ip_fw_chain *ch, ipfw_table *tbl)
 		return (EINVAL);
 	rnh = ch->tables[tbl->tbl];
 	tbl->cnt = 0;
-	rnh->rnh_walktree(rnh, dump_table_entry, tbl);
+	rnh->rnh_walktree(rnh, ipfw_dump_table_entry, tbl);
 	return (0);
 }
 /* end of file */
diff --git a/sys/netinet/raw_ip.c b/sys/netinet/raw_ip.c
index 7af7606..560cff3 100644
--- a/sys/netinet/raw_ip.c
+++ b/sys/netinet/raw_ip.c
@@ -558,6 +558,8 @@ rip_ctloutput(struct socket *so, struct sockopt *sopt)
 		case IP_FW_GET:
 		case IP_FW_TABLE_GETSIZE:
 		case IP_FW_TABLE_LIST:
+		case IP_FW_TABLE_GET_ENTRY:
+		case IP_FW_TABLE_ZERO_ENTRY_STATS:
 		case IP_FW_NAT_GET_CONFIG:
 		case IP_FW_NAT_GET_LOG:
 			if (V_ip_fw_ctl_ptr != NULL)
