Index: contrib/pf/pfctl/parse.y
===================================================================
RCS file: /root/freebsd/repo/src/contrib/pf/pfctl/parse.y,v
retrieving revision 1.3
diff -u -r1.3 parse.y
--- contrib/pf/pfctl/parse.y	7 Feb 2011 23:23:11 -0000	1.3
+++ contrib/pf/pfctl/parse.y	7 Feb 2011 23:23:18 -0000
@@ -129,6 +129,7 @@
 enum	{ PF_STATE_OPT_MAX, PF_STATE_OPT_NOSYNC, PF_STATE_OPT_SRCTRACK,
 	    PF_STATE_OPT_MAX_SRC_STATES, PF_STATE_OPT_MAX_SRC_CONN,
 	    PF_STATE_OPT_MAX_SRC_CONN_RATE, PF_STATE_OPT_MAX_SRC_NODES,
+	    PF_STATE_OPT_MAX_PACKETS, 
 	    PF_STATE_OPT_OVERLOAD, PF_STATE_OPT_STATELOCK,
 	    PF_STATE_OPT_TIMEOUT, PF_STATE_OPT_SLOPPY };
 
@@ -140,6 +141,7 @@
 		u_int32_t	 max_states;
 		u_int32_t	 max_src_states;
 		u_int32_t	 max_src_conn;
+		u_int32_t	 max_packets;
 		struct {
 			u_int32_t	limit;
 			u_int32_t	seconds;
@@ -207,6 +209,7 @@
 	int			 allowopts;
 	char			*label;
 	struct node_qassign	 queues;
+	u_int16_t		 divert;
 	char			*tag;
 	char			*match_tag;
 	u_int8_t		 match_tag_not;
@@ -434,7 +437,7 @@
 %token	LOAD RULESET_OPTIMIZATION
 %token	STICKYADDRESS MAXSRCSTATES MAXSRCNODES SOURCETRACK GLOBAL RULE
 %token	MAXSRCCONN MAXSRCCONNRATE OVERLOAD FLUSH SLOPPY
-%token	TAGGED TAG IFBOUND FLOATING STATEPOLICY ROUTE
+%token	TAGGED TAG IFBOUND FLOATING DIVERT STATEPOLICY ROUTE MAXPCKT
 %token	<v.string>		STRING
 %token	<v.i>			PORTBINARY
 %type	<v.interface>		interface if_list if_item_not if_item
@@ -462,6 +465,7 @@
 %type	<v.route>		route
 %type	<v.redirection>		redirection redirpool
 %type	<v.string>		label string tag anchorname
+%type	<v.number>		divert
 %type	<v.keep_state>		keep
 %type	<v.state_opt>		state_opt_spec state_opt_list state_opt_item
 %type	<v.logquick>		logquick quick log logopts logopt
@@ -1784,6 +1788,8 @@
 			r.rtableid = $9.rtableid;
 
 			r.af = $6;
+			if ($9.divert)
+				r.divert = $9.divert;
 			if ($9.tag)
 				if (strlcpy(r.tagname, $9.tag,
 				    PF_TAG_NAME_SIZE) >= PF_TAG_NAME_SIZE) {
@@ -1869,6 +1875,14 @@
 					}
 					r.rule_flag |= PFRULE_NOSYNC;
 					break;
+				case PF_STATE_OPT_MAX_PACKETS:
+					if (o->data.max_packets == 0) {
+						yyerror("max_packets must be"
+							"greater than 0");
+						YYERROR;
+					}
+					r.spare1 = o->data.max_packets;
+					break;
 				case PF_STATE_OPT_SRCTRACK:
 					if (srctrack) {
 						yyerror("state option "
@@ -2258,6 +2272,9 @@
 			filter_opts.dnpipe = $3;
 			filter_opts.free_flags |= PFRULE_DN_IS_QUEUE;
 		}
+		| DIVERT number				{
+			filter_opts.divert = $2;
+		}
 		| TAG string				{
 			filter_opts.tag = $2;
 		}
@@ -3333,6 +3350,15 @@
 			$$->next = NULL;
 			$$->tail = $$;
 		}
+		| MAXPCKT number                    	{
+                        $$ = calloc(1, sizeof(struct node_state_opt));
+                        if ($$ == NULL)
+                                err(1, "state_opt_item: calloc");
+                        $$->type = PF_STATE_OPT_MAX_PACKETS;
+                        $$->data.max_src_states = $2;
+                        $$->next = NULL;
+                        $$->tail = $$;
+                }
 		| MAXSRCCONN number			{
 			$$ = calloc(1, sizeof(struct node_state_opt));
 			if ($$ == NULL)
@@ -4005,6 +4031,10 @@
 		}
 		;
 
+divert		: /* empty */		{ $$ = 0; }
+		| DIVERT number		{ $$ = $2;   }
+		;
+
 tag		: /* empty */		{ $$ = NULL; }
 		| TAG STRING		{ $$ = $2; }
 		;
@@ -5098,6 +5128,7 @@
 		{ "code",		CODE},
 		{ "crop",		FRAGCROP},
 		{ "debug",		DEBUG},
+		{ "divert",             DIVERT},
 		{ "dnpipe",             DNPIPE},
                 { "dnqueue",            DNQUEUE},
 		{ "drop",		DROP},
@@ -5134,6 +5165,7 @@
 		{ "loginterface",	LOGINTERFACE},
 		{ "max",		MAXIMUM},
 		{ "max-mss",		MAXMSS},
+		{ "max-packets",	MAXPCKT},
 		{ "max-src-conn",	MAXSRCCONN},
 		{ "max-src-conn-rate",	MAXSRCCONNRATE},
 		{ "max-src-nodes",	MAXSRCNODES},
Index: contrib/pf/pfctl/pfctl_parser.c
===================================================================
RCS file: /root/freebsd/repo/src/contrib/pf/pfctl/pfctl_parser.c,v
retrieving revision 1.3
diff -u -r1.3 pfctl_parser.c
--- contrib/pf/pfctl/pfctl_parser.c	7 Feb 2011 23:23:11 -0000	1.3
+++ contrib/pf/pfctl/pfctl_parser.c	7 Feb 2011 23:23:18 -0000
@@ -914,6 +914,12 @@
 			printf("max-src-conn %u", r->max_src_conn);
 			opts = 0;
 		}
+		if (r->spare1) {
+                        if (!opts)
+                                printf(", ");
+                        printf("max-packets %u", r->spare1);
+                        opts = 0;
+		}
 		if (r->max_src_conn_rate.limit) {
 			if (!opts)
 				printf(", ");
@@ -1003,6 +1009,8 @@
 		printf(" queue(%s, %s)", r->qname, r->pqname);
 	else if (r->qname[0])
 		printf(" queue %s", r->qname);
+	if (r->divert)
+		printf(" divert %u", r->divert);
 	if (r->tagname[0])
 		printf(" tag %s", r->tagname);
 	if (r->match_tagname[0]) {
Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.3
diff -u -r1.3 pf.c
--- sys/contrib/pf/net/pf.c	7 Feb 2011 23:23:11 -0000	1.3
+++ sys/contrib/pf/net/pf.c	7 Feb 2011 23:23:41 -0000
@@ -138,6 +138,7 @@
 #include <netinet/ip_fw.h>
 #include <netinet/ipfw/ip_fw_private.h>
 #include <netinet/ip_dummynet.h>
+#include <netinet/ip_divert.h>
 
 #include <netinet6/ip6_var.h>
 #include <netinet6/in6_pcb.h>
@@ -457,6 +458,14 @@
 		}				\
 } while(0)
 
+#define	PF_DIVERT_MAXPACKETS_REACHED()					\
+	do {								\
+		if (r->spare1 &&					\
+		    s->packets[dir == PF_OUT] > r->spare1)		\
+			/* fake that divert already happened */         \
+			pd.pf_mtag->flags |= PF_PACKET_LOOPED;		\
+	} while(0)
+
 #define STATE_LOOKUP()							\
 	do {								\
 		if (direction == PF_IN)					\
@@ -7087,6 +7096,8 @@
 #ifdef __FreeBSD__
 	struct m_tag *dn_tag;
 	struct ip_fw_args	 dnflow;
+	u_int16_t		 divertcookie = 0;
+        u_int8_t		 divflags = 0;
 	int			 loopedfrom = 0;
 
 	PF_LOCK();
@@ -7159,12 +7170,17 @@
 		goto done;
 	}
 
-	if (ip_dn_io_ptr != NULL &&
+	if (((ip_dn_io_ptr != NULL) || (ip_divert_ptr != NULL)) &&
 	    ((dn_tag = m_tag_locate(m, MTAG_IPFW_RULE, 0, NULL)) != NULL)) {
 		struct ipfw_rule_ref *rr = (struct ipfw_rule_ref *)(dn_tag+1);
 		pd.pf_mtag->flags |= PF_PACKET_LOOPED;
 		if (rr->info & IPFW_IS_DUMMYNET)
 			loopedfrom = 1;
+		if (rr->info & IPFW_IS_DIVERT) {
+			divertcookie = rr->rulenum;
+			divflags = (u_int8_t)(divertcookie >> 8);
+			divertcookie &= ~PFSTATE_DIVERT_MASK;
+		}
 		if (pd.pf_mtag->flags & PF_FASTFWD_OURS_PRESENT) {
                         m->m_flags |= M_FASTFWD_OURS;
                         pd.pf_mtag->flags &= ~PF_FASTFWD_OURS_PRESENT;
@@ -7363,6 +7379,19 @@
 		    ("pf: dropping packet with ip options\n"));
 	}
 
+#ifdef __FreeBSD__
+        if (s) {
+                PF_DIVERT_MAXPACKETS_REACHED();
+
+                if (divflags) {
+                        s->divert_cookie = divertcookie;
+                        s->local_flags |= divflags;
+                } else {
+                        divertcookie = s->divert_cookie;
+                        divflags = s->local_flags;
+                }
+        }
+#endif
 	if ((s && s->tag) || r->rtableid)
 		pf_tag_packet(m, pd.pf_mtag, s ? s->tag : 0, r->rtableid);
 
@@ -7382,9 +7411,62 @@
 #endif /* ALTQ */
 
 #ifdef __FreeBSD__
-       	if (r->dnpipe && ip_dn_io_ptr != NULL && loopedfrom != 1) {
+	if (divflags & PFSTATE_DIVERT_TAG)
+		pd.pf_mtag->tag = divertcookie;
+	else if (divflags & PFSTATE_DIVERT_ALTQ)
+		pd.pf_mtag->qid = divertcookie;
+	else if (divflags & PFSTATE_DIVERT_ACTION) {
+		struct pf_rule *dlr;
+		action = PF_DROP;
+		if (s)
+			pf_unlink_state(s);
+		REASON_SET(&reason, PFRES_DIVERT);
+		log = 1;
+		DPFPRINTF(PF_DEBUG_MISC,
+		    ("pf: changing action to with overload from divert.\n"));
+		dlr = r;
+		PFLOG_PACKET(kif, h, m, AF_INET, dir, reason, dlr, a,
+		    ruleset, &pd);
+		PF_UNLOCK();
+		m_freem(*m0);
+                *m0 = NULL;
+		/* NOTE: Fake this to avoid divert giving errors to the application. */
+		return (PF_PASS);
+	}
+
+	if (r->divert && ip_divert_ptr != NULL && !PACKET_LOOPED(pd.pf_mtag)) {
+		if (!r->spare1 ||
+		    (s && s->packets[dir == PF_OUT] <= r->spare1)) {
+
+			dn_tag = m_tag_alloc(MTAG_IPFW_RULE, 0,
+			    sizeof(struct ipfw_rule_ref), M_NOWAIT | M_ZERO);
+			if (dn_tag != NULL) {
+				((struct ipfw_rule_ref *)(dn_tag+1))->info = r->divert;
+				((struct ipfw_rule_ref *)(dn_tag+1))->rulenum = dir;
 
-                        if (dir != r->direction && r->pdnpipe) {
+				m_tag_prepend(m, dn_tag);
+
+				if (s != NULL && s->nat_rule.ptr)
+					PACKET_UNDO_NAT(m, pd, s, dir);
+				PF_UNLOCK();
+			
+				if (m->m_flags & M_FASTFWD_OURS) {
+                                	pd.pf_mtag->flags |= PF_FASTFWD_OURS_PRESENT;
+                                	m->m_flags &= ~M_FASTFWD_OURS;
+                        	}
+
+				ip_divert_ptr(*m0,
+				    dir ==  PF_IN ? DIR_IN : DIR_OUT);
+				*m0 = NULL;
+				return (action);
+			}
+		}
+       	} else if ((r->dnpipe || (divflags & PFSTATE_DIVERT_DNCOOKIE)) &&
+	    ip_dn_io_ptr != NULL && loopedfrom != 1) {
+
+			if (divflags & PFSTATE_DIVERT_DNCOOKIE) {
+				dnflow.rule.info = divertcookie;
+                        } else if (dir != r->direction && r->pdnpipe) {
                                 dnflow.rule.info = r->pdnpipe;
                         } else if (dir == r->direction) {
                                 dnflow.rule.info = r->dnpipe;
@@ -7425,8 +7507,7 @@
                         	*m0 = NULL;
                         	return (action);
 			}
-        } else
-		pd.pf_mtag->flags &= ~PF_PACKET_LOOPED;
+        }
 continueprocessing:
 #endif
 
@@ -7439,8 +7520,13 @@
 	    pd.proto == IPPROTO_UDP) && s != NULL && s->nat_rule.ptr != NULL &&
 	    (s->nat_rule.ptr->action == PF_RDR ||
 	    s->nat_rule.ptr->action == PF_BINAT) &&
-	    (ntohl(pd.dst->v4.s_addr) >> IN_CLASSA_NSHIFT) == IN_LOOPBACKNET)
+	    (ntohl(pd.dst->v4.s_addr) >> IN_CLASSA_NSHIFT) == IN_LOOPBACKNET) {
 		pd.pf_mtag->flags |= PF_TAG_TRANSLATE_LOCALHOST;
+		if (PACKET_LOOPED(pd.pf_mtag) && !loopedfrom)
+                        m->m_flags |= M_FASTFWD_OURS;
+	}
+
+	pd.pf_mtag->flags &= ~PF_PACKET_LOOPED;
 
 	if (log) {
 		struct pf_rule *lr;
Index: sys/contrib/pf/net/pfvar.h
===================================================================
RCS file: /root/freebsd/repo/src/sys/contrib/pf/net/pfvar.h,v
retrieving revision 1.3
diff -u -r1.3 pfvar.h
--- sys/contrib/pf/net/pfvar.h	7 Feb 2011 23:23:11 -0000	1.3
+++ sys/contrib/pf/net/pfvar.h	7 Feb 2011 23:23:18 -0000
@@ -649,6 +649,7 @@
 	u_int16_t		 tag;
 	u_int16_t		 match_tag;
 	u_int16_t		 spare2;		/* netgraph */
+	u_int16_t		 divert;		/* divert */
 
 	struct pf_rule_uid	 uid;
 	struct pf_rule_gid	 gid;
@@ -830,7 +831,12 @@
 	u_int8_t	 direction;
 #ifdef __FreeBSD__
 	u_int8_t	 local_flags;
-#define	PFSTATE_EXPIRING 0x01
+#define	PFSTATE_EXPIRING 	0x01
+#define PFSTATE_DIVERT_ALTQ     0x10
+#define PFSTATE_DIVERT_DNCOOKIE 0x20
+#define PFSTATE_DIVERT_ACTION   0x40
+#define PFSTATE_DIVERT_TAG      0x80
+#define PFSTATE_DIVERT_MASK	0xFF00
 #else
 	u_int8_t	 pad;
 #endif
@@ -868,6 +874,7 @@
 	u_int32_t	 expire;
 	u_int32_t	 pfsync_time;
 	u_int16_t	 tag;
+	u_int16_t	 divert_cookie;
 };
 
 TAILQ_HEAD(pf_rulequeue, pf_rule);
@@ -1123,7 +1130,8 @@
 #define PFRES_MAXSTATES	12		/* State limit */
 #define PFRES_SRCLIMIT	13		/* Source node/conn limit */
 #define PFRES_SYNPROXY	14		/* SYN proxy */
-#define PFRES_MAX	15		/* total+1 */
+#define	PFRES_DIVERT	15		/* Divert override */
+#define PFRES_MAX	16		/* total+1 */
 
 #define PFRES_NAMES { \
 	"match", \
@@ -1141,6 +1149,7 @@
 	"state-limit", \
 	"src-limit", \
 	"synproxy", \
+	"divert", \
 	NULL \
 }
 
Index: sys/netinet/ip_divert.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/netinet/ip_divert.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 ip_divert.c
--- sys/netinet/ip_divert.c	7 Feb 2011 23:19:28 -0000	1.1.1.1
+++ sys/netinet/ip_divert.c	7 Feb 2011 23:23:18 -0000
@@ -289,8 +289,7 @@
 		 * this iface name will come along for the ride.
 		 * (see div_output for the other half of this.)
 		 */ 
-		strlcpy(divsrc.sin_zero, m->m_pkthdr.rcvif->if_xname,
-		    sizeof(divsrc.sin_zero));
+		*((u_short *)divsrc.sin_zero) = m->m_pkthdr.rcvif->if_index;
 	}
 
 	/* Put packet on socket queue, if any */
@@ -364,7 +363,7 @@
 
 	/* Loopback avoidance and state recovery */
 	if (sin) {
-		int i;
+		u_short idx;
 
 		/* set the starting point. We provide a non-zero slot,
 		 * but a non_matching chain_id to skip that info and use
@@ -372,7 +371,7 @@
 		 */
 		dt->slot = 1; /* dummy, chain_id is invalid */
 		dt->chain_id = 0;
-		dt->rulenum = sin->sin_port+1; /* host format ? */
+		dt->rulenum = sin->sin_port; /* host format ? */
 		dt->rule_id = 0;
 		/*
 		 * Find receive interface with the given name, stuffed
@@ -380,10 +379,9 @@
 		 * The name is user supplied data so don't trust its size
 		 * or that it is zero terminated.
 		 */
-		for (i = 0; i < sizeof(sin->sin_zero) && sin->sin_zero[i]; i++)
-			;
-		if ( i > 0 && i < sizeof(sin->sin_zero))
-			m->m_pkthdr.rcvif = ifunit(sin->sin_zero);
+		idx = *((u_short *)sin->sin_zero);
+		if ( idx > 0 )
+			m->m_pkthdr.rcvif = ifnet_byindex(idx);
 	}
 
 	/* Reinject packet into the system as incoming or outgoing */
@@ -812,5 +810,4 @@
 };
 
 DECLARE_MODULE(ipdivert, ipdivertmod, SI_SUB_PROTO_IFATTACHDOMAIN, SI_ORDER_ANY);
-MODULE_DEPEND(ipdivert, ipfw, 2, 2, 2);
 MODULE_VERSION(ipdivert, 1);
Index: sys/netinet/ipfw/ip_dummynet.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/netinet/ipfw/ip_dummynet.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 ip_dummynet.c
--- sys/netinet/ipfw/ip_dummynet.c	16 Feb 2012 15:19:13 -0000	1.1.1.1
+++ sys/netinet/ipfw/ip_dummynet.c	16 Feb 2012 15:34:29 -0000
@@ -2290,7 +2290,6 @@
 #define	DN_SI_SUB	SI_SUB_PROTO_IFATTACHDOMAIN
 #define	DN_MODEV_ORD	(SI_ORDER_ANY - 128) /* after ipfw */
 DECLARE_MODULE(dummynet, dummynet_mod, DN_SI_SUB, DN_MODEV_ORD);
-MODULE_DEPEND(dummynet, ipfw, 2, 2, 2);
 MODULE_VERSION(dummynet, 3);
 
 /*
