Index: net/if_ethersubr.c
===================================================================
RCS file: /home/ncvs/src/sys/net/if_ethersubr.c,v
retrieving revision 1.270.2.1
diff -u -r1.270.2.1 if_ethersubr.c
--- net/if_ethersubr.c	3 Aug 2009 08:13:06 -0000	1.270.2.1
+++ net/if_ethersubr.c	26 Oct 2009 01:02:45 -0000
@@ -967,8 +968,6 @@
 	for (i = 0; i < ifp->if_addrlen; i++)
 		if (lla[i] != 0)
 			break; 
-	if (i != ifp->if_addrlen)
-		if_printf(ifp, "Ethernet address: %6D\n", lla, ":");
 }
 
 /*
