Index: sys/netipsec/ipsec.c
===================================================================
RCS file: /root/freebsd/test/repo/src/sys/netipsec/ipsec.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 ipsec.c
--- sys/netipsec/ipsec.c	2 Apr 2010 00:52:25 -0000	1.1.1.1
+++ sys/netipsec/ipsec.c	2 Apr 2010 01:06:58 -0000
@@ -592,7 +592,7 @@
 	IPSEC_ASSERT(m->m_pkthdr.len >= sizeof(struct ip),("packet too short"));
 
 	/* NB: ip_input() flips it into host endian. XXX Need more checking. */
-	if (m->m_len < sizeof (struct ip)) {
+	if (m->m_len >= sizeof (struct ip)) {
 		struct ip *ip = mtod(m, struct ip *);
 		if (ip->ip_off & (IP_MF | IP_OFFMASK))
 			goto done;
