Index: sys/netgraph/ng_tcpmss.c
===================================================================
RCS file: /home/ncvs/src/sys/netgraph/ng_tcpmss.c,v
retrieving revision 1.5.2.2.2.1
diff -u -r1.5.2.2.2.1 ng_tcpmss.c
--- sys/netgraph/ng_tcpmss.c	14 Jun 2010 02:09:06 -0000	1.5.2.2.2.1
+++ sys/netgraph/ng_tcpmss.c	28 Jul 2011 15:48:54 -0000
@@ -341,6 +341,9 @@
 	if (correct_mss(tcp, tcphlen, priv->stats.maxMSS,
 	    m->m_pkthdr.csum_flags))
 		priv->stats.FixedPkts++;
+#ifdef __amd64__
+	be16enc(&ip->ip_sum, ip->ip_sum);
+#endif
 
 send:
 	/* Deliver frame out destination hook. */
