diff --git a/contrib/pf/pfctl/pf_print_state.c b/contrib/pf/pfctl/pf_print_state.c
index aa57f57..2fafbd1 100644
--- a/contrib/pf/pfctl/pf_print_state.c
+++ b/contrib/pf/pfctl/pf_print_state.c
@@ -294,6 +294,10 @@ print_state(struct pf_state *s, int opts)
 			printf(", anchor %u", s->anchor.nr);
 		if (s->rule.nr != -1)
 			printf(", rule %u", s->rule.nr);
+		if (s->pdnpipe)
+			printf(", pdnpipe %u", s->pdnpipe);
+		if (s->dnpipe)
+			printf(", dnpipe %u", s->dnpipe);
 		if (s->state_flags & PFSTATE_SLOPPY)
 			printf(", sloppy");
 		if (s->src_node != NULL)
diff --git a/sys/contrib/pf/net/pf.c b/sys/contrib/pf/net/pf.c
index da5a817..19db6d0 100644
--- a/sys/contrib/pf/net/pf.c
+++ b/sys/contrib/pf/net/pf.c
@@ -3436,6 +3436,12 @@ pf_rule_to_actions(struct pf_rule *r, struct pf_rule_actions *a)
 		a->qid = r->qid;
 	if (r->pqid)
 		a->pqid = r->pqid;
+	if (r->pdnpipe)
+		a->pdnpipe = r->pdnpipe;
+	if (r->dnpipe)
+		a->dnpipe = r->dnpipe;
+	if (r->free_flags & PFRULE_DN_IS_PIPE)
+		a->flags |= PFRULE_DN_IS_PIPE;
 }
 
 int
@@ -3733,6 +3739,9 @@ cleanup:
 		s->proto = IPPROTO_TCP;
 		s->qid = act.qid;
 		s->pqid = act.pqid;
+		s->pdnpipe = act.pdnpipe;
+		s->dnpipe = act.dnpipe;
+		s->state_flags |= act.flags;
 		s->direction = direction;
 		s->af = af;
 		if (direction == PF_OUT) {
@@ -4152,6 +4161,9 @@ cleanup:
 		s->proto = IPPROTO_UDP;
 		s->qid = act.qid;
                 s->pqid = act.pqid;
+		s->pdnpipe = act.pdnpipe;
+		s->dnpipe = act.dnpipe;
+		s->state_flags |= act.flags;
 		s->direction = direction;
 		s->af = af;
 		if (direction == PF_OUT) {
@@ -4487,6 +4499,9 @@ cleanup:
 		s->proto = pd->proto;
 		s->qid = act.qid;
                 s->pqid = act.pqid;
+		s->pdnpipe = act.pdnpipe;
+		s->dnpipe = act.dnpipe;
+		s->state_flags |= act.flags;
 		s->direction = direction;
 		s->af = af;
 		if (direction == PF_OUT) {
@@ -4796,6 +4811,9 @@ cleanup:
 		s->proto = pd->proto;
 		s->qid = act.qid;
                 s->pqid = act.pqid;
+		s->pdnpipe = act.pdnpipe;
+		s->dnpipe = act.dnpipe;
+		s->state_flags |= act.flags;
 		s->direction = direction;
 		s->af = af;
 		if (direction == PF_OUT) {
@@ -7743,11 +7761,15 @@ done:
 				return (action);
 			}
 		}
-       	} else if ((r->dnpipe || (divflags & PFSTATE_DIVERT_DNCOOKIE)) &&
+       	} else if (((s && s->dnpipe) || r->dnpipe || (divflags & PFSTATE_DIVERT_DNCOOKIE)) &&
 	    ip_dn_io_ptr != NULL && loopedfrom != 1) {
 
 			if (divflags & PFSTATE_DIVERT_DNCOOKIE) {
 				dnflow.rule.info = divertcookie;
+                        } else if (dir != r->direction && s && s->pdnpipe) {
+                                dnflow.rule.info = s->pdnpipe;
+                        } else if (dir == r->direction && s && s->dnpipe) {
+                                dnflow.rule.info = s->dnpipe;
                         } else if (dir != r->direction && r->pdnpipe) {
                                 dnflow.rule.info = r->pdnpipe;
                         } else if (dir == r->direction) {
@@ -7757,6 +7779,8 @@ done:
 
 			if (r->free_flags & PFRULE_DN_IS_PIPE)
 				dnflow.rule.info |= IPFW_IS_PIPE;
+			if (s && s->state_flags & PFRULE_DN_IS_PIPE)
+				dnflow.rule.info |= IPFW_IS_PIPE;
                         dnflow.f_id.addr_type = 4; /* IPv4 type */
                         dnflow.f_id.proto = pd.proto;
 			if (dir == PF_OUT && s != NULL && s->nat_rule.ptr != NULL &&
diff --git a/sys/contrib/pf/net/pfvar.h b/sys/contrib/pf/net/pfvar.h
index 4ab8ba9..209fdaf 100644
--- a/sys/contrib/pf/net/pfvar.h
+++ b/sys/contrib/pf/net/pfvar.h
@@ -574,6 +574,9 @@ struct pf_osfp_ioctl {
 struct pf_rule_actions {
 	u_int16_t	qid;
 	u_int16_t	pqid;
+	u_int32_t	pdnpipe;
+	u_int32_t	dnpipe;
+	u_int8_t	flags;
 };
 
 union pf_rule_ptr {
@@ -884,6 +887,8 @@ struct pf_state {
 	u_int32_t	 pfsync_time;
 	u_int16_t	 qid;
 	u_int16_t	 pqid;
+	u_int32_t        pdnpipe;
+	u_int32_t        dnpipe;
 	u_int16_t	 tag;
 	u_int16_t	 divert_cookie;
 };
