diff --git a/contrib/pf/pfctl/pf_print_state.c b/contrib/pf/pfctl/pf_print_state.c
index 2fafbd1..fd7692d 100644
--- a/contrib/pf/pfctl/pf_print_state.c
+++ b/contrib/pf/pfctl/pf_print_state.c
@@ -294,6 +294,10 @@ print_state(struct pf_state *s, int opts)
 			printf(", anchor %u", s->anchor.nr);
 		if (s->rule.nr != -1)
 			printf(", rule %u", s->rule.nr);
+		if (s->pqid)
+			printf(", priqueue %u", s->pqid);
+		if (s->qid)
+			printf(", queue %u", s->qid);
 		if (s->pdnpipe)
 			printf(", pdnpipe %u", s->pdnpipe);
 		if (s->dnpipe)
