diff --git a/sys/contrib/pf/net/pf.c b/sys/contrib/pf/net/pf.c
index faacde1..197f38e 100644
--- a/sys/contrib/pf/net/pf.c
+++ b/sys/contrib/pf/net/pf.c
@@ -6645,6 +6645,34 @@ pf_route(struct mbuf **m, struct pf_rule *r, int dir, struct ifnet *oifp,
 	if (oifp != ifp) {
 		if (in_broadcast(ip->ip_dst, oifp)) /* XXX: LOCKING of address list?! */
 			goto donelocal;
+
+		if (s && r->rt == PF_ROUTETO && pd->nat_rule != NULL &&
+		    r->direction == PF_OUT && r->direction == dir && pd->pf_mtag->routed < 2) {
+			int off = ip->ip_hl << 2;
+			switch (pd->proto) {
+			case IPPROTO_TCP: {
+				struct tcphdr *th = pd->hdr.tcp;
+				pf_change_ap(pd->src, &pd->hdr.tcp->th_sport, pd->ip_sum,
+                            		&pd->hdr.tcp->th_sum, &s->lan.addr,
+                            		s->lan.port, 0, pd->af);
+                		m_copyback(*m, off, sizeof(*th), (caddr_t)th);
+			}	break;
+			case IPPROTO_UDP: {
+				struct udphdr *uh = pd->hdr.udp;
+				pf_change_ap(pd->src, &pd->hdr.udp->uh_sport, pd->ip_sum,
+                            		&pd->hdr.udp->uh_sum, &s->lan.addr,
+                            		s->lan.port, 1, pd->af);
+                		m_copyback(*m, off, sizeof(*uh), (caddr_t)uh);
+			}	break;
+			case IPPROTO_ICMP:
+				/* XXX: If we want to do this for icmp is probably wrong!?! */
+				break;
+			default:
+				pf_change_a(&pd->src->v4.s_addr,
+                                    pd->ip_sum, s->lan.addr.v4.s_addr, 0);
+				break;
+			}
+		}
 #ifdef __FreeBSD__
 		PF_UNLOCK();
 		if (pf_test(PF_OUT, ifp, &m0, NULL, NULL) != PF_PASS) {
