diff --git a/sys/contrib/pf/net/pf.c b/sys/contrib/pf/net/pf.c
index 64f27bf..ea089e7 100644
--- a/sys/contrib/pf/net/pf.c
+++ b/sys/contrib/pf/net/pf.c
@@ -6800,14 +6800,14 @@ pf_route(struct mbuf **m, struct pf_rule *r, int dir, struct ifnet *oifp,
 		goto bad;
 	}
 
-	for (m0 = m1; m0; m0 = m1) {
+	for (; m0; m0 = m1) {
 		m1 = m0->m_nextpkt;
 		m0->m_nextpkt = 0;
 #ifdef __FreeBSD__
 		if (error == 0) {
 			PF_UNLOCK();
 			error = (*ifp->if_output)(ifp, m0, sintosa(dst),
-			    NULL);
+			    ro);
 			PF_LOCK();
 		} else
 #else
