Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.8
diff -u -r1.8 pf.c
--- sys/contrib/pf/net/pf.c	8 Feb 2011 12:25:35 -0000	1.8
+++ sys/contrib/pf/net/pf.c	8 Feb 2011 12:25:46 -0000
@@ -6710,6 +6711,7 @@
 				ip->ip_sum = in_cksum(m0, ip->ip_hl << 2);
 			}
 		}
+		m0->m_flags &= ~(M_PROTOFLAGS);
 		PF_UNLOCK();
 		error = (*ifp->if_output)(ifp, m0, sintosa(dst), ro);
 		PF_LOCK();
@@ -6811,6 +6813,7 @@
 		m0->m_nextpkt = 0;
 #ifdef __FreeBSD__
 		if (error == 0) {
+			m0->m_flags &= ~(M_PROTOFLAGS);
 			PF_UNLOCK();
 			error = (*ifp->if_output)(ifp, m0, sintosa(dst),
 			    NULL);
