Index: proc.h
===================================================================
RCS file: /root/freebsd/repo/src/sys/sys/proc.h,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 proc.h
--- proc.h	16 Feb 2012 15:19:08 -0000	1.1.1.1
+++ proc.h	16 Feb 2012 15:44:07 -0000
@@ -673,7 +673,7 @@
  * We use process IDs <= PID_MAX; PID_MAX + 1 must also fit in a pid_t,
  * as it is used to represent "no process group".
  */
-#define	PID_MAX		99999
+#define	PID_MAX		64000
 #define	NO_PID		100000
 
 #define	SESS_LEADER(p)	((p)->p_session->s_leader == (p))
