Index: sys/contrib/pf/net/pf.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/contrib/pf/net/pf.c,v
retrieving revision 1.9
diff -u -r1.9 pf.c
--- sys/contrib/pf/net/pf.c	25 Nov 2010 15:55:25 -0000	1.9
+++ sys/contrib/pf/net/pf.c	25 Nov 2010 16:00:11 -0000
@@ -6636,6 +6636,7 @@
 				ip->ip_sum = in_cksum(m0, ip->ip_hl << 2);
 			}
 		}
+		m0->m_flags &= ~(M_PROTOFLAGS);
 		PF_UNLOCK();
 		error = (*ifp->if_output)(ifp, m0, sintosa(dst), ro);
 		PF_LOCK();
@@ -6737,6 +6738,7 @@
 		m0->m_nextpkt = 0;
 #ifdef __FreeBSD__
 		if (error == 0) {
+			m0->m_flags &= ~(M_PROTOFLAGS);
 			PF_UNLOCK();
 			error = (*ifp->if_output)(ifp, m0, sintosa(dst),
 			    NULL);
