Index: sys/netinet/if_ether.c
===================================================================
RCS file: /root/freebsd/repo/src/sys/netinet/if_ether.c,v
retrieving revision 1.2
diff -u -r1.2 if_ether.c
--- sys/netinet/if_ether.c	30 May 2011 12:21:38 -0000	1.2
+++ sys/netinet/if_ether.c	30 May 2011 12:32:53 -0000
@@ -188,16 +188,17 @@
 			LLE_REMREF(lle);
 			(void) llentry_free(lle);
 			ARPSTAT_INC(timeouts);
-		} 
+		} else {
 #ifdef DIAGNOSTIC
-		else {
 			struct sockaddr *l3addr = L3_ADDR(lle);
 			log(LOG_INFO, 
 			    "arptimer issue: %p, IPv4 address: \"%s\"\n", lle,
 			    inet_ntoa(
 			        ((const struct sockaddr_in *)l3addr)->sin_addr));
-		}
+		
 #endif
+			LLE_WUNLOCK(lle);
+		}
 	}
 	IF_AFDATA_UNLOCK(ifp);
 	CURVNET_RESTORE();