--- sys/modules/Makefile.orig	2012-05-31 18:14:20.000000000 +0000
+++ sys/modules/Makefile	2012-05-31 18:16:16.000000000 +0000
@@ -309,6 +309,7 @@
 	usb \
 	utopia \
 	${_vesa} \
+	${_virtio} \
 	vge \
 	${_viawd} \
 	vkbd \
@@ -519,6 +520,7 @@
 _s3=		s3
 _twa=		twa
 _vesa=		vesa
+_virtio=	virtio
 _x86bios=	x86bios
 .elif ${MACHINE} == "pc98"
 _canbepm=	canbepm
@@ -613,6 +615,7 @@
 _twa=		twa
 _vesa=		vesa
 _viawd=		viawd
+_virtio=	virtio
 _x86bios=	x86bios
 _wi=		wi
 _wpi=		wpi
