--- usbdevs	2008-02-08 21:43:26.000000000 +0000
+++ usbdevs	2008-02-08 21:43:02.000000000 +0000
@@ -1542,6 +1542,9 @@
 /* Sierra Wireless products */
 product SIERRA MC5720		0x0218	MC5720 Wireless Modem
 
+product SIERRA 875U   	        0x1199  MC5720 Wireless Modem
+product SIERRA 875UA   	        0x6820  MC5720 Wireless Modem
+
 /* Sigmatel products */
 product SIGMATEL I_BEAD100	0x8008	i-Bead 100 MP3 Player
 
