--- src/inline.c.orig	2010-04-19 10:39:22.579837563 +1200
+++ src/inline.c	2010-04-19 10:39:43.146515746 +1200
@@ -360,6 +360,15 @@
 
         if (select(32, &fdset, (fd_set *)NULL, (fd_set *)NULL, (struct timeval *)NULL) == -1)
         {
+            /* Do the signal check. If we don't do this we will
+             * evaluate the signal only when we receive an actual
+             * packet. We don't want to depend on this. */
+            if (SignalCheck())
+            {
+#ifndef SNORT_RELOAD
+                Restart();
+#endif
+            }
             printf("select failed");
             continue;
         }
