Index: hostres_processor_tbl.c
===================================================================
RCS file: /home/ncvs/src/usr.sbin/bsnmpd/modules/snmp_hostres/hostres_processor_tbl.c,v
retrieving revision 1.6
diff -u -r1.6 hostres_processor_tbl.c
--- hostres_processor_tbl.c	14 Jul 2006 09:07:56 -0000	1.6
+++ hostres_processor_tbl.c	14 Aug 2008 16:41:28 -0000
@@ -120,7 +120,7 @@
 processor_getpcpu(struct kinfo_proc *ki_p)
 {
 
-	if (ccpu == 0 || fscale == 0)
+	if (fscale == 0)
 		return (0.0);
 
 #define	fxtofl(fixpt) ((double)(fixpt) / fscale)
