Index: sys/netgraph/netgraph.h
===================================================================
RCS file: /mnt/repo/RELENG_7/sys/netgraph/netgraph.h,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 netgraph.h
--- sys/netgraph/netgraph.h	6 Apr 2009 12:30:19 -0000	1.1.1.1
+++ sys/netgraph/netgraph.h	6 Apr 2009 12:38:22 -0000
@@ -1124,6 +1124,7 @@
 struct	ng_type *ng_findtype(const char *type);
 int	ng_make_node_common(struct ng_type *typep, node_p *nodep);
 int	ng_name_node(node_p node, const char *name);
+node_p  ng_name2noderef(node_p node, const char *name);
 int	ng_newtype(struct ng_type *tp);
 ng_ID_t ng_node2ID(node_p node);
 item_p	ng_package_data(struct mbuf *m, int flags);
Index: sys/netgraph/ng_base.c
===================================================================
RCS file: /mnt/repo/RELENG_7/sys/netgraph/ng_base.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 ng_base.c
--- sys/netgraph/ng_base.c	6 Apr 2009 12:30:19 -0000	1.1.1.1
+++ sys/netgraph/ng_base.c	6 Apr 2009 12:38:22 -0000
@@ -216,7 +216,6 @@
 
 /* Imported, these used to be externally visible, some may go back. */
 void	ng_destroy_hook(hook_p hook);
-node_p	ng_name2noderef(node_p node, const char *name);
 int	ng_path2noderef(node_p here, const char *path,
 	node_p *dest, hook_p *lasthook);
 int	ng_make_node(const char *type, node_p *nodepp);
Index: sys/netgraph/ng_eiface.c
===================================================================
RCS file: /mnt/repo/RELENG_7/sys/netgraph/ng_eiface.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 ng_eiface.c
--- sys/netgraph/ng_eiface.c	6 Apr 2009 12:30:19 -0000	1.1.1.1
+++ sys/netgraph/ng_eiface.c	6 Apr 2009 12:38:51 -0000
@@ -41,6 +41,7 @@
 
 #include <net/if.h>
 #include <net/if_types.h>
+#include <net/if_dl.h>
 #include <net/netisr.h>
 
 #include <netgraph/ng_message.h>
@@ -62,6 +63,13 @@
 	},
 	{
 	  NGM_EIFACE_COOKIE,
+	  NGM_EIFACE_SET_IFNAME,
+	  "setifname",
+	  &ng_parse_string_type,
+	  NULL
+	},
+	{
+	  NGM_EIFACE_COOKIE,
 	  NGM_EIFACE_SET,
 	  "set",
 	  &ng_parse_enaddr_type,
@@ -413,6 +421,11 @@
 	struct ng_mesg *resp = NULL;
 	int error = 0;
 	struct ng_mesg *msg;
+	char *new_name;
+        size_t namelen, onamelen;
+        struct sockaddr_dl *sdl = NULL;
+        struct ifaddr *ifa = NULL;
+	node_p ethernode;
 
 	NGI_GET_MSG(item, msg);
 	switch (msg->header.typecookie) {
@@ -438,6 +451,46 @@
 			}
 			strlcpy(resp->data, ifp->if_xname, IFNAMSIZ);
 			break;
+		case NGM_EIFACE_SET_IFNAME:
+			new_name = (char *)msg->data;
+                        
+                	/* Deny request if interface is UP */
+                	if ((ifp->if_flags & IFF_UP) != 0) {
+                  		error = EBUSY;
+                  		break;
+                        }
+                        
+                	EVENTHANDLER_INVOKE(ifnet_departure_event, ifp);
+
+			ethernode = ng_name2noderef(node, ifp->if_xname);
+                        if (ethernode != NULL)
+                                ng_name_node(ethernode, new_name);
+
+                	strlcpy(ifp->if_xname, new_name, sizeof(ifp->if_xname));
+                	ifa = ifp->if_addr;
+                	IFA_LOCK(ifa);
+                	sdl = (struct sockaddr_dl *)ifa->ifa_addr;
+                	namelen = strlen(new_name) + 1;
+                	onamelen = sdl->sdl_nlen;
+                	/*
+                 	* Move the address if needed.  This is safe because we
+                 	* allocate space for a name of length IFNAMSIZ when we
+                 	* create this in if_attach().
+                	 */
+        	        if (namelen != onamelen) {
+	                        bcopy(sdl->sdl_data + onamelen,
+                        	    sdl->sdl_data + namelen, sdl->sdl_alen);
+                	}
+        	        bcopy(new_name, sdl->sdl_data, namelen);
+	                sdl->sdl_nlen = namelen;
+                	sdl = (struct sockaddr_dl *)ifa->ifa_netmask;
+        	        bzero(sdl->sdl_data, onamelen);
+	                while (namelen != 0)
+                        	sdl->sdl_data[--namelen] = 0xff;
+                	IFA_UNLOCK(ifa);
+
+                	EVENTHANDLER_INVOKE(ifnet_arrival_event, ifp);
+			break;
 
 		case NGM_EIFACE_GET_IFADDRS:
 		    {
Index: sys/netgraph/ng_eiface.h
===================================================================
RCS file: /mnt/repo/RELENG_7/sys/netgraph/ng_eiface.h,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 ng_eiface.h
--- sys/netgraph/ng_eiface.h	6 Apr 2009 12:30:19 -0000	1.1.1.1
+++ sys/netgraph/ng_eiface.h	6 Apr 2009 12:38:23 -0000
@@ -54,6 +54,7 @@
 	NGM_EIFACE_GET_IFNAME = 1,	/* get the interface name */
 	NGM_EIFACE_GET_IFADDRS,		/* returns list of addresses */
 	NGM_EIFACE_SET,			/* set ethernet address */
+	NGM_EIFACE_SET_IFNAME,
 };
 
 #endif /* _NETGRAPH_NG_EIFACE_H_ */
Index: usr.sbin/ngctl/main.c
===================================================================
RCS file: /mnt/repo/RELENG_7/usr.sbin/ngctl/main.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 main.c
--- usr.sbin/ngctl/main.c	6 Apr 2009 12:30:51 -0000	1.1.1.1
+++ usr.sbin/ngctl/main.c	6 Apr 2009 12:39:45 -0000
@@ -216,7 +216,6 @@
 			continue;
 		if ((rtn = DoParseCommand(line)) != 0) {
 			warnx("line %d: error in file", num);
-			return (rtn);
 		}
 	}
 	return (CMDRTN_OK);
