Index: rc.conf
===================================================================
RCS file: /home/ncvs/src/etc/defaults/rc.conf,v
retrieving revision 1.318.2.1.2.2
diff -u -r1.318.2.1.2.2 rc.conf
--- rc.conf	29 Jan 2008 00:22:33 -0000	1.318.2.1.2.2
+++ rc.conf	6 Sep 2008 20:56:58 -0000
@@ -41,7 +41,7 @@
 powerd_enable="NO" 	# Run powerd to lower our power usage.
 powerd_flags=""		# Flags to powerd (if enabled).
 tmpmfs="AUTO"		# Set to YES to always create an mfs /tmp, NO to never
-tmpsize="20m"		# Size of mfs /tmp if created
+tmpsize="50m"		# Size of mfs /tmp if created
 tmpmfs_flags="-S"	# Extra mdmfs options for the mfs /tmp
 varmfs="AUTO"		# Set to YES to always create an mfs /var, NO to never
 varsize="32m"		# Size of mfs /var if created
