Index: sockbuf.h
===================================================================
RCS file: /home/ncvs/src/sys/sys/sockbuf.h,v
retrieving revision 1.5.2.1
diff -u -r1.5.2.1 sockbuf.h
--- sockbuf.h	31 Jul 2008 20:35:44 -0000	1.5.2.1
+++ sockbuf.h	26 Oct 2008 15:40:48 -0000
@@ -37,7 +37,7 @@
 #include <sys/_mutex.h>
 #include <sys/_sx.h>
 
-#define	SB_MAX		(256*1024)	/* default for max chars in sockbuf */
+#define	SB_MAX		(512*1024)	/* default for max chars in sockbuf */
 
 /*
  * Constants for sb_flags field of struct sockbuf.
