--- ata-chipset.c.orig	Thu Mar 16 21:28:51 2006
+++ ata-chipset.c	Mon May 15 02:06:36 2006
@@ -288,15 +288,34 @@
 static int
 ata_sata_connect(struct ata_channel *ch)
 {
-    u_int32_t status;
-    int timeout;
+    u_int32_t status, error;
+    int timeout, device;
+    struct ata_pci_controller *ctlr = device_get_softc(device_get_parent(ch->dev));
+    int something_there = 0;
 
+    if (ctlr->chip->chipid == ATA_I82801GB_S1) {
+       status = ATA_IDX_INL(ch, ATA_SSTATUS);
+       error = ATA_IDX_INL(ch, ATA_SERROR);
+       if (status == (ATA_SS_DET_DEV_PRESENT | ATA_SS_DET_PHY_OFFLINE) ||
+           status == (ATA_SS_DET_PHY_ONLINE | ATA_SS_SPD_GEN1 | ATA_SS_IPM_ACTIVE)) {
+           /* reset port */
+           device = 1 << (ch->unit + 1);
+           pci_write_config(ch->dev, 0x92,
+               pci_read_config(ch->dev, 0x92, 2) & ~device, 2);
+           pci_write_config(ch->dev, 0x92,
+               pci_read_config(ch->dev, 0x92, 2) & ~device, 2);
+       }
+    }
     /* wait up to 1 second for "connect well" */
     for (timeout = 0; timeout < 100 ; timeout++) {
 	status = ATA_IDX_INL(ch, ATA_SSTATUS);
+	error = ATA_IDX_INL(ch, ATA_SERROR);
+	ATA_IDX_OUTL(ch, ATA_SERROR, error);
 	if ((status & ATA_SS_CONWELL_MASK) == ATA_SS_CONWELL_GEN1 ||
-	    (status & ATA_SS_CONWELL_MASK) == ATA_SS_CONWELL_GEN2)
+	    (status & ATA_SS_CONWELL_MASK) == ATA_SS_CONWELL_GEN2) {	
+	    something_there = 1;
 	    break;
+	}
 	ata_udelay(10000);
     }
     if (timeout >= 100) {
@@ -319,7 +338,7 @@
     }
     if (bootverbose)
 	device_printf(ch->dev, "SATA connect ready time=%dms\n", timeout * 10);
-    if (timeout < 1000) {
+    if (something_there || timeout < 1000) {
 	if ((ATA_IDX_INB(ch, ATA_CYL_LSB) == ATAPI_MAGIC_LSB) &&
 	    (ATA_IDX_INB(ch, ATA_CYL_MSB) == ATAPI_MAGIC_MSB))
 	    ch->devices = ATA_ATAPI_MASTER;
@@ -1650,6 +1669,12 @@
 	/* force all ports active "the legacy way" */
 	pci_write_config(dev, 0x92, pci_read_config(dev, 0x92, 2) | 0x0f,2);
 
+           if (ctlr->chip->chipid == ATA_I82801GB_S1) {
+               /* enable AHCI register compat mode */
+               pci_write_config(dev, 0x94, pci_read_config(dev, 0x94, 4) | 1 << 9, 4);
+               ATA_OUTL(ctlr->r_res2, 0x0C, ATA_INL(ctlr->r_res2, 0x0C) | 0xf);
+           }
+
 	ctlr->allocate = ata_intel_allocate;
 	ctlr->reset = ata_intel_reset;
 
@@ -1721,8 +1746,10 @@
     struct ata_channel *ch = device_get_softc(dev);
     int mask, timeout;
 
-    /* ICH6 & ICH7 in compat mode has 4 SATA ports as master/slave on 2 ch's */
-    if (ctlr->chip->cfg1) {
+    /* ICH6/7 has 4 SATA ports as master/slave on 2 channels so deal with pairs */
+    if (ctlr->chip->cfg1 || ctlr->chip->chipid == ATA_I82801FB_M || ctlr->chip->chipid == ATA_I82801GB_S1 ||
+	 ctlr->chip->chipid == ATA_I82801GB_R1 ||
+	 ctlr->chip->chipid == ATA_I82801GB_M) {
 	mask = (0x0005 << ch->unit);
     }
     else {
