--- contrib/bsnmp/snmpd/main.c.orig	Wed Dec 28 19:44:25 2005
+++ contrib/bsnmp/snmpd/main.c	Wed Dec 28 19:46:19 2005
@@ -1457,8 +1457,8 @@
 	/*
 	 * Get standard communities
 	 */
-	(void)comm_define(1, "SNMP read", NULL, "public");
-	(void)comm_define(2, "SNMP write", NULL, "public");
+	(void)comm_define(1, "SNMP read", NULL, NULL);
+	(void)comm_define(2, "SNMP write", NULL, NULL);
 	community = COMM_INITIALIZE;
 
 	trap_reqid = reqid_allocate(512, NULL);
