Index: ip_carp.h
===================================================================
RCS file: /home/ncvs/src/sys/netinet/ip_carp.h,v
retrieving revision 1.1
diff -u -r1.1 ip_carp.h
--- ip_carp.h	22 Feb 2005 13:04:03 -0000	1.1
+++ ip_carp.h	10 Aug 2006 17:31:06 -0000
@@ -1,4 +1,4 @@
-/*	$FreeBSD: src/sys/netinet/ip_carp.h,v 1.1 2005/02/22 13:04:03 glebius Exp $	*/
+/*	$FreeBSD: src/sys/netinet/ip_carp.h,v 1.1.4.1 2006/08/10 10:10:12 glebius Exp $	*/
 /*	$OpenBSD: ip_carp.h,v 1.8 2004/07/29 22:12:15 mcbride Exp $	*/
 
 /*
@@ -148,7 +148,6 @@
 }
 
 #ifdef _KERNEL
-void		 carp_ifdetach (struct ifnet *);
 void		 carp_carpdev_state(void *);
 void		 carp_input (struct mbuf *, int);
 int		 carp6_input (struct mbuf **, int *, int);
