Index: pfctl_altq.c
===================================================================
RCS file: /cvs/src/sbin/pfctl/pfctl_altq.c,v
retrieving revision 1.81
diff -u -r1.81 pfctl_altq.c
--- pfctl_altq.c	10 Feb 2004 22:26:56 -0000	1.81
+++ pfctl_altq.c	16 Feb 2004 16:19:08 -0000
@@ -338,8 +338,9 @@
 		pa->qlimit = DEFAULT_QLIMIT;
 
 	if (pa->scheduler == ALTQT_CBQ || pa->scheduler == ALTQT_HFSC) {
-		if ((pa->bandwidth = eval_bwspec(bw,
-		    parent == NULL ? 0 : parent->bandwidth)) == 0) {
+		pa->bandwidth = eval_bwspec(bw,
+		    parent == NULL ? 0 : parent->bandwidth);
+		if (pa->scheduler == ALTQT_CBQ && pa->bandwidth == 0) {
 			fprintf(stderr, "bandwidth for %s invalid (%d / %d)\n",
 			    pa->qname, bw->bw_absolute, bw->bw_percent);
 			return (1);

